

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bioat.crisprtools &mdash; bioat Documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=48106516"></script>
      <script src="../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            bioat
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">Modules</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">bioat</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bioat.crisprtools</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bioat.crisprtools</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">crisprtools.py</span>

<span class="sd">This module provides a toolbox for mining CRISPR-related sequences in metagenomic data.</span>
<span class="sd">It includes functionality for identifying Cas candidates associated with CRISPR loci</span>
<span class="sd">and for annotating Cas13 candidates from protein sequences. The main class, `CrisprTools`,</span>
<span class="sd">provides methods for calling external executables such as Prodigal and Pilercr for protein</span>
<span class="sd">prediction and CRISPR identification.</span>

<span class="sd">Classes:</span>
<span class="sd">    CrisprTools: A class containing methods for identifying Cas candidates and Cas13</span>
<span class="sd">                 candidates from genomic data.</span>

<span class="sd">Methods:</span>
<span class="sd">    cas_finder: A method for de novo annotation of Cas candidates from CRISPR loci,</span>
<span class="sd">                utilizing input fasta files and producing various output files.</span>
<span class="sd">    cas13_finder: A method for the annotation of Cas13 candidates from protein</span>
<span class="sd">                  fasta files.</span>

<span class="sd">Usage:</span>
<span class="sd">    To use this module, create an instance of `CrisprTools` and call the methods</span>
<span class="sd">    `cas_finder` or `cas13_finder` with the appropriate parameters to perform</span>
<span class="sd">    CRISPR analysis on your dataset.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.lib.libfastx</span><span class="w"> </span><span class="kn">import</span> <span class="n">cas13_finder</span><span class="p">,</span> <span class="n">cas_finder</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.lib.libpath</span><span class="w"> </span><span class="kn">import</span> <span class="n">check_executable</span>


<div class="viewcode-block" id="CrisprTools">
<a class="viewcode-back" href="../../bioat.html#bioat.crisprtools.CrisprTools">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">CrisprTools</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;CRISPR mining toolbox.</span>

<span class="sd">    This class provides methods for performing CRISPR analysis on datasets, including</span>
<span class="sd">    finding Cas proteins and CRISPR sequences.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        None</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">pass</span>

<div class="viewcode-block" id="CrisprTools.cas_finder">
<a class="viewcode-back" href="../../bioat.html#bioat.crisprtools.CrisprTools.cas_finder">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">cas_finder</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">input_fa</span><span class="p">,</span>
        <span class="n">output_faa</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">output_contig_fa</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">output_crispr_info_tab</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">lmin</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span>  <span class="c1"># 3001 in Nature Methods paper</span>
        <span class="n">lmax</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">extend</span><span class="o">=</span><span class="mi">10_000</span><span class="p">,</span>
        <span class="n">temp_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">prodigal</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">prodigal_mode</span><span class="o">=</span><span class="s2">&quot;meta&quot;</span><span class="p">,</span>
        <span class="n">pilercr</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">rm_temp</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;INFO&quot;</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;De novo annotation for Cas candidates from neighbors of CRISPR loci.</span>

<span class="sd">        Args:</span>
<span class="sd">            input_fa (str): Path to the input metagenome fasta file containing many contigs.</span>
<span class="sd">            output_faa (str, optional): Path to save the de novo annotated Cas candidates.</span>
<span class="sd">            output_contig_fa (str, optional): Path to save the whole contigs of de novo annotated Cas candidates.</span>
<span class="sd">            output_crispr_info_tab (str, optional): Path to save the de novo annotated CRISPR info table (CSV format).</span>
<span class="sd">            lmin (int, optional): Minimum length for a contig. Defaults to 3000.</span>
<span class="sd">            lmax (int, optional): Maximum length for a contig. Defaults to None.</span>
<span class="sd">            extend (int, optional): Distance over which proteins are considered from the start/end of the CRISPR loci. Defaults to 10000.</span>
<span class="sd">            temp_dir (str, optional): Directory to store temporary files. Defaults to None.</span>
<span class="sd">            prodigal (str, optional): Path to the Prodigal executable. Defaults to None.</span>
<span class="sd">            prodigal_mode (str, optional): Mode for Prodigal annotation. Can be &quot;meta&quot; or &quot;single&quot;. Defaults to &quot;meta&quot;.</span>
<span class="sd">            pilercr (str, optional): Path to the Pilercr executable. Defaults to None.</span>
<span class="sd">            rm_temp (bool, optional): If False, temporary files will be kept. Defaults to True.</span>
<span class="sd">            log_level (str, optional): Logging level; set to &quot;DEBUG&quot; to see logs from prodigal and pilercr. Defaults to &quot;INFO&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">check_executable</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">prodigal</span><span class="p">)</span>
        <span class="n">check_executable</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">pilercr</span><span class="p">)</span>
        <span class="n">cas_finder</span><span class="p">(</span>
            <span class="n">input_fa</span><span class="o">=</span><span class="n">input_fa</span><span class="p">,</span>
            <span class="n">output_faa</span><span class="o">=</span><span class="n">output_faa</span><span class="p">,</span>
            <span class="n">output_contig_fa</span><span class="o">=</span><span class="n">output_contig_fa</span><span class="p">,</span>
            <span class="n">output_crispr_info_tab</span><span class="o">=</span><span class="n">output_crispr_info_tab</span><span class="p">,</span>
            <span class="n">lmin</span><span class="o">=</span><span class="n">lmin</span><span class="p">,</span>
            <span class="n">lmax</span><span class="o">=</span><span class="n">lmax</span><span class="p">,</span>
            <span class="n">extend</span><span class="o">=</span><span class="n">extend</span><span class="p">,</span>
            <span class="n">temp_dir</span><span class="o">=</span><span class="n">temp_dir</span><span class="p">,</span>
            <span class="n">prodigal</span><span class="o">=</span><span class="n">prodigal</span><span class="p">,</span>
            <span class="n">prodigal_mode</span><span class="o">=</span><span class="n">prodigal_mode</span><span class="p">,</span>
            <span class="n">pilercr</span><span class="o">=</span><span class="n">pilercr</span><span class="p">,</span>
            <span class="n">rm_temp</span><span class="o">=</span><span class="n">rm_temp</span><span class="p">,</span>
            <span class="n">log_level</span><span class="o">=</span><span class="n">log_level</span><span class="p">,</span>
        <span class="p">)</span></div>


    <span class="c1"># def cas12_finder(self):</span>
    <span class="c1">#     &quot;&quot;&quot;Cas12 mining toolbox.</span>
    <span class="c1">#</span>
    <span class="c1">#     Under development!</span>
    <span class="c1">#     &quot;&quot;&quot;</span>
    <span class="c1">#     pass</span>

<div class="viewcode-block" id="CrisprTools.cas13_finder">
<a class="viewcode-back" href="../../bioat.html#bioat.crisprtools.CrisprTools.cas13_finder">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">cas13_finder</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">input_faa</span><span class="p">,</span> <span class="n">output_faa</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">lmin</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">lmax</span><span class="o">=</span><span class="mi">1500</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;INFO&quot;</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;De novo annotation for Cas13 candidates from proteins.faa.</span>

<span class="sd">        Args:</span>
<span class="sd">            input_faa (str): The input file containing Cas candidates in .faa format.</span>
<span class="sd">            output_faa (str, optional): The output file for Cas13 candidates in .faa format. Defaults to None.</span>
<span class="sd">            lmin (int, optional): Minimum length for a Cas candidate. Defaults to 200.</span>
<span class="sd">            lmax (int, optional): Maximum length for a Cas candidate. Defaults to 1500.</span>
<span class="sd">            log_level (str, optional): The logging level. Options are &#39;CRITICAL&#39;, &#39;ERROR&#39;, &#39;WARNING&#39;, &#39;INFO&#39;, &#39;DEBUG&#39;, &#39;NOTSET&#39;. Defaults to &#39;INFO&#39;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">cas13_finder</span><span class="p">(</span>
            <span class="n">input_faa</span><span class="o">=</span><span class="n">input_faa</span><span class="p">,</span>
            <span class="n">output_faa</span><span class="o">=</span><span class="n">output_faa</span><span class="p">,</span>
            <span class="n">lmin</span><span class="o">=</span><span class="n">lmin</span><span class="p">,</span>
            <span class="n">lmax</span><span class="o">=</span><span class="n">lmax</span><span class="p">,</span>
            <span class="n">log_level</span><span class="o">=</span><span class="n">log_level</span><span class="p">,</span>
        <span class="p">)</span></div>
</div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="k">pass</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>