

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bioat.foldtools &mdash; bioat Documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=241ed0f2"></script>
      <script src="../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            bioat
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">Modules</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">bioat</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bioat.foldtools</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bioat.foldtools</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">Bio</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.lib.libpdb</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_cut2ref_aln_info</span><span class="p">,</span> <span class="n">pdb2fasta</span><span class="p">,</span> <span class="n">show_ref_cut</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.logger</span><span class="w"> </span><span class="kn">import</span> <span class="n">LoggerManager</span>

<span class="n">lm</span> <span class="o">=</span> <span class="n">LoggerManager</span><span class="p">(</span><span class="n">mod_name</span><span class="o">=</span><span class="s2">&quot;bioat.foldtools&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="FoldTools">
<a class="viewcode-back" href="../../bioat.html#bioat.foldtools.FoldTools">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">FoldTools</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Folding toolbox.&quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">pass</span>

<div class="viewcode-block" id="FoldTools.show_ref_cut">
<a class="viewcode-back" href="../../bioat.html#bioat.foldtools.FoldTools.show_ref_cut">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">show_ref_cut</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">ref_seq</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Seq</span><span class="p">,</span>
        <span class="n">ref_pdb</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Bio</span><span class="o">.</span><span class="n">PDB</span><span class="o">.</span><span class="n">Structure</span><span class="o">.</span><span class="n">Structure</span><span class="p">,</span>
        <span class="n">cut_seq</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Seq</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">cut_pdb</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Bio</span><span class="o">.</span><span class="n">PDB</span><span class="o">.</span><span class="n">Structure</span><span class="o">.</span><span class="n">Structure</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ref_color</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;blue&quot;</span><span class="p">,</span>
        <span class="n">ref_map_colors</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ref_map_values</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">cut_color</span><span class="o">=</span><span class="s2">&quot;green&quot;</span><span class="p">,</span>
        <span class="n">gap_color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span>
        <span class="n">ref_style</span><span class="o">=</span><span class="s2">&quot;cartoon&quot;</span><span class="p">,</span>
        <span class="n">cut_style</span><span class="o">=</span><span class="s2">&quot;cartoon&quot;</span><span class="p">,</span>
        <span class="n">gap_style</span><span class="o">=</span><span class="s2">&quot;cartoon&quot;</span><span class="p">,</span>
        <span class="n">ref_map_value_random</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">output_fig</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">col</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
        <span class="n">scale</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
        <span class="n">annotate</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">text_interval</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
        <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Visualizes the alignment of sequences and highlights changes in PDB structures using py3Dmol.</span>

<span class="sd">        Args:</span>
<span class="sd">            ref_seq (str or Seq): Amino acid sequence content for the ref protein.</span>
<span class="sd">            ref_pdb (str or Bio.PDB.Structure.Structure): Path to the PDB file of the reference structure.</span>
<span class="sd">            cut_seq (str, Seq or None, optional): Amino acid sequence content for the cut protein.</span>
<span class="sd">            cut_pdb (str, Bio.PDB.Structure.Structure or None, optional): Path to the PDB file of the cut structure.</span>
<span class="sd">            ref_color (str, optional): Color for reference residues.</span>
<span class="sd">            ref_map_colors (tuple[str, str] or None, optional): ref_map_colors will be used as color bar from ref_map_colors[0] to ref_map_colors[1]. If None, do not apply color mapping. Defaults to None.</span>
<span class="sd">            ref_map_values (dict or None, optional): A dictionary of values for the ref color map, it will be normalized to the range of [0 - 1]. If None, all residues will be colored with the same color. e.g. ref_value_dict = {&#39;V_0&#39;: 0.4177215189873418, &#39;S_1&#39;: 0.8185654008438819, &#39;K_2&#39;: 0.9915611814345991, &#39;G_3&#39;: 0.42616033755274263, ...}</span>
<span class="sd">            cut_color (str, optional): Color for cut residues.</span>
<span class="sd">            gap_color (str, optional): Color for gaps or removed residues.</span>
<span class="sd">            ref_style (str, optional): &quot;stick&quot;, &quot;sphere&quot;, &quot;cartoon&quot;, or &quot;line&quot;</span>
<span class="sd">            cut_style (str, optional): &quot;stick&quot;, &quot;sphere&quot;, &quot;cartoon&quot;, or &quot;line&quot;</span>
<span class="sd">            gap_style (str, optional): &quot;stick&quot;, &quot;sphere&quot;, &quot;cartoon&quot;, or &quot;line&quot;</span>
<span class="sd">            ref_map_value_random (bool, optional): If True, ref_value_dict will be randomly generated. Defaults to False.</span>
<span class="sd">            output_fig (str or None, optional): Output figure file path. If None, the figure will not be saved in html format. Defaults to None.</span>
<span class="sd">            col (int, optional): Number of columns for the visualization. Defaults to 3.</span>
<span class="sd">            scale (float, optional): Scale factor for the visualization. Defaults to 1.0.</span>
<span class="sd">            annotate (bool, optional): Whether to annotate the visualization with labels. Defaults to True.</span>
<span class="sd">            text_interval (int, optional): The interval between text annotations. Defaults to 5.</span>
<span class="sd">            log_level (str, optional): Log level. Defaults to &quot;WARNING&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">show_ref_cut</span><span class="p">(</span>
            <span class="n">ref_seq</span><span class="o">=</span><span class="n">ref_seq</span><span class="p">,</span>
            <span class="n">cut_seq</span><span class="o">=</span><span class="n">cut_seq</span><span class="p">,</span>
            <span class="n">ref_pdb</span><span class="o">=</span><span class="n">ref_pdb</span><span class="p">,</span>
            <span class="n">cut_pdb</span><span class="o">=</span><span class="n">cut_pdb</span><span class="p">,</span>
            <span class="n">ref_color</span><span class="o">=</span><span class="n">ref_color</span><span class="p">,</span>
            <span class="n">ref_map_colors</span><span class="o">=</span><span class="n">ref_map_colors</span><span class="p">,</span>
            <span class="n">ref_map_values</span><span class="o">=</span><span class="n">ref_map_values</span><span class="p">,</span>
            <span class="n">cut_color</span><span class="o">=</span><span class="n">cut_color</span><span class="p">,</span>
            <span class="n">gap_color</span><span class="o">=</span><span class="n">gap_color</span><span class="p">,</span>
            <span class="n">ref_style</span><span class="o">=</span><span class="n">ref_style</span><span class="p">,</span>
            <span class="n">cut_style</span><span class="o">=</span><span class="n">cut_style</span><span class="p">,</span>
            <span class="n">gap_style</span><span class="o">=</span><span class="n">gap_style</span><span class="p">,</span>
            <span class="n">ref_map_value_random</span><span class="o">=</span><span class="n">ref_map_value_random</span><span class="p">,</span>
            <span class="n">output_fig</span><span class="o">=</span><span class="n">output_fig</span><span class="p">,</span>
            <span class="n">col</span><span class="o">=</span><span class="n">col</span><span class="p">,</span>
            <span class="n">scale</span><span class="o">=</span><span class="n">scale</span><span class="p">,</span>
            <span class="n">annotate</span><span class="o">=</span><span class="n">annotate</span><span class="p">,</span>
            <span class="n">text_interval</span><span class="o">=</span><span class="n">text_interval</span><span class="p">,</span>
            <span class="n">log_level</span><span class="o">=</span><span class="n">log_level</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="FoldTools.pdb2fasta">
<a class="viewcode-back" href="../../bioat.html#bioat.foldtools.FoldTools.pdb2fasta">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">pdb2fasta</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="nb">input</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Converts a PDB file to a FASTA file.</span>

<span class="sd">        This function processes the provided PDB file and extracts protein, DNA,</span>
<span class="sd">        RNA sequences, and other molecules appropriately to create a FASTA file.</span>

<span class="sd">        Details:</span>
<span class="sd">            1. **Proteins**:</span>
<span class="sd">               The protein sequence for each chain will be extracted as &quot;Chain X Protein&quot;.</span>
<span class="sd">            2. **DNA and RNA**:</span>
<span class="sd">               Bases for DNA (A, T, G, C) will be saved as &quot;Chain X DNA&quot;, and bases for RNA (A, U, G, C) will be saved as &quot;Chain X RNA&quot;.</span>
<span class="sd">            3. **Other molecules**:</span>
<span class="sd">               Any unrecognized molecules (e.g., ions, modified molecules) will be labeled as [residue] and stored as &quot;Chain X Other molecules&quot;.</span>
<span class="sd">            4. **Multi-chain complexes**:</span>
<span class="sd">               The program supports multi-chain structures in complexes, and the content of each chain will be recorded separately.</span>

<span class="sd">        Args:</span>
<span class="sd">            input (str):</span>
<span class="sd">                Input file path.</span>
<span class="sd">            output (str, optional):</span>
<span class="sd">                Output file path. If None, the output file will be named as the</span>
<span class="sd">                basename of the input file with a &quot;.fa&quot; extension. Defaults to None.</span>
<span class="sd">            log_level (str, optional):</span>
<span class="sd">                Logging level. Defaults to &quot;WARNING&quot;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            None</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pdb2fasta</span><span class="p">(</span><span class="n">pdb</span><span class="o">=</span><span class="nb">input</span><span class="p">,</span> <span class="n">output_fasta</span><span class="o">=</span><span class="n">output</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="n">log_level</span><span class="p">)</span></div>


<div class="viewcode-block" id="FoldTools.get_cut2ref_aln_info">
<a class="viewcode-back" href="../../bioat.html#bioat.foldtools.FoldTools.get_cut2ref_aln_info">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_cut2ref_aln_info</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">ref</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Bio</span><span class="o">.</span><span class="n">PDB</span><span class="o">.</span><span class="n">Structure</span><span class="o">.</span><span class="n">Structure</span><span class="p">,</span>
        <span class="n">cut</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Bio</span><span class="o">.</span><span class="n">PDB</span><span class="o">.</span><span class="n">Structure</span><span class="o">.</span><span class="n">Structure</span><span class="p">,</span>
        <span class="n">cal_rmsd</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">cal_tmscore</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">label1</span><span class="o">=</span><span class="s2">&quot;ref&quot;</span><span class="p">,</span>
        <span class="n">label2</span><span class="o">=</span><span class="s2">&quot;cut&quot;</span><span class="p">,</span>
        <span class="n">usalign_bin</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;usalign&quot;</span><span class="p">,</span>
        <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Align cutted pdb to ref pdb using the CA atoms.</span>
<span class="sd">        Aligns a truncated protein structure (cut) to its full-length reference structure (ref)</span>
<span class="sd">        using Cα atoms and Biopython&#39;s Superimposer.</span>

<span class="sd">        This function:</span>
<span class="sd">        - Extracts all Cα atoms from `ref` and `cut`</span>
<span class="sd">        - Removes atoms from `ref` at the indices listed in `gap_indices`</span>
<span class="sd">        - Aligns the remaining atoms from `cut` to the corresponding positions in `ref`</span>
<span class="sd">        - Modifies the `cut` structure in-place to match the aligned orientation</span>
<span class="sd">        - Returns both structures and the RMSD value of the alignment</span>

<span class="sd">        It assumes:</span>
<span class="sd">        - One-to-one correspondence between residues after gap removal</span>
<span class="sd">        - Structures are predicted by AlphaFold2 / ESMFold (no missing atoms)</span>

<span class="sd">        Args:</span>
<span class="sd">            ref (str or Bio.PDB.Structure.Structure): Reference structure path or loaded Structure.</span>
<span class="sd">            cut (str or Bio.PDB.Structure.Structure): Truncated structure path or loaded Structure.</span>
<span class="sd">            cal_rmsd (bool, optional): Whether to calculate RMSD. Default is True.</span>
<span class="sd">            cal_tmscore (bool, optional): Whether to calculate TM-score using USalign. Default is False.</span>
<span class="sd">            label1 (str, optional): Name for the reference structure. Default is &quot;ref&quot;.</span>
<span class="sd">            label2 (str, optional): Name for the cut structure. Default is &quot;cut&quot;.</span>
<span class="sd">            usalign_bin (str, optional): Path to the USalign binary for TM-score calculation. Default is &quot;usalign&quot;.</span>
<span class="sd">            log_level (str, optional): Logging level. Default is &quot;WARNING&quot;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: {</span>
<span class="sd">                    &quot;{label1}&quot;: aln label1 structure,  # if cal_rmsd is True, unaltered label1 structure</span>
<span class="sd">                    &quot;{label2}}&quot;: fixed label2 structure,  # if cal_rmsd is True, fix label2 coords in-place</span>
<span class="sd">                    &quot;RMSD&quot;: 0.123  # if cal_rmsd is True, the RMSD value between label1 and label2</span>
<span class="sd">                    f&quot;{label1}_seq&quot;: ref_seq,  # if cal_rmsd is True, the sequence of label1 structure</span>
<span class="sd">                    f&quot;{label2}_seq&quot;: cut_seq,  # if cal_rmsd is True, the sequence of label2 structure</span>
<span class="sd">                    &quot;alignment_dict&quot;: alignment_dict,  # if cal_rmsd is True, the alignment dict of label1 and label2</span>
<span class="sd">                    &quot;gap_indices&quot;: gap_indices,  # if cal_rmsd is True, the indices of gaps in label1 structure</span>
<span class="sd">                    &quot;TM-score:mean&quot;: 0.623,  # if cal_tmscore is True, the mean TM-score value</span>
<span class="sd">                    &quot;TM-score:TM1&quot;: 0.456,  # if cal_tmscore is True, use label1 as ref &lt;L_N&gt; in calculation</span>
<span class="sd">                    &quot;TM-score:TM2&quot;: 0.789,  # if cal_tmscore is True, use label2 as ref &lt;L_N&gt; in calculation</span>
<span class="sd">                    ...</span>
<span class="sd">                }</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">aln_info</span> <span class="o">=</span> <span class="n">get_cut2ref_aln_info</span><span class="p">(</span>
            <span class="n">ref</span><span class="o">=</span><span class="n">ref</span><span class="p">,</span>
            <span class="n">cut</span><span class="o">=</span><span class="n">cut</span><span class="p">,</span>
            <span class="n">cal_rmsd</span><span class="o">=</span><span class="n">cal_rmsd</span><span class="p">,</span>
            <span class="n">cal_tmscore</span><span class="o">=</span><span class="n">cal_tmscore</span><span class="p">,</span>
            <span class="n">label1</span><span class="o">=</span><span class="n">label1</span><span class="p">,</span>
            <span class="n">label2</span><span class="o">=</span><span class="n">label2</span><span class="p">,</span>
            <span class="n">usalign_bin</span><span class="o">=</span><span class="n">usalign_bin</span><span class="p">,</span>
            <span class="n">log_level</span><span class="o">=</span><span class="n">log_level</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">aln_info</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">value</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>