

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bioat.hictools &mdash; bioat Documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=27e171c2"></script>
      <script src="../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../contents.html" class="icon icon-home">
            bioat
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../QuickStart.html">QuickStart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../bioat.html">bioat package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../contents.html">bioat</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../contents.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bioat.hictools</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bioat.hictools</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">gc</span>
<span class="kn">import</span> <span class="nn">time</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">bioat.devtools</span> <span class="kn">import</span> <span class="n">profile</span>
<span class="kn">from</span> <span class="nn">bioat.logger</span> <span class="kn">import</span> <span class="n">LoggerManager</span>

<span class="n">lm</span> <span class="o">=</span> <span class="n">LoggerManager</span><span class="p">(</span><span class="n">mod_name</span><span class="o">=</span><span class="s2">&quot;bioat.hictools&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="HiCTools">
<a class="viewcode-back" href="../../bioat.html#bioat.hictools.HiCTools">[docs]</a>
<span class="k">class</span> <span class="nc">HiCTools</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Hi-C toolbox.&quot;&quot;&quot;</span>

    <span class="n">lm</span><span class="o">.</span><span class="n">set_names</span><span class="p">(</span><span class="n">cls_name</span><span class="o">=</span><span class="s2">&quot;HiCTools&quot;</span><span class="p">)</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">pass</span>

    <span class="nd">@profile</span>
    <span class="k">def</span> <span class="nf">get_effective_resolutions</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">genome_index</span><span class="p">,</span>
            <span class="n">valid_pairs</span><span class="p">,</span>
            <span class="n">output</span><span class="p">,</span>
            <span class="n">log_level</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;INFO&#39;</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get deepest resolution of cis-interation.</span>

<span class="sd">        输入为result文件，输出为matrix的大致resolution.</span>

<span class="sd">        :param genome_index: genome.fa.fai</span>
<span class="sd">        :param valid_pairs: rm_dup_pairs.allValidPairs by HiCTools-Pro</span>
<span class="sd">        :param output: table_output, TSV file</span>
<span class="sd">        :param log_level: &#39;CRITICAL&#39;, &#39;ERROR&#39;, &#39;WARNING&#39;, &#39;INFO&#39;, &#39;DEBUG&#39;, &#39;NOTSET&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">set_names</span><span class="p">(</span><span class="n">func_name</span><span class="o">=</span><span class="s2">&quot;get_effective_resolutions&quot;</span><span class="p">)</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">set_level</span><span class="p">(</span><span class="n">log_level</span><span class="p">)</span>

        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Develop mode is on&quot;</span><span class="p">)</span>
        <span class="c1"># load ref genome lengths</span>
        <span class="n">df_chromosome</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">genome_index</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span>
            <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
            <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chromosome&#39;</span><span class="p">,</span> <span class="s1">&#39;length&#39;</span><span class="p">],</span>
            <span class="n">index_col</span><span class="o">=</span><span class="s1">&#39;chromosome&#39;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># load sample valid pairs by HiCTools-Pro after remove duplication and filter.</span>
        <span class="n">reader</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">valid_pairs</span><span class="p">,</span>
            <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">names</span><span class="o">=</span><span class="p">[</span>
                <span class="s1">&#39;read_name&#39;</span><span class="p">,</span> <span class="s1">&#39;chr_A&#39;</span><span class="p">,</span> <span class="s1">&#39;start_A&#39;</span><span class="p">,</span> <span class="s1">&#39;strand_A&#39;</span><span class="p">,</span> <span class="s1">&#39;chr_B&#39;</span><span class="p">,</span> <span class="s1">&#39;start_B&#39;</span><span class="p">,</span> <span class="s1">&#39;strand_B&#39;</span><span class="p">,</span> <span class="s1">&#39;insert_size&#39;</span><span class="p">,</span>
                <span class="s1">&#39;fragment_name_A&#39;</span><span class="p">,</span> <span class="s1">&#39;fragment_name_B&#39;</span><span class="p">,</span> <span class="s1">&#39;mapQ_A&#39;</span><span class="p">,</span> <span class="s1">&#39;mapQ_B&#39;</span><span class="p">,</span> <span class="s1">&#39;allele_specific_info&#39;</span>
            <span class="p">],</span>
            <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr_A&#39;</span><span class="p">,</span> <span class="s1">&#39;start_A&#39;</span><span class="p">,</span> <span class="s1">&#39;chr_B&#39;</span><span class="p">,</span> <span class="s1">&#39;start_B&#39;</span><span class="p">],</span>
            <span class="n">dtype</span><span class="o">=</span><span class="p">{</span>
                <span class="s1">&#39;chr_A&#39;</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                <span class="s1">&#39;start_A&#39;</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
                <span class="s1">&#39;chr_B&#39;</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                <span class="s1">&#39;start_B&#39;</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
            <span class="p">},</span>
            <span class="n">iterator</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
        <span class="n">chunks</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">chunk</span> <span class="o">=</span> <span class="n">reader</span><span class="o">.</span><span class="n">get_chunk</span><span class="p">(</span><span class="mi">10_000_000</span><span class="p">)</span>
                <span class="n">chunks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">StopIteration</span><span class="p">:</span>
                <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;pandas reader iteration is done.&quot;</span><span class="p">)</span>
                <span class="k">break</span>

        <span class="n">df_valid_pairs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">chunks</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">df_valid_pairs</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">memory_usage</span><span class="o">=</span><span class="s2">&quot;deep&quot;</span><span class="p">))</span>
        <span class="c1"># 思路：二分法确认new_bin下线</span>
        <span class="n">output</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="s1">&#39;wt&#39;</span><span class="p">)</span>
        <span class="n">write_lines</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;test_bin(kb)</span><span class="se">\t</span><span class="s1">total_bins</span><span class="se">\t</span><span class="s1">threshold(%)</span><span class="se">\t</span><span class="s1">is_effective</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">write_lines</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="c1"># 思路：二分法确认new_bin下线</span>
        <span class="n">circle</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">new_bin</span> <span class="o">=</span> <span class="mi">512_000</span>  <span class="c1"># 512kb</span>

        <span class="k">def</span> <span class="nf">get_bin_index</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
            <span class="n">cumsum</span> <span class="o">=</span> <span class="n">df_chromosome</span><span class="p">[</span><span class="s1">&#39;bins_cumsum&#39;</span><span class="p">][</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;chr_A&#39;</span><span class="p">]]</span>
            <span class="n">bin_count_in_this_chrom_a</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;start_A&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">new_bin</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">bin_count_in_this_chrom_b</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;start_B&#39;</span><span class="p">]</span> <span class="o">/</span> <span class="n">new_bin</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">bin_index_a</span> <span class="o">=</span> <span class="n">cumsum</span> <span class="o">+</span> <span class="n">bin_count_in_this_chrom_a</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">bin_index_b</span> <span class="o">=</span> <span class="n">cumsum</span> <span class="o">+</span> <span class="n">bin_count_in_this_chrom_b</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="k">return</span> <span class="n">bin_index_a</span><span class="p">,</span> <span class="n">bin_index_b</span>

        <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
            <span class="c1"># 每次迭代都把new_bin / 2</span>
            <span class="n">new_bin</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">new_bin</span> <span class="o">/</span> <span class="n">circle</span><span class="p">)</span>
            <span class="n">circle</span> <span class="o">=</span> <span class="mi">2</span>

            <span class="c1"># calculate total bins</span>
            <span class="n">df_chromosome</span><span class="p">[</span><span class="s2">&quot;bins&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_chromosome</span><span class="p">[</span><span class="s2">&quot;length&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">new_bin</span><span class="o">=</span><span class="n">new_bin</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span> <span class="o">/</span> <span class="n">new_bin</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="p">)</span>
            <span class="n">df_chromosome</span><span class="p">[</span><span class="s1">&#39;bins_cumsum&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_chromosome</span><span class="p">[</span><span class="s1">&#39;bins&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cumsum</span><span class="p">()</span>
            <span class="n">total_bins</span> <span class="o">=</span> <span class="n">df_chromosome</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>

            <span class="c1"># calculate mapped bin index</span>
            <span class="n">df_valid_pairs</span><span class="p">[[</span><span class="s2">&quot;bin_index_A&quot;</span><span class="p">,</span> <span class="s2">&quot;bin_index_B&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">df_valid_pairs</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                <span class="c1"># df_valid_pairs[[&#39;bin_index_A&#39;, &#39;bin_index_B&#39;]] = df_valid_pairs.apply(</span>
                <span class="n">get_bin_index</span><span class="p">,</span>
                <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                <span class="n">result_type</span><span class="o">=</span><span class="s2">&quot;expand&quot;</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># print(df_valid_pairs)</span>
            <span class="c1"># threshold calculation</span>
            <span class="n">s_bin_idx_count1</span> <span class="o">=</span> <span class="n">df_valid_pairs</span><span class="p">[</span><span class="s1">&#39;bin_index_A&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
            <span class="n">s_bin_idx_count2</span> <span class="o">=</span> <span class="n">df_valid_pairs</span><span class="p">[</span><span class="s1">&#39;bin_index_B&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
            <span class="n">df_bin_interaction_count</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">s_bin_idx_count1</span><span class="p">,</span> <span class="n">s_bin_idx_count2</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">df_bin_interaction_count</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">df_bin_interaction_count</span><span class="p">[</span><span class="s1">&#39;total&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df_bin_interaction_count</span><span class="p">[</span><span class="s1">&#39;bin_index_A&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">df_bin_interaction_count</span><span class="p">[</span>
                <span class="s1">&#39;bin_index_B&#39;</span>
            <span class="p">]</span>
            <span class="n">threshold</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">df_bin_interaction_count</span><span class="p">[</span><span class="s1">&#39;total&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">1000</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="n">total_bins</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">threshold</span> <span class="o">&gt;</span> <span class="mi">80</span><span class="p">:</span>
                <span class="c1"># 输出结果为有效,进入下一个循环</span>
                <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">new_bin</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">1000</span><span class="p">)</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">total_bins</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">threshold</span><span class="si">}</span><span class="se">\t</span><span class="s1">YES</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="n">write_lines</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># if write_lines != 1:</span>
                <span class="c1">#     # 第1+n次计算无效,直接终止程序</span>
                <span class="c1">#     output.write(f&#39;{int(new_bin / 1000)}\t{total_bins}\t{threshold}\tNO\n&#39;)</span>
                <span class="c1"># else:</span>
                <span class="c1">#     # 第一次计算就无效,直接终止程序</span>
                <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="nb">int</span><span class="p">(</span><span class="n">new_bin</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">1000</span><span class="p">)</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">total_bins</span><span class="si">}</span><span class="se">\t</span><span class="si">{</span><span class="n">threshold</span><span class="si">}</span><span class="se">\t</span><span class="s1">NO</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="k">break</span>
            <span class="k">del</span> <span class="n">s_bin_idx_count1</span><span class="p">,</span> <span class="n">s_bin_idx_count2</span><span class="p">,</span> <span class="n">df_bin_interaction_count</span>
            <span class="n">gc</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
        <span class="n">output</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">a</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">hic</span> <span class="o">=</span> <span class="n">HiCTools</span><span class="p">()</span>
    <span class="n">hic</span><span class="o">.</span><span class="n">get_effective_resolutions</span><span class="p">(</span>
        <span class="n">genome_index</span><span class="o">=</span><span class="s2">&quot;/Volumes/zhaohn_HD/Bio/1.database/db_genomes/genome_fa/genome_ucsc_hg38/genome_ucsc_hg38.fa.fai&quot;</span><span class="p">,</span>
        <span class="n">valid_pairs</span><span class="o">=</span><span class="s2">&quot;../../data/hic/test.rm_dup_pairs.allValidPairs&quot;</span><span class="p">,</span>
        <span class="n">output</span><span class="o">=</span><span class="s2">&quot;/Users/zhaohuanan/Downloads/testout&quot;</span>
    <span class="p">)</span>
    <span class="n">a</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">a</span>
    <span class="c1"># print(f&#39;\n{a:.3f}s&#39;)</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>