

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bioat.lib package &mdash; bioat Documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=687cf729"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="bioat package" href="bioat.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            bioat
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">Modules</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="bioat.html">bioat package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="bioat.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">bioat.lib package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.about">bioat.about module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.bamtools">bioat.bamtools module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.bedtools">bioat.bedtools module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.cli">bioat.cli module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.crisprtools">bioat.crisprtools module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.devtools">bioat.devtools module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.exceptions">bioat.exceptions module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.fastxtools">bioat.fastxtools module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.foldtools">bioat.foldtools module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.hictools">bioat.hictools module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.logger">bioat.logger module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.metatools">bioat.metatools module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.searchtools">bioat.searchtools module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.systemtools">bioat.systemtools module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.tabletools">bioat.tabletools module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat.target_seq">bioat.target_seq module</a></li>
<li class="toctree-l3"><a class="reference internal" href="bioat.html#module-bioat">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">bioat</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">Modules</a></li>
          <li class="breadcrumb-item"><a href="bioat.html">bioat package</a></li>
      <li class="breadcrumb-item active">bioat.lib package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/bioat.lib.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bioat-lib-package">
<h1>bioat.lib package<a class="headerlink" href="#bioat-lib-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-bioat.lib.libalignment">
<span id="bioat-lib-libalignment-module"></span><h2>bioat.lib.libalignment module<a class="headerlink" href="#module-bioat.lib.libalignment" title="Link to this heading"></a></h2>
<p>Library for alignment, depends on Biopython</p>
<p>author: Herman Huanan Zhao
email: <a class="reference external" href="mailto:hermanzhaozzzz&#37;&#52;&#48;gmail&#46;com">hermanzhaozzzz<span>&#64;</span>gmail<span>&#46;</span>com</a>
homepage: <a class="reference external" href="https://github.com/hermanzhaozzzz">https://github.com/hermanzhaozzzz</a></p>
<p>Library for alignment, depends on Biopython</p>
<dl>
<dt>example 1:</dt><dd><dl>
<dt>bioat.lib.libalignment</dt><dd><dl>
<dt>&lt;in python consolo&gt;:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">bioat.lib.libalignment</span><span class="w"> </span><span class="kn">import</span> <span class="n">instantiate_pairwise_aligner</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">aligner</span> <span class="o">=</span> <span class="n">instantiate_pairwise_aligner</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">get_best_alignment</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GGCACTGCGGCTGGAAAAAAAAAAAAAAAGT&quot;</span><span class="p">),</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GGCAGCGGCTGGAAAAAAAAAAAAAAAAGGT&quot;</span><span class="p">),</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">aligner</span><span class="o">=</span><span class="n">aligner</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">consider_strand</span><span class="o">=</span><span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">get_aligned_seq</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">get_aligned_seq</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">get_alignment_info</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">res</span> <span class="o">=</span> <span class="n">get_alignment_info</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libalignment.get_aligned_seq">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libalignment.</span></span><span class="sig-name descname"><span class="pre">get_aligned_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letn_match</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="reference internal" href="_modules/bioat/lib/libalignment.html#get_aligned_seq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libalignment.get_aligned_seq" title="Link to this definition"></a></dt>
<dd><p>Retrieve aligned sequence information from a Bio.Align.Alignment object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> (<em>Bio.Align.Alignment</em>) – A <cite>Bio.Align.Alignment</cite> object containing the alignment information.</p></li>
<li><p><strong>reverse</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to return reversed sequence information. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A dictionary containing the alignment details with the following keys:</dt><dd><ul class="simple">
<li><p>’reference_seq’ (str): The aligned reference sequence.</p></li>
<li><p>’aln_info’ (str): The alignment information string, indicating matches (‘|’),
mismatches (‘.’), and gaps (‘-‘).</p></li>
<li><p>’target_seq’ (str): The aligned target sequence.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">get_aligned_seq</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<span class="go">{</span>
<span class="go">    &#39;reference_seq&#39;: &#39;GGCACTGCGGCTGGAAAAAAAAAAAAAAA--GT&#39;,</span>
<span class="go">    &#39;aln_info&#39;:      &#39;--..|.|||||||||||||||||||||||--||&#39;,</span>
<span class="go">    &#39;target_seq&#39;:    &#39;--GGCAGCGGCTGGAAAAAAAAAAAAAAAAGGT&#39;</span>
<span class="go">}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libalignment.get_alignment_info">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libalignment.</span></span><span class="sig-name descname"><span class="pre">get_alignment_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">alignment</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Alignment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letn_match</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'WARNING'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libalignment.html#get_alignment_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libalignment.get_alignment_info" title="Link to this definition"></a></dt>
<dd><p>Analyze a Bio.Align.Alignment object to extract alignment metrics.</p>
<p>This function calculates the match count, mismatch count, gap count, alignment score,
and start and end indices for a given alignment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alignment</strong> (<em>Bio.Align.Alignment</em>) – A <cite>Bio.Align.Alignment</cite> object containing the alignment information.</p></li>
<li><p><strong>reverse</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to return reversed sequence information. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A list containing the following metrics:</p>
<blockquote>
<div><ul class="simple">
<li><p>match_count (int): The number of matches in the alignment.</p></li>
<li><p>mismatch_count (int): The number of mismatches in the alignment.</p></li>
<li><p>gap_count (int): The number of gaps within the sgRNA alignment region. Note: This counts only gaps within the alignment region, not total gaps.</p></li>
<li><p>alignment_score (float): The alignment score.</p></li>
<li><p>start_index (int): The start index of the sgRNA alignment region.</p></li>
<li><p>end_index (int): The end index of the sgRNA alignment region.</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The gap count specifically counts gaps within the sgRNA alignment region, not the total gaps in the entire sequence.
For example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Reference</span><span class="p">:</span>  <span class="n">AGTGGTAAGAAGAAGACGAGACATAATGAG</span>
<span class="gp">&gt;&gt;&gt; </span>            <span class="o">------||||||||||||||.|----||||</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Target</span><span class="p">:</span>     <span class="o">------</span><span class="n">AAGAAGAAGACGAGCC</span><span class="o">----</span><span class="n">TGAG</span>
</pre></div>
</div>
<p>In this example: Gap count = 4 (within sgRNA alignment region), not 10 (total gaps).</p>
<p class="rubric">Example</p>
<p>The returned list will be as follows: return_list = [match_count, mismatch_count, gap_count, alignment_score, start_index, end_index]</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libalignment.get_best_alignment">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libalignment.</span></span><span class="sig-name descname"><span class="pre">get_best_alignment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_a</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_b</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aligner</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">PairwiseAligner</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">consider_strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libalignment.html#get_best_alignment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libalignment.get_best_alignment" title="Link to this definition"></a></dt>
<dd><p>Get best alignment by heightest alignment score</p>
<p>Get best alignment by heightest alignment score,
default to consider strand, that means, it will attempt to use
seq_b.reverse_complement() to align</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq_a</strong> (<em>Seq</em>) – seq_a</p></li>
<li><p><strong>seq_b</strong> (<em>Seq</em>) – seq_b</p></li>
<li><p><strong>aligner</strong> (<em>PairwiseAligner</em>) – aligner returns from instantiate_pairwise_aligner</p></li>
<li><p><strong>consider_strand</strong> (<em>bool</em><em>, </em><em>optional</em>) – consider to use seq_b.reverse_complement() or not, defaults to True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Alignment with a new attribute: alignment.is_a_reverse_complement_alignment (True / False)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Alignment</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libalignment.instantiate_pairwise_aligner">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libalignment.</span></span><span class="sig-name descname"><span class="pre">instantiate_pairwise_aligner</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scoring_match</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">penalty_mismatch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">penalty_gap_open</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">-5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">penalty_gap_extension</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">-2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">penalty_query_left_gap_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">penalty_query_right_gap_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'global'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">letn_match</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_matrix_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DEBUG'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">PairwiseAligner</span></span></span><a class="reference internal" href="_modules/bioat/lib/libalignment.html#instantiate_pairwise_aligner"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libalignment.instantiate_pairwise_aligner" title="Link to this definition"></a></dt>
<dd><p>Returns a PairwiseAligner object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scoring_match</strong> (<em>float</em><em>, </em><em>optional</em>) – Scoring for matches. Defaults to 1.</p></li>
<li><p><strong>penalty_mismatch</strong> (<em>float</em><em>, </em><em>optional</em>) – Penalty for mismatches. Defaults to 0.8.</p></li>
<li><p><strong>penalty_gap_open</strong> (<em>float</em><em>, </em><em>optional</em>) – Penalty for opening a gap. Defaults to -5.</p></li>
<li><p><strong>penalty_gap_extension</strong> (<em>float</em><em>, </em><em>optional</em>) – Penalty for extending a gap. Defaults to -2.</p></li>
<li><p><strong>penalty_query_left_gap_score</strong> (<em>float</em><em>, </em><em>optional</em>) – Penalty for left query gap. Defaults to 0.</p></li>
<li><p><strong>penalty_query_right_gap_score</strong> (<em>float</em><em>, </em><em>optional</em>) – Penalty for right query gap. Defaults to 0.</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) – Alignment mode, either “global” or “local”. Defaults to “global”.</p></li>
<li><p><strong>letn_match</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to treat base ‘N’ as a match. Defaults to False.</p></li>
<li><p><strong>log_level</strong> (<em>str</em><em>, </em><em>optional</em>) – Logging level. Defaults to “DEBUG”.</p></li>
<li><p><strong>score_matrix_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – Custom scoring matrix for bases or amino acids. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An object instantiated from PairwiseAligner.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>PairwiseAligner</p>
</dd>
</dl>
<dl>
<dt>Details for <cite>score_matrix_dict</cite>:</dt><dd><p>You can use any base and any score, such as AGCTN, or 20AA + 2 abnormal AAs,
or any characters you want. For example:</p>
<blockquote>
<div><p>score_matrix_dict = {</p>
<blockquote>
<div><p>(“A”, “A”): scoring_match,
(“A”, “G”): penalty_mismatch,
(“A”, “C”): penalty_mismatch,
(“A”, “T”): penalty_mismatch,
(“A”, “N”): scoring_match,
(“G”, “A”): penalty_mismatch,
(“G”, “G”): scoring_match,
(“G”, “C”): penalty_mismatch,
(“G”, “T”): penalty_mismatch,
(“G”, “N”): scoring_match,
(“C”, “A”): penalty_mismatch,
(“C”, “G”): penalty_mismatch,
(“C”, “C”): scoring_match,
(“C”, “T”): penalty_mismatch,
(“C”, “N”): scoring_match,
(“T”, “A”): penalty_mismatch,
(“T”, “G”): penalty_mismatch,
(“T”, “C”): penalty_mismatch,
(“T”, “T”): scoring_match,
(“T”, “N”): scoring_match,
(“N”, “A”): scoring_match,
(“N”, “G”): scoring_match,
(“N”, “C”): scoring_match,
(“N”, “T”): scoring_match,
(“N”, “N”): scoring_match</p>
</div></blockquote>
<p>}</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-bioat.lib.libcircos">
<span id="bioat-lib-libcircos-module"></span><h2>bioat.lib.libcircos module<a class="headerlink" href="#module-bioat.lib.libcircos" title="Link to this heading"></a></h2>
<p>Doc.
Adapt from <a class="reference external" href="https://github.com/ponnhide/pyCircos">https://github.com/ponnhide/pyCircos</a>
and thanks for this github user.</p>
<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Garc">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libcircos.</span></span><span class="sig-name descname"><span class="pre">Garc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arc_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interspace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(500,</span> <span class="pre">550)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.75</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labelposition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labelsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_visible</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Garc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Garc" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class representation for an arc section in a graphic circle.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>colorlist</strong> (<em>list</em><em> of </em><em>str</em>) – List of color codes for the arc sections.</p></li>
<li><p><strong>_arcnum</strong> (<em>int</em>) – Counter for the number of arcs.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Garc.__setitem__">
<span class="sig-name descname"><span class="pre">__setitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">item</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Garc.__setitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Garc.__setitem__" title="Link to this definition"></a></dt>
<dd><p>Sets the attribute <cite>key</cite> with the value <cite>item</cite>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Garc.__getitem__">
<span class="sig-name descname"><span class="pre">__getitem__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Garc.__getitem__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Garc.__getitem__" title="Link to this definition"></a></dt>
<dd><p>Retrieves the value of the attribute <cite>key</cite>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Garc.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arc_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interspace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(500,</span> <span class="pre">550)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.75</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labelposition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labelsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_visible</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Garc.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Garc.__init__" title="Link to this definition"></a></dt>
<dd><p>Initializes the Garc object with specified parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>arc_id</strong> (<em>str</em><em>, </em><em>optional</em>) – Unique identifier for the Garc class object. If None, a unique ID is generated. Default is None.</p></li>
<li><p><strong>record</strong> (<em>Bio.SeqRecord</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – Bio.SeqRecord object or NCBI accession number for an annotated sequence. Default is None.</p></li>
<li><p><strong>size</strong> (<em>int</em><em>, </em><em>optional</em>) – Width of the arc section, default is 1000. Adjusted if record is provided.</p></li>
<li><p><strong>interspace</strong> (<em>float</em><em>, </em><em>optional</em>) – Distance angle (degrees) to the adjacent arc section. Default is 3.</p></li>
<li><p><strong>raxis_range</strong> (<em>tuple</em><em> of </em><em>int</em><em>, </em><em>optional</em>) – Radial axis range for line plotting. Default is (500, 550).</p></li>
<li><p><strong>facecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Color for filling the arc section. Default is automatically set.</p></li>
<li><p><strong>edgecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Color for the edge of the filled area. Default is “#303030”.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em>, </em><em>optional</em>) – Width of the edge line. Default is 0.75.</p></li>
<li><p><strong>label</strong> (<em>str</em><em>, </em><em>optional</em>) – Label for the arc section. Default is None.</p></li>
<li><p><strong>labelposition</strong> (<em>int</em><em>, </em><em>optional</em>) – Relative label height from the center. Default is 0.</p></li>
<li><p><strong>labelsize</strong> (<em>int</em><em>, </em><em>optional</em>) – Font size of the label. Default is 10.</p></li>
<li><p><strong>label_visible</strong> (<em>bool</em><em>, </em><em>optional</em>) – Determines if the label is visible on the arc section. Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If no match for the provided NCBI accession number is found in <cite>record</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Garc.calc_density">
<span class="sig-name descname"><span class="pre">calc_density</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">positions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Garc.calc_density"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Garc.calc_density" title="Link to this definition"></a></dt>
<dd><p>Calculates density values in a sliding window based on given positions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>positions</strong> (<em>list</em><em> of </em><em>int</em><em> or </em><em>tuple</em>) – A list of x-coordinate values
or a tuple containing two x-coordinate values. Each value
should be in the range of 0 to the size of the Garc object.</p></li>
<li><p><strong>window_size</strong> (<em>int</em><em>, </em><em>optional</em>) – The size of the sliding window.
Defaults to 1000.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If an inappropriate value or values are provided
    for positions.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A list containing the calculated density values.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Garc.calc_nnratio">
<span class="sig-name descname"><span class="pre">calc_nnratio</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'G'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Garc.calc_nnratio"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Garc.calc_nnratio" title="Link to this definition"></a></dt>
<dd><p>Calculates the ratio of nucleotide base frequencies for two specified bases over multiple windows along a sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n1</strong> (<em>str</em>) – The first nucleotide base to be compared.
Must be one of “ATGC”. Default is “G”.</p></li>
<li><p><strong>n2</strong> (<em>str</em>) – The second nucleotide base to be compared.
Must be one of “ATGC”. Default is “C”.</p></li>
<li><p><strong>window_size</strong> (<em>int</em>) – Size of the sliding window.
Default is 1000.</p></li>
<li><p><strong>step_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the sliding step.
If not provided, defaults to window_size.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If no record is provided.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An array containing the computed ratios of n1 to n2 for each window.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Garc.calc_nnskew">
<span class="sig-name descname"><span class="pre">calc_nnskew</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'G'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'C'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Garc.calc_nnskew"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Garc.calc_nnskew" title="Link to this definition"></a></dt>
<dd><p>Calculates n1,n2 skew (n1-n2)/(n1+n2) for multiple windows along
the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n1</strong> (<em>str</em><em>, </em><em>optional</em>) – The first of the two nucleotide bases to be compared.
The default is “G”.</p></li>
<li><p><strong>n2</strong> (<em>str</em><em>, </em><em>optional</em>) – The second of the two nucleotide bases to be compared.
The default is “C”.</p></li>
<li><p><strong>window_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the sliding window. The default is 1000.</p></li>
<li><p><strong>step_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the sliding step. The default is window_size.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If no record is provided, will return an error.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An array of the skews computed by this method.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Garc.colorlist">
<span class="sig-name descname"><span class="pre">colorlist</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['#ff8a80',</span> <span class="pre">'#ff80ab',</span> <span class="pre">'#ea80fc',</span> <span class="pre">'#b388ff',</span> <span class="pre">'#8c9eff',</span> <span class="pre">'#82b1ff',</span> <span class="pre">'#84ffff',</span> <span class="pre">'#a7ffeb',</span> <span class="pre">'#b9f6ca',</span> <span class="pre">'#ccff90',</span> <span class="pre">'#f4ff81',</span> <span class="pre">'#ffff8d',</span> <span class="pre">'#ffe57f',</span> <span class="pre">'#ffd180',</span> <span class="pre">'#ff9e80',</span> <span class="pre">'#bcaaa4',</span> <span class="pre">'#eeeeee',</span> <span class="pre">'#b0bec5',</span> <span class="pre">'#ff5252',</span> <span class="pre">'#ff4081',</span> <span class="pre">'#e040fb',</span> <span class="pre">'#7c4dff',</span> <span class="pre">'#536dfe',</span> <span class="pre">'#448aff',</span> <span class="pre">'#18ffff',</span> <span class="pre">'#64ffda',</span> <span class="pre">'#69f0ae',</span> <span class="pre">'#b2ff59',</span> <span class="pre">'#eeff41',</span> <span class="pre">'#ffff00',</span> <span class="pre">'#ffd740',</span> <span class="pre">'#ffab40',</span> <span class="pre">'#ff6e40',</span> <span class="pre">'#a1887f',</span> <span class="pre">'#e0e0e0',</span> <span class="pre">'#90a4ae']</span></em><a class="headerlink" href="#bioat.lib.libcircos.Garc.colorlist" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libcircos.</span></span><span class="sig-name descname"><span class="pre">Gcircle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A Gcircle class object provides a circle whose diameter is 1000 (a.u.) as a
drawing space. Any graph (line plot, scatter plot, barplot, heatmap, and chordplot)
can be placed on the space by specifying the raxis_range (from 0 to 1000) and
the corresponding Garc class object.</p>
<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.__init__" title="Link to this definition"></a></dt>
<dd><p>Initializes the circular map.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fig</strong> (<em>matplotlib.pyplot.Figure</em><em>, </em><em>optional</em>) – A Matplotlib Figure object. If not provided, a new figure will be created.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – A tuple specifying the size of the figure for the circular map, e.g., (width, height).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.add_garc">
<span class="sig-name descname"><span class="pre">add_garc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">garc</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.add_garc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.add_garc" title="Link to this definition"></a></dt>
<dd><p>Add a new Garc class object to the garc_dict.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>garc</strong> (<a class="reference internal" href="#bioat.lib.libcircos.Garc" title="bioat.lib.libcircos.Garc"><em>Garc</em></a>) – The Garc class object to be added.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.barplot">
<span class="sig-name descname"><span class="pre">barplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">garc_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(550,</span> <span class="pre">600)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rlim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.barplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.barplot" title="Link to this definition"></a></dt>
<dd><p>Draws a bar plot within the specified arc of the Garc class object.</p>
<p>This method visualizes numerical data as bars in a sector corresponding
to the specified arc of a Garc class object identified by <cite>garc_id</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>garc_id</strong> (<em>str</em>) – The ID of the Garc class object. It should exist
in the Gcircle object’s garc_dict.</p></li>
<li><p><strong>data</strong> (<em>list</em><em> or </em><em>numpy.ndarray</em>) – The numerical data used to generate
the plot.</p></li>
<li><p><strong>positions</strong> (<em>list</em><em> or </em><em>numpy.ndarray</em><em>, </em><em>optional</em>) – The x-coordinates of
the values in <cite>data</cite> when plotted in rectangular coordinates.
Each value should be within the range of the Garc class object’s
size identified by <cite>garc_id</cite>. If not provided, coordinates are
automatically generated based on the length of <cite>data</cite>.</p></li>
<li><p><strong>width</strong> (<em>float</em><em> or </em><em>list</em><em> of </em><em>float</em><em>, </em><em>optional</em>) – The width(s) of the bars.
The default width is determined by <cite>garc_object.size / len(data)</cite>.</p></li>
<li><p><strong>raxis_range</strong> (<em>tuple</em><em> of </em><em>int</em><em>, </em><em>optional</em>) – Range for the radial axis
where the line plot is drawn, defaulting to (550, 600).</p></li>
<li><p><strong>rlim</strong> (<em>tuple</em><em> of </em><em>int</em><em>, </em><em>optional</em>) – Specifies the top and bottom limits
for the radial data coordinates. If not provided, it defaults
to the minimum and maximum values of <cite>data</cite>.</p></li>
<li><p><strong>base_value</strong> (<em>float</em><em>, </em><em>optional</em>) – Height of the baseline in data
coordinates; the area between this line and the data line is filled with
<cite>facecolor</cite>. Defaults to None.</p></li>
<li><p><strong>facecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em> or </em><em>list</em><em>, </em><em>optional</em>) – Specifies the face color(s)
of the bars. If a list is provided, its length must match that of <cite>data</cite>.
Default is None.</p></li>
<li><p><strong>edgecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Color of the edges of the bars,
defaulting to “#303030”.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em>, </em><em>optional</em>) – Width of the edge lines of the bars;
defaults to 0.0 (no lines).</p></li>
<li><p><strong>spine</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, displays the spines of the Garc
object on the arc section. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>This method does not return any value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.chord_plot">
<span class="sig-name descname"><span class="pre">chord_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.chord_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.chord_plot" title="Link to this definition"></a></dt>
<dd><p>Visualize interrelationships between data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start_list</strong> (<em>tuple</em>) – Start data location of linked data. The tuple is composed of four parameters:
- <cite>arc_id</cite> (str): The ID of the first Garc class object to be compared. The ID should be in <cite>Gcircle</cite> object’s <cite>garc_dict</cite>.
- <cite>edge_position1</cite> (int): The minimal x coordinate on the Garc class object when the plot is drawn on the rectangular coordinates.
- <cite>edge_position2</cite> (int): The maximal x coordinate on the Garc class object when the plot is drawn on the rectangular coordinates.
- <cite>raxis_position</cite> (int): The base height for the drawing chord.</p></li>
<li><p><strong>end_list</strong> (<em>tuple</em>) – End data location of linked data. The tuple is composed of four parameters:
- <cite>arc_id</cite> (str): The ID of the second Garc class object to be compared. the ID should be in <cite>Gcircle</cite> object’s <cite>garc_dict</cite>.
- <cite>edge_position1</cite> (int): The minimal x coordinate on the Garc class object when the plot is drawn on the rectangular coordinates.
- <cite>edge_position2</cite> (int): The maximal x coordinate on the Garc class object when the plot is drawn on the rectangular coordinates.
- <cite>raxis_position</cite> (int): The base height for the drawing chord.</p></li>
<li><p><strong>facecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Facecolor of the link. The default is None.</p></li>
<li><p><strong>edgecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Edge color of the link. The default is “#303030”.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em>, </em><em>optional</em>) – Edge line width of the link. The default is 0.0.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.cmaps">
<span class="sig-name descname"><span class="pre">cmaps</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">[&lt;matplotlib.colors.LinearSegmentedColormap</span> <span class="pre">object&gt;,</span> <span class="pre">&lt;matplotlib.colors.LinearSegmentedColormap</span> <span class="pre">object&gt;,</span> <span class="pre">&lt;matplotlib.colors.LinearSegmentedColormap</span> <span class="pre">object&gt;,</span> <span class="pre">&lt;matplotlib.colors.LinearSegmentedColormap</span> <span class="pre">object&gt;]</span></em><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.cmaps" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.colors">
<span class="sig-name descname"><span class="pre">colors</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['#f44336',</span> <span class="pre">'#e91e63',</span> <span class="pre">'#9c27b0',</span> <span class="pre">'#673ab7',</span> <span class="pre">'#3f51b5',</span> <span class="pre">'#2196f3',</span> <span class="pre">'#00bcd4',</span> <span class="pre">'#009688',</span> <span class="pre">'#4caf50',</span> <span class="pre">'#8bc34a',</span> <span class="pre">'#cddc39',</span> <span class="pre">'#ffeb3b',</span> <span class="pre">'#ffc107',</span> <span class="pre">'#ff9800',</span> <span class="pre">'#ff5722',</span> <span class="pre">'#795548',</span> <span class="pre">'#9e9e9e',</span> <span class="pre">'#607d8b']</span></em><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.colors" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.featureplot">
<span class="sig-name descname"><span class="pre">featureplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">garc_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feature_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(550,</span> <span class="pre">600)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.featureplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.featureplot" title="Link to this definition"></a></dt>
<dd><p>Visualize sequence features with bar plots in the sector corresponding
to the arc of the Garc class object specified by garc_id.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>garc_id</strong> (<em>str</em>) – ID of the Garc class object. The ID should be in Gcircle object.garc_dict.</p></li>
<li><p><strong>feature_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Biological nature of the Bio.Seqfeature class objects.
Accepts any value, but GenBank format requires registering a biological
nature category for each sequence feature. If the value is “all”,
all features in source will be drawn in the sector of the Garc
class object specified by garc_id. The default is ‘all’.</p></li>
<li><p><strong>source</strong> (<em>list</em><em> of </em><em>Bio.SeqFeature</em><em>, </em><em>optional</em>) – List of Bio.SeqFeature class objects. If not provided,
record.features of the Garc class object specified by garc_id is used.
The default is record.features of the Garc class object specified by garc_id.</p></li>
<li><p><strong>raxis_range</strong> (<em>tuple</em><em> of </em><em>int</em><em>, </em><em>optional</em>) – Radial axis range where the feature plot is drawn.
The default is (550, 600).</p></li>
<li><p><strong>facecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Facecolor(s) of the bars. If a list is provided, its length
should match the data length. The default is None.</p></li>
<li><p><strong>edgecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Edge color of the bars. The default is “#303030”.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em>, </em><em>optional</em>) – Edge line width of the bars. The default is 0.0.</p></li>
<li><p><strong>spine</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, spines of the Garc object are shown on the arc section.
The default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.fillplot">
<span class="sig-name descname"><span class="pre">fillplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">garc_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(550,</span> <span class="pre">600)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rlim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.fillplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.fillplot" title="Link to this definition"></a></dt>
<dd><p>Fill a specified area in the sector corresponding to the arc of the
Garc class object specified by garc_id.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>garc_id</strong> (<em>str</em>) – ID of the Garc class object. The ID should be in Gcircle object.garc_dict.</p></li>
<li><p><strong>data</strong> (<em>list</em><em> or </em><em>numpy.ndarray</em>) – Numerical data used for plot generation.</p></li>
<li><p><strong>positions</strong> (<em>list</em><em> or </em><em>numpy.ndarray</em><em>, </em><em>optional</em>) – The x coordinates of the values in data on the Garc class object
when the plot is drawn on the rectangular coordinates. Each
coordinate value should be in the range 0 to the size of the Garc class
object specified by garc_id. If positions are not given, proper
coordinate values are generated according to the length of data.
Default is None.</p></li>
<li><p><strong>raxis_range</strong> (<em>tuple</em><em> of </em><em>int</em><em>, </em><em>optional</em>) – Radial axis range where line plot is drawn. Default is (550, 600).</p></li>
<li><p><strong>rlim</strong> (<em>tuple</em><em> of </em><em>int</em><em>, </em><em>optional</em>) – The top and bottom r limits in data coordinates. If not given,
the maximum and minimum values in data will be set to top and bottom, respectively.
Default is (min(data), max(data)).</p></li>
<li><p><strong>base_value</strong> (<em>float</em><em>, </em><em>optional</em>) – Base line height in data coordinates. The area between the base
line and the data line is filled by facecolor. Default is None.</p></li>
<li><p><strong>facecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Color for filling. Default is None.</p></li>
<li><p><strong>edgecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Edge color of the filled area. Default is “#303030”.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em>, </em><em>optional</em>) – Edge line width. The default is 0.0.</p></li>
<li><p><strong>spine</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, spines of the Garc object are shown on the arc section.
Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>This function does not return any value.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.heatmap">
<span class="sig-name descname"><span class="pre">heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">garc_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(550,</span> <span class="pre">600)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.heatmap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.heatmap" title="Link to this definition"></a></dt>
<dd><p>Visualize magnitudes of data values by color scale in the sector
corresponding to the arc of the Garc class object specified by garc_id.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>garc_id</strong> (<em>str</em>) – ID of the Garc class object. The ID should be in Gcircle object.garc_dict.</p></li>
<li><p><strong>data</strong> (<em>list</em><em> or </em><em>numpy.ndarray</em>) – Numerical data to be used for plot generation.</p></li>
<li><p><strong>positions</strong> (<em>list</em><em> or </em><em>numpy.ndarray</em><em>, </em><em>optional</em>) – The x coordinates of the values in data on the Garc class object
when the plot is drawn on the rectangular coordinates. Each
coordinate value should be in the range 0 to size of the Garc class
object specified by garc_id. If positions are not given, proper
coordinates values are generated according to the length of data.
Defaults to None.</p></li>
<li><p><strong>width</strong> (<em>float</em><em> or </em><em>list</em><em> of </em><em>float</em><em>, </em><em>optional</em>) – Width(s) of the bars. Defaults to <cite>garc_object.size / len(data)</cite>.</p></li>
<li><p><strong>raxis_range</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Radial axis range where heatmap is drawn. Default is (550, 600).</p></li>
<li><p><strong>cmap</strong> (<em>str</em><em> or </em><em>matplotlib.colors.Colormap</em><em>, </em><em>optional</em>) – The mapping from data values to color space. Default is ‘Reds’.</p></li>
<li><p><strong>vmin</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimum data threshold for color scale. Defaults to min(data).</p></li>
<li><p><strong>vmax</strong> (<em>float</em><em>, </em><em>optional</em>) – Maximum data threshold for color scale. Defaults to max(data).</p></li>
<li><p><strong>edgecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Edge color of the bars. Default is “#303030”.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em>, </em><em>optional</em>) – Edge line width of the bars. Default is 0.0.</p></li>
<li><p><strong>spine</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, spines of the Garc object are shown on the arc section.
Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.lineplot">
<span class="sig-name descname"><span class="pre">lineplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">garc_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(550,</span> <span class="pre">600)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rlim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linestyle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'solid'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.lineplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.lineplot" title="Link to this definition"></a></dt>
<dd><p>Plot a line in the sector corresponding to the arc of the Garc class
object specified by garc_id.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>garc_id</strong> (<em>str</em>) – ID of the Garc class object. The ID should be in Gcircle
object.garc_dict.</p></li>
<li><p><strong>data</strong> (<em>list</em><em> or </em><em>numpy.ndarray</em>) – Numerical data to be used for plot generation.</p></li>
<li><p><strong>positions</strong> (<em>list</em><em> or </em><em>numpy.ndarray</em><em>, </em><em>optional</em>) – The x coordinates of the values in data on the Garc class object
when the plot is drawn on rectangular coordinates. Each coordinate
value should be in the range 0 to size of the Garc class object
specified by garc_id. If positions are not given, proper coordinate
values are generated according to the length of data. The default
is None.</p></li>
<li><p><strong>raxis_range</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Radial axis range where line plot is drawn. The default is
(550, 600).</p></li>
<li><p><strong>rlim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – The top and bottom r limits in data coordinates. If rlim value
is not given, the maximum value and the minimum value in data
will be set to top and bottom, respectively. The default is None.</p></li>
<li><p><strong>linestyle</strong> (<em>str</em><em>, </em><em>optional</em>) – Line style. The default is “solid”. Possible line styles are documented at
<a class="reference external" href="https://matplotlib.org/stable/gallery/lines_bars_and_markers/linestyles.html">https://matplotlib.org/stable/gallery/lines_bars_and_markers/linestyles.html</a></p></li>
<li><p><strong>linecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Color of the line plot. If linecolor value is not given, the color
will be set according to the default color set of matplotlib. To
specify the opacity for a line color, please use <cite>(r,g,b,a)</cite> or
<cite>#FFFFFF</cite> format. The default is None.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em>, </em><em>optional</em>) – Edge line width. The default is 1.0.</p></li>
<li><p><strong>spine</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, spines of the Garc object are shown on the arc section.
The default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'test'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.save" title="Link to this definition"></a></dt>
<dd><p>Saves the figure object of the Gcircle class to a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the file to save the figure. Defaults to “test”.</p></li>
<li><p><strong>format</strong> (<em>str</em><em>, </em><em>optional</em>) – The file format for the saved figure. Defaults to “pdf”.</p></li>
<li><p><strong>dpi</strong> (<em>int</em><em>, </em><em>optional</em>) – The resolution of the saved figure in dots per inch. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.scatterplot">
<span class="sig-name descname"><span class="pre">scatterplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">garc_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(550,</span> <span class="pre">600)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rlim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markershape</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'o'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markersize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.scatterplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.scatterplot" title="Link to this definition"></a></dt>
<dd><p>Draws a scatter plot on the sector corresponding to the arc of the Garc class object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>garc_id</strong> (<em>str</em>) – ID of the Garc class object. The ID should be in Gcircle object.garc_dict.</p></li>
<li><p><strong>data</strong> (<em>list</em><em> or </em><em>numpy.ndarray</em>) – Numerical data used for plot generation.</p></li>
<li><p><strong>positions</strong> (<em>list</em><em> or </em><em>numpy.ndarray</em><em>, </em><em>optional</em>) – The x coordinates of the values in data on the Garc class object
when the plot is drawn on rectangular coordinates. Each coordinate
value should be in the range 0 to the size of the Garc class
object specified by garc_id. If positions are not provided,
proper coordinate values are generated according to the length of data. Default is None.</p></li>
<li><p><strong>raxis_range</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Radial axis range where the line plot is drawn. Default is (550, 600).</p></li>
<li><p><strong>rlim</strong> (<em>tuple</em><em>, </em><em>optional</em>) – The top and bottom r limits in data coordinates. Default is
(min(data), max(data)) if not provided.</p></li>
<li><p><strong>markershape</strong> (<em>str</em><em>, </em><em>optional</em>) – Marker shape. Default is “o”.
Possible markers are listed at
<a class="reference external" href="https://matplotlib.org/stable/gallery/lines_bars_and_markers/marker_reference.html">https://matplotlib.org/stable/gallery/lines_bars_and_markers/marker_reference.html</a>.</p></li>
<li><p><strong>markersize</strong> (<em>float</em><em> or </em><em>list</em><em> of </em><em>float</em><em>, </em><em>optional</em>) – Size(s) of the marker(s). Default is 5.</p></li>
<li><p><strong>facecolor</strong> (<em>str</em><em>, </em><em>tuple</em><em> or </em><em>list</em><em>, </em><em>optional</em>) – Face color(s) of the markers. If a list is provided, its length
should match the length of data. Default is None.</p></li>
<li><p><strong>edgecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Edge color of the markers. Default is “#303030”.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em>, </em><em>optional</em>) – Edge line width of the markers. Default is 0.0.</p></li>
<li><p><strong>spine</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, spines of the Garc object are shown in the arc section.
Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.set_garcs">
<span class="sig-name descname"><span class="pre">set_garcs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">360</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.set_garcs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.set_garcs" title="Link to this definition"></a></dt>
<dd><p>Visualize arc rectangles for Garc class objects in .garc_dict.</p>
<p>This method draws the arc rectangles of the Garc class objects in the drawing space.
After executing this method, no new Garc class objects can be added to garc_dict,
and a figure parameter representing a matplotlib.pyplot.figure object will be created
in the Gcircle object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – Start angle of the circos plot. The range is -360 to 360.
Defaults to 0.</p></li>
<li><p><strong>end</strong> (<em>int</em><em>, </em><em>optional</em>) – End angle of the circos plot. The range is -360 to 360.
Defaults to 360.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.setspine">
<span class="sig-name descname"><span class="pre">setspine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">garc_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(550,</span> <span class="pre">600)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#30303000'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.75</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.setspine"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.setspine" title="Link to this definition"></a></dt>
<dd><p>Sets the spines in the sector corresponding to the arc of
the Garc class object specified by garc_id.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>garc_id</strong> (<em>str</em>) – ID of the Garc class object. The ID must be present in Gcircle object.garc_dict.</p></li>
<li><p><strong>raxis_range</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Radial axis range where the line plot is drawn.
Defaults to (550, 600).</p></li>
<li><p><strong>facecolor</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for the spines area. Default is “#30303000”.</p></li>
<li><p><strong>edgecolor</strong> (<em>str</em><em>, </em><em>optional</em>) – Edge color of the spines boundary area. Default is “#303030”.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em>, </em><em>optional</em>) – Edge line width of the spines boundary area. Default is 0.75.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Gcircle.tickplot">
<span class="sig-name descname"><span class="pre">tickplot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">garc_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tickinterval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tickpositions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticklabels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tickwidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tickcolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticklabelsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticklabelcolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticklabelmargin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tickdirection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'outer'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ticklabelorientation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'vertical'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Gcircle.tickplot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Gcircle.tickplot" title="Link to this definition"></a></dt>
<dd><p>Plots ticks on the arc of the Garc class object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>garc_id</strong> (<em>str</em>) – The ID of the Garc class object. The ID should be in
Gcircle object.garc_dict.</p></li>
<li><p><strong>raxis_range</strong> (<em>tuple</em><em> of </em><em>int</em><em>, </em><em>optional</em>) – Radial axis range where tick plot is drawn.
If direction is “inner”, the default is
<cite>(r0 - 0.5 * abs(r1 - r0), r0)</cite>. If direction is
“outer”, the default is <cite>(r1, r1 + 0.5 * abs(r1 - r0))</cite>.
<cite>r0</cite> and <cite>r1</cite> are defined as
<cite>Garc_object.raxis_range[0], Garc_object.raxis_range[1]</cite>.</p></li>
<li><p><strong>tickinterval</strong> (<em>int</em><em>, </em><em>optional</em>) – The interval between ticks. The default value is 1000.
If <cite>tickpositions</cite> is provided, this value will be ignored.</p></li>
<li><p><strong>tickpositions</strong> (<em>list</em><em> of </em><em>int</em><em>, </em><em>optional</em>) – Specific positions on the arc of the Garc class object
where ticks should be placed. The values should be
less than <cite>Garc_object.size</cite>.</p></li>
<li><p><strong>ticklabels</strong> (<em>list</em><em> of </em><em>int</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – Labels for the ticks on the arc of the Garc class object.
The default value is the same as <cite>tickpositions</cite>.</p></li>
<li><p><strong>tickwidth</strong> (<em>float</em><em>, </em><em>optional</em>) – The width of the ticks. The default value is 1.0.</p></li>
<li><p><strong>tickcolor</strong> (<em>str</em><em> or </em><em>float</em><em>, </em><em>optional</em>) – The color of the ticks. The default value is “#303030”.</p></li>
<li><p><strong>ticklabelsize</strong> (<em>float</em><em>, </em><em>optional</em>) – The font size of the tick labels. The default value is 10.</p></li>
<li><p><strong>ticklabelcolor</strong> (<em>str</em><em>, </em><em>optional</em>) – The color of the tick labels. The default value is “#303030”.</p></li>
<li><p><strong>ticklabelmargin</strong> (<em>float</em><em>, </em><em>optional</em>) – The margin of the tick labels. The default value is 10.</p></li>
<li><p><strong>tickdirection</strong> (<em>str</em><em>, </em><em>optional</em>) – The direction of the ticks (“outer” or “inner”).
The default value is “outer”.</p></li>
<li><p><strong>ticklabelorientation</strong> (<em>str</em><em>, </em><em>optional</em>) – The orientation of the tick labels (“vertical” or “horizontal”).
The default value is “vertical”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Tarc">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libcircos.</span></span><span class="sig-name descname"><span class="pre">Tarc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arc_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'newick'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interspace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(900,</span> <span class="pre">950)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labelposition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labelsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_visible</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Tarc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Tarc" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#bioat.lib.libcircos.Garc" title="bioat.lib.libcircos.Garc"><code class="xref py py-class docutils literal notranslate"><span class="pre">Garc</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Tarc.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">arc_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tree</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'newick'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interspace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raxis_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(900,</span> <span class="pre">950)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">facecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labelposition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labelsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_visible</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Tarc.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Tarc.__init__" title="Link to this definition"></a></dt>
<dd><p>Initializes a Garc class object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>arc_id</strong> (<em>str</em><em>, </em><em>optional</em>) – Unique identifier for the Garc class object. If not provided,
an original unique ID is automatically generated. Default is None.</p></li>
<li><p><strong>tree</strong> (<em>str</em>) – File name of phylogenetic tree.</p></li>
<li><p><strong>format</strong> (<em>str</em>) – Format of the phylogenetic tree. Default is “newick”.</p></li>
<li><p><strong>interspace</strong> (<em>float</em><em>, </em><em>optional</em>) – Distance angle (degrees) to the adjacent arc section in clockwise
sequence. The actual interspace size is determined by the ratio
of size to the combined sum of the size and interspace values of
the Garc class objects in the Gcircle class object. Default is 3.</p></li>
<li><p><strong>raxis_range</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Radial axis range where the line plot is drawn. Default is (900, 950).</p></li>
<li><p><strong>facecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Color for filling. Default is None.</p></li>
<li><p><strong>edgecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Edge color of the filled area. Default is “#303030”.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em>, </em><em>optional</em>) – Edge line width. Default is 0.</p></li>
<li><p><strong>label</strong> (<em>str</em><em>, </em><em>optional</em>) – Label of the arc section. Default is None.</p></li>
<li><p><strong>labelposition</strong> (<em>int</em><em>, </em><em>optional</em>) – Relative label height from the center of the arc section. Default is 0.</p></li>
<li><p><strong>labelsize</strong> (<em>int</em><em>, </em><em>optional</em>) – Font size of the label. Default is 10.</p></li>
<li><p><strong>label_visible</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, the label of the Garc object is shown on the arc section.
Default is False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Tcircle">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libcircos.</span></span><span class="sig-name descname"><span class="pre">Tcircle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Tcircle"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Tcircle" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#bioat.lib.libcircos.Gcircle" title="bioat.lib.libcircos.Gcircle"><code class="xref py py-class docutils literal notranslate"><span class="pre">Gcircle</span></code></a></p>
<p>Tcircle class is the subclass of Gcircle. All methods implemented in the
Gcircle class also can be used. Then, the two additional methods set_tarc,
plot_tree and plot_highlight is provided in the Tcircle class.</p>
<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Tcircle.__getattr__">
<span class="sig-name descname"><span class="pre">__getattr__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Tcircle.__getattr__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Tcircle.__getattr__" title="Link to this definition"></a></dt>
<dd><p>Retrieves attributes of the object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>name</strong> (<em>str</em>) – The name of the attribute to retrieve.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The corresponding _garc_dict if name is ‘tarc_dict’.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Tcircle.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Tcircle.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Tcircle.__init__" title="Link to this definition"></a></dt>
<dd><p>Initializes the circular mapping object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fig</strong> (<em>matplotlib.pyplot.figure</em><em>, </em><em>optional</em>) – Matplotlib Figure class object to be used.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Size of the figure for the circular map.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Tcircle.add_tarc">
<span class="sig-name descname"><span class="pre">add_tarc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tarc</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Tcircle.add_tarc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Tcircle.add_tarc" title="Link to this definition"></a></dt>
<dd><p>Adds a new Tarc or Garc object to the tarc_dict.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tarc</strong> (<a class="reference internal" href="#bioat.lib.libcircos.Tarc" title="bioat.lib.libcircos.Tarc"><em>Tarc</em></a><em> or </em><a class="reference internal" href="#bioat.lib.libcircos.Garc" title="bioat.lib.libcircos.Garc"><em>Garc</em></a>) – The Tarc or Garc object to be added.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Tcircle.plot_highlight">
<span class="sig-name descname"><span class="pre">plot_highlight</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tarc_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">highlight_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Tcircle.plot_highlight"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Tcircle.plot_highlight" title="Link to this definition"></a></dt>
<dd><p>Add highlight for a specific clade under the given internal clade.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tarc_id</strong> (<em>str</em>) – ID of the Tarc class object. The ID should be in Tcircle object’s
tarc_dict.</p></li>
<li><p><strong>highlight_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – <p>A dictionary composed of pairs of internal clade name and a sub-dict.
Instead of clade name, tuples of terminal clade names can also be used.
A sub-dict is composed of the following key-value pairs:</p>
<ul>
<li><dl class="simple">
<dt>color (str):</dt><dd><p>Color of the highlight for clades. The default is “#000000”.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>alpha (float):</dt><dd><p>Alpha of the highlight for clades. The default is 0.25.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>label (str, optional):</dt><dd><p>Label for the highlight. The default is None.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>fontsize (float):</dt><dd><p>Font size of the label. The default is 10.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>y (float):</dt><dd><p>Y location of the text. The default is at the bottom edge of the highlight.</p>
</dd>
</dl>
</li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Tcircle.plot_tree">
<span class="sig-name descname"><span class="pre">plot_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tarc_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rlim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">700)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cladevisual_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">highlight_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#303030'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Tcircle.plot_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Tcircle.plot_tree" title="Link to this definition"></a></dt>
<dd><p>Draw a circular phylogenetic tree.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tarc_id</strong> (<em>str</em>) – The ID of the Tarc class object. The ID should be in
Tcircle object.tarc_dict.</p></li>
<li><p><strong>rlim</strong> (<em>tuple</em><em> of </em><em>int</em><em>, </em><em>optional</em>) – The top and bottom radial limits in data coordinates.
Defaults to (0, 700).</p></li>
<li><p><strong>cladevisual_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – <p>A dictionary containing clade visualization parameters,
structured as pairs of clade name and a sub-dictionary with
the following keys:</p>
<ul>
<li><p><cite>size</cite> (float): Size of the dot. Defaults to 5.</p></li>
<li><p><cite>color</cite> (str or float): Face color of the dot.
Defaults to “#303030”.</p></li>
<li><p><cite>edgecolor</cite> (str or float): Edge line color of the dot.
Defaults to “#303030”.</p></li>
<li><p><cite>linewidth</cite> (float): Edge line width of the dot.
Defaults to 0.5.</p></li>
</ul>
</p></li>
<li><p><strong>highlight_dict</strong> (<em>dict</em><em>, </em><em>optional</em>) – <p>A dictionary containing clade highlight parameters,
which can also use tuples of terminal clade names.
The structure includes:</p>
<ul>
<li><p><cite>color</cite> (str): Color for highlighting clades.
Defaults to “#000000”.</p></li>
<li><p><cite>alpha</cite> (float): Transparency level for highlights.
Defaults to 0.25.</p></li>
<li><p><cite>label</cite> (str): Label for the highlight. Defaults to None.</p></li>
<li><p><cite>fontsize</cite> (float): Font size of the label.
Defaults to 10.</p></li>
<li><p><cite>y</cite> (float): Y location of the text.
Defaults to the bottom edge of the highlight.</p></li>
</ul>
</p></li>
<li><p><strong>linecolor</strong> (<em>str</em><em> or </em><em>tuple</em><em>, </em><em>optional</em>) – Color of the tree line. Defaults to “#303030”.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em>) – Line width of the tree. Defaults to 0.5.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libcircos.Tcircle.set_tarcs">
<span class="sig-name descname"><span class="pre">set_tarcs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">360</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#Tcircle.set_tarcs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.Tcircle.set_tarcs" title="Link to this definition"></a></dt>
<dd><p>Visualizes the arc rectangles of Tarc objects in the .garc_dict.</p>
<p>After execution, no new Tarc objects can be added to garc_dict,
and a matplotlib.pyplot.figure will be created in the Tcircle object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>start</strong> (<em>int</em><em>, </em><em>optional</em>) – Start angle of the circos plot, in the range of -360 to 360. Defaults to 0.</p></li>
<li><p><strong>end</strong> (<em>int</em><em>, </em><em>optional</em>) – End angle of the circos plot, in the range of -360 to 360. Defaults to 360.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcircos.table_hg38_chromosome_length">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcircos.</span></span><span class="sig-name descname"><span class="pre">table_hg38_chromosome_length</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#table_hg38_chromosome_length"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.table_hg38_chromosome_length" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcircos.table_hg38_cytoband">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcircos.</span></span><span class="sig-name descname"><span class="pre">table_hg38_cytoband</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#table_hg38_cytoband"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.table_hg38_cytoband" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcircos.table_mm10_chromosome_length">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcircos.</span></span><span class="sig-name descname"><span class="pre">table_mm10_chromosome_length</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#table_mm10_chromosome_length"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.table_mm10_chromosome_length" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcircos.table_mm10_cytoband">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcircos.</span></span><span class="sig-name descname"><span class="pre">table_mm10_cytoband</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcircos.html#table_mm10_cytoband"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcircos.table_mm10_cytoband" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-bioat.lib.libcolor">
<span id="bioat-lib-libcolor-module"></span><h2>bioat.lib.libcolor module<a class="headerlink" href="#module-bioat.lib.libcolor" title="Link to this heading"></a></h2>
<p>_summary_</p>
<p>author: Herman Huanan Zhao
email: <a class="reference external" href="mailto:hermanzhaozzzz&#37;&#52;&#48;gmail&#46;com">hermanzhaozzzz<span>&#64;</span>gmail<span>&#46;</span>com</a>
homepage: <a class="reference external" href="https://github.com/hermanzhaozzzz">https://github.com/hermanzhaozzzz</a></p>
<p>_description_</p>
<dl>
<dt>example 1:</dt><dd><dl>
<dt>bioat list</dt><dd><dl>
<dt>&lt;in shell&gt;:</dt><dd><p>$ bioat list</p>
</dd>
<dt>&lt;in python consolo&gt;:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">bioat.cli</span><span class="w"> </span><span class="kn">import</span> <span class="n">Cli</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bioat</span> <span class="o">=</span> <span class="n">Cli</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bioat</span><span class="o">.</span><span class="n">list</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bioat</span><span class="o">.</span><span class="n">list</span><span class="p">())</span>
</pre></div>
</div>
</dd>
</dl>
</dd>
</dl>
</dd>
<dt>example 2:</dt><dd><p>_example_</p>
</dd>
</dl>
<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcolor.convert_hex_to_rgb">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcolor.</span></span><span class="sig-name descname"><span class="pre">convert_hex_to_rgb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hex_color</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span></span></span><a class="reference internal" href="_modules/bioat/lib/libcolor.html#convert_hex_to_rgb"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcolor.convert_hex_to_rgb" title="Link to this definition"></a></dt>
<dd><p>Convert HEX color to RGB color.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>hex_color</strong> – str, like ‘#FFFFAA’</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple, like (255, 255, 170)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcolor.convert_rgb_to_hex">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcolor.</span></span><span class="sig-name descname"><span class="pre">convert_rgb_to_hex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rgb_color</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/bioat/lib/libcolor.html#convert_rgb_to_hex"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcolor.convert_rgb_to_hex" title="Link to this definition"></a></dt>
<dd><p>Convert RGB color to HEX color.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rgb_color</strong> – tuple, like (255, 255, 170)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>str, like ‘#FFFFAA’</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcolor.make_color_list">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcolor.</span></span><span class="sig-name descname"><span class="pre">make_color_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">low_color_RGB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">high_color_RGB</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length_out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_fmt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'HEX'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DEBUG'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcolor.html#make_color_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcolor.make_color_list" title="Link to this definition"></a></dt>
<dd><dl class="simple">
<dt>INPUT</dt><dd><dl class="simple">
<dt>&lt;low_color_RGB&gt; &lt;high_color_RGB&gt;</dt><dd><p>Format like (210, 179, 150), tuple, list, or np.array</p>
</dd>
<dt>&lt;back_format&gt;</dt><dd><p>Hex OR RGB</p>
</dd>
</dl>
</dd>
<dt>RETURN</dt><dd><p>&lt;color_list&gt;</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcolor.map_color">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcolor.</span></span><span class="sig-name descname"><span class="pre">map_color</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value_vec</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">breaks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcolor.html#map_color"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcolor.map_color" title="Link to this definition"></a></dt>
<dd><dl class="simple">
<dt>INPUT:</dt><dd><dl class="simple">
<dt>&lt;value_vec&gt;</dt><dd><p>np.array or a list of values</p>
</dd>
<dt>&lt;breaks&gt;</dt><dd><p>A sorted value list, which can split all num into len(color_list) intervals.
e.g. [0.01, 0.1, 0.5, 1] make all real num into 5 intervals, (-Inf,0.01], (0.01,0.1], (0.1, 0.5],  (0.5, 1], (1, +Inf]</p>
</dd>
<dt>&lt;color_list&gt;</dt><dd><p>A hex-format color list, which have to match with breaks</p>
</dd>
</dl>
</dd>
<dt>RETURN</dt><dd><dl class="simple">
<dt>&lt;value_color_vec&gt;</dt><dd><p>A list map the value_vec with breaks</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcolor.map_colors_between_two">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcolor.</span></span><span class="sig-name descname"><span class="pre">map_colors_between_two</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">base_color</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_color</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcolor.html#map_colors_between_two"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcolor.map_colors_between_two" title="Link to this definition"></a></dt>
<dd><p>在两个指定颜色之间根据数值映射生成颜色。</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>base_color</strong> (<em>str</em>) – 起始颜色，格式为 ‘#RRGGBB’。</p></li>
<li><p><strong>target_color</strong> (<em>str</em>) – 结束颜色，格式为 ‘#RRGGBB’。</p></li>
<li><p><strong>values</strong> (<em>list</em><em> or </em><em>array</em>) – 数值数组，范围可为任意值。</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>对应颜色数组（#RRGGBB 格式）。</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcolor.plot_colortable">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcolor.</span></span><span class="sig-name descname"><span class="pre">plot_colortable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">colors</span></span></em>, <em class="sig-param"><span class="keyword-only-separator o"><abbr title="Keyword-only parameters separator (PEP 3102)"><span class="pre">*</span></abbr></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort_colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcolor.html#plot_colortable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcolor.plot_colortable" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-bioat.lib.libcrispr">
<span id="bioat-lib-libcrispr-module"></span><h2>bioat.lib.libcrispr module<a class="headerlink" href="#module-bioat.lib.libcrispr" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcrispr.cmp_align_list">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcrispr.</span></span><span class="sig-name descname"><span class="pre">cmp_align_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">aln_a</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aln_b</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libcrispr.html#cmp_align_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcrispr.cmp_align_list" title="Link to this definition"></a></dt>
<dd><p>Compare function for alignment lists.</p>
<p>This function processes alignment data and performs comparisons based on the input structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>input_data</strong> (<em>list</em><em> | </em><em>dict</em>) – The alignment data to be processed. Can be provided as:
- A list, e.g.: [17, 3, 0, 73.0, 1, 22, ‘AAGAAGAAGACGAGTCTGCA’, ‘||||||||||||||X|||XX’, ‘AAGAAGAAGACGAGCCTGAG’]
- A dictionary, e.g.: {‘match_count’: 26, ‘mismatch_count’: 3, ‘gap_count’: 4, ‘aln_score’: 96.0, ‘ref_aln_start’: 0, ‘ref_aln_end’: 32, ‘alignment’: {‘reference_seq’: ‘GGCACTGCGGCTGGAAAAAAAAAAAAAAA–GT’,’aln_info’: ‘–..|.|||||||||||||||||||||||–||’, ‘target_seq’: ‘–GGCAGCGGCTGGAAAAAAAAAAAAAAAAGGT’}}</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function does not return a value but performs comparisons.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">compare_alignments</span><span class="p">([</span>
<span class="gp">... </span>    <span class="mi">17</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">73.0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">22</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s1">&#39;AAGAAGAAGACGAGTCTGCA&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s1">&#39;||||||||||||||X|||XX&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="s1">&#39;AAGAAGAAGACGAGCCTGAG&#39;</span>
<span class="gp">... </span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libcrispr.run_target_seq_align">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libcrispr.</span></span><span class="sig-name descname"><span class="pre">run_target_seq_align</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aligner</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">PairwiseAligner</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PAM</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'WARNING'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="reference internal" href="_modules/bioat/lib/libcrispr.html#run_target_seq_align"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libcrispr.run_target_seq_align" title="Link to this definition"></a></dt>
<dd><p>Perform global alignment for the target sequence.</p>
<p>This function performs global alignment between a reference sequence and a target sequence (usually the sgRNA sequence without PAM) using the specified aligner.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_seq</strong> (<em>Seq</em>) – A Seq object from BioPython representing the reference sequence
for the targeted deep sequencing.</p></li>
<li><p><strong>target_seq</strong> (<em>Seq</em>) – A Seq object from BioPython representing the target region sequence
for the editing window (usually the sgRNA sequence without PAM).</p></li>
<li><p><strong>aligner</strong> (<em>object</em>) – A PairwiseAligner object from BioPython used for performing the alignment.</p></li>
<li><p><strong>PAM</strong> (<em>dict</em><em>, </em><em>optional</em>) – A dictionary containing PAM information with the following keys:
- ‘PAM’ (str): The PAM sequence (e.g., “AGG”).
- ‘position’ (int): The insertion site of <cite>target_seq</cite>.
- ‘weight’ (float): The priority weight for PAM alignment. The alignment score will be multiplied by this weight.
- Example: {‘PAM’: ‘AGG’, ‘position’: 20, ‘weight’: 1.0}.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of dictionaries, each containing alignment results. Each dictionary has the following keys:
- ‘match_count’ (int): The number of matching bases in the alignment.
- ‘mismatch_count’ (int): The number of mismatched bases.
- ‘gap_count’ (int): The number of gaps.
- ‘aln_score’ (float): The alignment score.
- ‘ref_aln_start’ (int): The starting position of the alignment on the reference sequence.
- ‘ref_aln_end’ (int): The ending position of the alignment on the reference sequence.
- ‘alignment’ (dict): The alignment details containing:’reference_seq’ (str): The aligned reference sequence.’aln_info’ (str): The alignment information (e.g., matches, mismatches, and gaps).’target_seq’ (str): The aligned target sequence.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[dict]</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>An example alignment result might look like this:[{‘match_count’: 26,’mismatch_count’: 3,’gap_count’: 4,’aln_score’: 96.0,’ref_aln_start’: 0,’ref_aln_end’: 32,’alignment’: {‘reference_seq’: ‘GGCACTGCGGCTGGAAAAAAAAAAAAAAA–GT’,’aln_info’: ‘–..|.|||||||||||||||||||||||–||’,’target_seq’: ‘–GGCAGCGGCTGGAAAAAAAAAAAAAAAAGGT’}}]</p>
</dd></dl>

</section>
<section id="module-bioat.lib.libdataclasses">
<span id="bioat-lib-libdataclasses-module"></span><h2>bioat.lib.libdataclasses module<a class="headerlink" href="#module-bioat.lib.libdataclasses" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Assembly">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libdataclasses.</span></span><span class="sig-name descname"><span class="pre">Assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libdataclasses.html#Assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libdataclasses.Assembly" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Assembly.contigs">
<span class="sig-name descname"><span class="pre">contigs</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#bioat.lib.libdataclasses.Fasta" title="bioat.lib.libdataclasses.Fasta"><span class="pre">Fasta</span></a><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Assembly.contigs" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Assembly.length">
<span class="sig-name descname"><span class="pre">length</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Assembly.length" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Assembly.path">
<span class="sig-name descname"><span class="pre">path</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Assembly.path" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Bam">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libdataclasses.</span></span><span class="sig-name descname"><span class="pre">Bam</span></span><a class="reference internal" href="_modules/bioat/lib/libdataclasses.html#Bam"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libdataclasses.Bam" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Bed">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libdataclasses.</span></span><span class="sig-name descname"><span class="pre">Bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chromosome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libdataclasses.html#Bed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libdataclasses.Bed" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Bed.chromosome">
<span class="sig-name descname"><span class="pre">chromosome</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Bed.chromosome" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Bed.end">
<span class="sig-name descname"><span class="pre">end</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Bed.end" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Bed.name">
<span class="sig-name descname"><span class="pre">name</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Bed.name" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Bed.score">
<span class="sig-name descname"><span class="pre">score</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Bed.score" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Bed.start">
<span class="sig-name descname"><span class="pre">start</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Bed.start" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Bed.strand">
<span class="sig-name descname"><span class="pre">strand</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Bed.strand" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Fasta">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libdataclasses.</span></span><span class="sig-name descname"><span class="pre">Fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libdataclasses.html#Fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libdataclasses.Fasta" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="simple">
<dt>usage:</dt><dd><p>fa = Fasta(‘this_header’, ‘AGCTACGTCCCCTGA’)
fa
print(fa)</p>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Fasta.header">
<span class="sig-name descname"><span class="pre">header</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Fasta.header" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Fasta.length">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">length</span></span><a class="headerlink" href="#bioat.lib.libdataclasses.Fasta.length" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Fasta.sequence">
<span class="sig-name descname"><span class="pre">sequence</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Fasta.sequence" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Fastq">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libdataclasses.</span></span><span class="sig-name descname"><span class="pre">Fastq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extras</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quality</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libdataclasses.html#Fastq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libdataclasses.Fastq" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#bioat.lib.libdataclasses.Fasta" title="bioat.lib.libdataclasses.Fasta"><code class="xref py py-class docutils literal notranslate"><span class="pre">Fasta</span></code></a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Fastq.extras">
<span class="sig-name descname"><span class="pre">extras</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Fastq.extras" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Fastq.header">
<span class="sig-name descname"><span class="pre">header</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Fastq.header" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Fastq.is_valid">
<span class="sig-name descname"><span class="pre">is_valid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libdataclasses.html#Fastq.is_valid"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libdataclasses.Fastq.is_valid" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Fastq.quality">
<span class="sig-name descname"><span class="pre">quality</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Fastq.quality" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.Fastq.sequence">
<span class="sig-name descname"><span class="pre">sequence</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#bioat.lib.libdataclasses.Fastq.sequence" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libdataclasses.VCF">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libdataclasses.</span></span><span class="sig-name descname"><span class="pre">VCF</span></span><a class="reference internal" href="_modules/bioat/lib/libdataclasses.html#VCF"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libdataclasses.VCF" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
</dd></dl>

</section>
<section id="module-bioat.lib.libdetect_seq">
<span id="bioat-lib-libdetect-seq-module"></span><h2>bioat.lib.libdetect_seq module<a class="headerlink" href="#module-bioat.lib.libdetect_seq" title="Link to this heading"></a></h2>
</section>
<section id="module-bioat.lib.libfastx">
<span id="bioat-lib-libfastx-module"></span><h2>bioat.lib.libfastx module<a class="headerlink" href="#module-bioat.lib.libfastx" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libfastx.calculate_length_distribution">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libfastx.</span></span><span class="sig-name descname"><span class="pre">calculate_length_distribution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plt_show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'WARNING'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libfastx.html#calculate_length_distribution"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libfastx.calculate_length_distribution" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libfastx.cas13_finder">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libfastx.</span></span><span class="sig-name descname"><span class="pre">cas13_finder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_faa</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_faa</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lmin</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lmax</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DEBUG'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/bioat/lib/libfastx.html#cas13_finder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libfastx.cas13_finder" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libfastx.cas_finder">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libfastx.</span></span><span class="sig-name descname"><span class="pre">cas_finder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_fa</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_faa</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_contig_fa</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_crispr_info_tab</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lmin</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lmax</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temp_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prodigal</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prodigal_mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'meta'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pilercr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rm_temp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'INFO'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/bioat/lib/libfastx.html#cas_finder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libfastx.cas_finder" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libfastx.format_this_fastx">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libfastx.</span></span><span class="sig-name descname"><span class="pre">format_this_fastx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">old_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'DEBUG'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libfastx.html#format_this_fastx"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libfastx.format_this_fastx" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-bioat.lib.libjgi">
<span id="bioat-lib-libjgi-module"></span><h2>bioat.lib.libjgi module<a class="headerlink" href="#module-bioat.lib.libjgi" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIConfig">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libjgi.</span></span><span class="sig-name descname"><span class="pre">JGIConfig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">overwrite_conf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libjgi.html#JGIConfig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libjgi.JGIConfig" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIConfig.FILENAME_CONFIG_PATH">
<span class="sig-name descname"><span class="pre">FILENAME_CONFIG_PATH</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'/home/runner/.bioat/JGI/account.conf'</span></em><a class="headerlink" href="#bioat.lib.libjgi.JGIConfig.FILENAME_CONFIG_PATH" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIConfig.FILENAME_COOKIE">
<span class="sig-name descname"><span class="pre">FILENAME_COOKIE</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'/home/runner/.bioat/JGI/cookie'</span></em><a class="headerlink" href="#bioat.lib.libjgi.JGIConfig.FILENAME_COOKIE" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIConfig.FILENAME_TEMPLATE_LOG_FAIL">
<span class="sig-name descname"><span class="pre">FILENAME_TEMPLATE_LOG_FAIL</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'jgi-xml-query.failed_{}.log'</span></em><a class="headerlink" href="#bioat.lib.libjgi.JGIConfig.FILENAME_TEMPLATE_LOG_FAIL" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIConfig.FILENAME_TEMPLATE_XML">
<span class="sig-name descname"><span class="pre">FILENAME_TEMPLATE_XML</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'jgi-xml-query.result_{}.xml'</span></em><a class="headerlink" href="#bioat.lib.libjgi.JGIConfig.FILENAME_TEMPLATE_XML" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIConfig.URL_JGI_FETCH_XML">
<span class="sig-name descname"><span class="pre">URL_JGI_FETCH_XML</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'https://genome.jgi.doe.gov/portal/ext-api/downloads/get-directory'</span></em><a class="headerlink" href="#bioat.lib.libjgi.JGIConfig.URL_JGI_FETCH_XML" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIConfig.URL_JGI_LOGIN">
<span class="sig-name descname"><span class="pre">URL_JGI_LOGIN</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'https://signon.jgi.doe.gov/signon/create'</span></em><a class="headerlink" href="#bioat.lib.libjgi.JGIConfig.URL_JGI_LOGIN" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIConfig.URL_JGI_MAIN">
<span class="sig-name descname"><span class="pre">URL_JGI_MAIN</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'https://genome.jgi.doe.gov'</span></em><a class="headerlink" href="#bioat.lib.libjgi.JGIConfig.URL_JGI_MAIN" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIConfig.input_user_info">
<span class="sig-name descname"><span class="pre">input_user_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libjgi.html#JGIConfig.input_user_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libjgi.JGIConfig.input_user_info" title="Link to this definition"></a></dt>
<dd><p>Dialog with user to gather user information for
use with the curl query. Returns a dict.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIConfig.load_config">
<span class="sig-name descname"><span class="pre">load_config</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libjgi.html#JGIConfig.load_config"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libjgi.JGIConfig.load_config" title="Link to this definition"></a></dt>
<dd><p>Reads “user”, “password” and “categories” entries
from config file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIConfig.save_config">
<span class="sig-name descname"><span class="pre">save_config</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libjgi.html#JGIConfig.save_config"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libjgi.JGIConfig.save_config" title="Link to this definition"></a></dt>
<dd><p>Creates a config file &lt;config_path&gt; using
credentials from dict &lt;config_info&gt;.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIDoc">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libjgi.</span></span><span class="sig-name descname"><span class="pre">JGIDoc</span></span><a class="reference internal" href="_modules/bioat/lib/libjgi.html#JGIDoc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libjgi.JGIDoc" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIDoc.DEFAULT_CATEGORIES">
<span class="sig-name descname"><span class="pre">DEFAULT_CATEGORIES</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['ESTs',</span> <span class="pre">'EST</span> <span class="pre">Clusters',</span> <span class="pre">'Assembled</span> <span class="pre">scaffolds</span> <span class="pre">(unmasked)',</span> <span class="pre">'Assembled</span> <span class="pre">scaffolds</span> <span class="pre">(masked)',</span> <span class="pre">'Transcripts',</span> <span class="pre">'Genes',</span> <span class="pre">'CDS',</span> <span class="pre">'Proteins',</span> <span class="pre">'Additional</span> <span class="pre">Files']</span></em><a class="headerlink" href="#bioat.lib.libjgi.JGIDoc.DEFAULT_CATEGORIES" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIDoc.select_blurb">
<span class="sig-name descname"><span class="pre">select_blurb</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">&quot;</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">#</span> <span class="pre">SYNTAX</span> <span class="pre">///////////////////////////////////////////////////////////////////////\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">Using</span> <span class="pre">the</span> <span class="pre">following</span> <span class="pre">format</span> <span class="pre">syntax</span> <span class="pre">to</span> <span class="pre">download</span> <span class="pre">selected</span> <span class="pre">file:\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">&lt;category</span> <span class="pre">number&gt;:&lt;i&gt;[,&lt;i&gt;,</span> <span class="pre">&lt;i&gt;];&lt;category</span> <span class="pre">number&gt;:&lt;i&gt;-&lt;i&gt;;...\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">Indices</span> <span class="pre">(&lt;i&gt;)</span> <span class="pre">may</span> <span class="pre">be</span> <span class="pre">a</span> <span class="pre">mixture</span> <span class="pre">of</span> <span class="pre">comma-separated</span> <span class="pre">values</span> <span class="pre">and</span> <span class="pre">hyphen-separated</span> <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">ranges.</span> <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">Example:\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">'3:4,5;</span> <span class="pre">7:1-10,13'</span> <span class="pre">will</span> <span class="pre">select</span> <span class="pre">elements</span> <span class="pre">4</span> <span class="pre">and</span> <span class="pre">5</span> <span class="pre">from</span> <span class="pre">category</span> <span class="pre">3,</span> <span class="pre">and</span> <span class="pre">1-10</span> <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">plus</span> <span class="pre">13</span> <span class="pre">from</span> <span class="pre">category</span> <span class="pre">7.\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">#</span> <span class="pre">/SYNTAX</span> <span class="pre">///////////////////////////////////////////////////////////////////////\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">&quot;</span></em><a class="headerlink" href="#bioat.lib.libjgi.JGIDoc.select_blurb" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIDoc.usage_example_blurb">
<span class="sig-name descname"><span class="pre">usage_example_blurb</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">This</span> <span class="pre">tool</span> <span class="pre">will</span> <span class="pre">retrieve</span> <span class="pre">files</span> <span class="pre">from</span> <span class="pre">JGI\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">It</span> <span class="pre">will</span> <span class="pre">return</span> <span class="pre">a</span> <span class="pre">list</span> <span class="pre">of</span> <span class="pre">possible</span> <span class="pre">files</span> <span class="pre">for</span> <span class="pre">downloading.\n</span>&#160;&#160;&#160; <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">To</span> <span class="pre">get</span> <span class="pre">&lt;jgi_address&gt;,</span> <span class="pre">go</span> <span class="pre">to:</span> <span class="pre">http://genome.jgi.doe.gov/</span> <span class="pre">and</span> <span class="pre">search</span> <span class="pre">for</span> <span class="pre">your\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">species</span> <span class="pre">of</span> <span class="pre">interest.</span> <span class="pre">Click</span> <span class="pre">through</span> <span class="pre">until</span> <span class="pre">you</span> <span class="pre">are</span> <span class="pre">at</span> <span class="pre">the</span> <span class="pre">&quot;Info&quot;</span> <span class="pre">page.</span> <span class="pre">For\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">\x1b[3mNematostella</span> <span class="pre">vectensis\x1b[23m,</span> <span class="pre">the</span> <span class="pre">appropriate</span> <span class="pre">page</span> <span class="pre">is\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">&quot;http://genome.jgi.doe.gov/Nemve1/Nemve1.info.html&quot;.\n</span>&#160;&#160;&#160; <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">To</span> <span class="pre">query</span> <span class="pre">using</span> <span class="pre">only</span> <span class="pre">the</span> <span class="pre">name</span> <span class="pre">simply</span> <span class="pre">requires</span> <span class="pre">the</span> <span class="pre">specific</span> <span class="pre">JGI</span> <span class="pre">organism\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">abbreviation,</span> <span class="pre">as</span> <span class="pre">referenced</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">full</span> <span class="pre">url.\n</span>&#160;&#160;&#160; <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">For</span> <span class="pre">the</span> <span class="pre">above</span> <span class="pre">example,</span> <span class="pre">the</span> <span class="pre">proper</span> <span class="pre">input</span> <span class="pre">syntax</span> <span class="pre">for</span> <span class="pre">this</span> <span class="pre">script</span> <span class="pre">would</span> <span class="pre">be:\n</span>&#160;&#160;&#160; <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">$</span> <span class="pre">bioat</span> <span class="pre">meta</span> <span class="pre">JGI_query</span> <span class="pre">-q</span> <span class="pre">http://genome.jgi.doe.gov/Nemve1/Nemve1.info.html\n</span>&#160;&#160;&#160; <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">-or-\n</span>&#160;&#160;&#160; <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">$</span> <span class="pre">bioat</span> <span class="pre">meta</span> <span class="pre">JGI_query</span> <span class="pre">-q</span> <span class="pre">Nemve1\n</span>&#160;&#160;&#160; <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">If</span> <span class="pre">you</span> <span class="pre">already</span> <span class="pre">have</span> <span class="pre">the</span> <span class="pre">XML</span> <span class="pre">file</span> <span class="pre">for</span> <span class="pre">the</span> <span class="pre">query</span> <span class="pre">in</span> <span class="pre">the</span> <span class="pre">directory,</span> <span class="pre">you</span> <span class="pre">may</span> <span class="pre">use\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">the</span> <span class="pre">--xml</span> <span class="pre">flag</span> <span class="pre">to</span> <span class="pre">avoid</span> <span class="pre">redownloading</span> <span class="pre">it</span> <span class="pre">(particularly</span> <span class="pre">useful</span> <span class="pre">if</span> <span class="pre">querying\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">large,</span> <span class="pre">top-level</span> <span class="pre">groups</span> <span class="pre">with</span> <span class="pre">many</span> <span class="pre">sub-species,</span> <span class="pre">such</span> <span class="pre">as</span> <span class="pre">&quot;fungi&quot;):\n</span>&#160;&#160;&#160; <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">$</span> <span class="pre">bioat</span> <span class="pre">meta</span> <span class="pre">JGI_query</span> <span class="pre">-x</span> <span class="pre">&lt;your_xml_index&gt;\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">If</span> <span class="pre">the</span> <span class="pre">XML</span> <span class="pre">filename</span> <span class="pre">is</span> <span class="pre">omitted</span> <span class="pre">when</span> <span class="pre">using</span> <span class="pre">the</span> <span class="pre">-x/--xml</span> <span class="pre">flag,</span> <span class="pre">it</span> <span class="pre">is</span> <span class="pre">assumed</span> <span class="pre">that\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">the</span> <span class="pre">XML</span> <span class="pre">file</span> <span class="pre">is</span> <span class="pre">named</span> <span class="pre">&quot;jgi-xml-query.result_&lt;organism-name&gt;.xml&quot;.</span> <span class="pre">In</span> <span class="pre">such</span> <span class="pre">cases,</span> <span class="pre">the\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">organism</span> <span class="pre">name</span> <span class="pre">is</span> <span class="pre">required.\n</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">'</span></em><a class="headerlink" href="#bioat.lib.libjgi.JGIDoc.usage_example_blurb" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIOperator">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libjgi.</span></span><span class="sig-name descname"><span class="pre">JGIOperator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_info</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xml</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_fails</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nretry</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regex</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">all_get</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite_conf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_files</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proxy_pool</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">just_query_xml</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">syntax_help</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">usage</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'INFO'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libjgi.html#JGIOperator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libjgi.JGIOperator" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIOperator.download">
<span class="sig-name descname"><span class="pre">download</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libjgi.html#JGIOperator.download"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libjgi.JGIOperator.download" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIOperator.parse_xml">
<span class="sig-name descname"><span class="pre">parse_xml</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libjgi.html#JGIOperator.parse_xml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libjgi.JGIOperator.parse_xml" title="Link to this definition"></a></dt>
<dd><p>Moves through the xml document &lt;xml_file&gt; and returns information
about matches to elements in &lt;DESIRED_CATEGORIES&gt; if
&lt;filter_categories&gt; is True, or all files otherwise</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libjgi.JGIOperator.query">
<span class="sig-name descname"><span class="pre">query</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libjgi.html#JGIOperator.query"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libjgi.JGIOperator.query" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-bioat.lib.libpandas">
<span id="bioat-lib-libpandas-module"></span><h2>bioat.lib.libpandas module<a class="headerlink" href="#module-bioat.lib.libpandas" title="Link to this heading"></a></h2>
<p>_summary_</p>
<p>author: Herman Huanan Zhao
email: <a class="reference external" href="mailto:hermanzhaozzzz&#37;&#52;&#48;gmail&#46;com">hermanzhaozzzz<span>&#64;</span>gmail<span>&#46;</span>com</a>
homepage: <a class="reference external" href="https://github.com/hermanzhaozzzz">https://github.com/hermanzhaozzzz</a></p>
<p>_description_</p>
<dl>
<dt>example 1:</dt><dd><dl>
<dt>bioat list</dt><dd><dl>
<dt>&lt;in shell&gt;:</dt><dd><p>$ bioat list</p>
</dd>
<dt>&lt;in python consolo&gt;:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">bioat.cli</span><span class="w"> </span><span class="kn">import</span> <span class="n">Cli</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bioat</span> <span class="o">=</span> <span class="n">Cli</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bioat</span><span class="o">.</span><span class="n">list</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bioat</span><span class="o">.</span><span class="n">list</span><span class="p">())</span>
</pre></div>
</div>
</dd>
</dl>
</dd>
</dl>
</dd>
<dt>example 2:</dt><dd><p>_example_</p>
</dd>
</dl>
<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libpandas.set_option">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libpandas.</span></span><span class="sig-name descname"><span class="pre">set_option</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">progres_bar</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_colwidth</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">40</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">display_width</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">120</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">display_max_columns</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">display_max_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'INFO'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libpandas.html#set_option"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libpandas.set_option" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-bioat.lib.libpatentseq">
<span id="bioat-lib-libpatentseq-module"></span><h2>bioat.lib.libpatentseq module<a class="headerlink" href="#module-bioat.lib.libpatentseq" title="Link to this heading"></a></h2>
<p>_summary_</p>
<p>author: Herman Huanan Zhao
email: <a class="reference external" href="mailto:hermanzhaozzzz&#37;&#52;&#48;gmail&#46;com">hermanzhaozzzz<span>&#64;</span>gmail<span>&#46;</span>com</a>
homepage: <a class="reference external" href="https://github.com/hermanzhaozzzz">https://github.com/hermanzhaozzzz</a></p>
<p>_description_</p>
<dl>
<dt>example 1:</dt><dd><dl>
<dt>bioat list</dt><dd><dl>
<dt>&lt;in shell&gt;:</dt><dd><p>$ bioat list</p>
</dd>
<dt>&lt;in python consolo&gt;:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">bioat.cli</span><span class="w"> </span><span class="kn">import</span> <span class="n">Cli</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bioat</span> <span class="o">=</span> <span class="n">Cli</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bioat</span><span class="o">.</span><span class="n">list</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bioat</span><span class="o">.</span><span class="n">list</span><span class="p">())</span>
</pre></div>
</div>
</dd>
</dl>
</dd>
</dl>
</dd>
<dt>example 2:</dt><dd><p>_example_</p>
</dd>
</dl>
<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libpatentseq.run">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libpatentseq.</span></span><span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">playwright</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">username</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">password</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_header</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proxy_server</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">headless</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nretry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_browser</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rm_fail_cookie</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/bioat/lib/libpatentseq.html#run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libpatentseq.run" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-bioat.lib.libpath">
<span id="bioat-lib-libpath-module"></span><h2>bioat.lib.libpath module<a class="headerlink" href="#module-bioat.lib.libpath" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libpath.check_cmd">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libpath.</span></span><span class="sig-name descname"><span class="pre">check_cmd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'WARNING'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/bioat/lib/libpath.html#check_cmd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libpath.check_cmd" title="Link to this definition"></a></dt>
<dd><p>Check if a command is available in the system’s PATH.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>x</strong> (<em>str</em>) – The command name to check.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the command is executable and found in PATH, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libpath.check_executable">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libpath.</span></span><span class="sig-name descname"><span class="pre">check_executable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'WARNING'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/bioat/lib/libpath.html#check_executable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libpath.check_executable" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libpath.is_file">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libpath.</span></span><span class="sig-name descname"><span class="pre">is_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'WARNING'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/bioat/lib/libpath.html#is_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libpath.is_file" title="Link to this definition"></a></dt>
<dd><p>Check if the given path is a file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>x</strong> (<em>str</em><em> | </em><em>Path</em>) – The path to check.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if the path is a file, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-bioat.lib.libpdb">
<span id="bioat-lib-libpdb-module"></span><h2>bioat.lib.libpdb module<a class="headerlink" href="#module-bioat.lib.libpdb" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libpdb.get_cut2ref_aln_info">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libpdb.</span></span><span class="sig-name descname"><span class="pre">get_cut2ref_aln_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cal_rmsd</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cal_tmscore</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ref'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cut'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">usalign_bin</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'usalign'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'WARNING'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="reference internal" href="_modules/bioat/lib/libpdb.html#get_cut2ref_aln_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libpdb.get_cut2ref_aln_info" title="Link to this definition"></a></dt>
<dd><p>Align cutted pdb to ref pdb using the CA atoms.
Aligns a truncated protein structure (cut) to its full-length reference structure (ref)
using Cα atoms and Biopython’s Superimposer.</p>
<p>This function:
- Extracts all Cα atoms from <cite>ref</cite> and <cite>cut</cite>
- Removes atoms from <cite>ref</cite> at the indices listed in <cite>gap_indices</cite>
- Aligns the remaining atoms from <cite>cut</cite> to the corresponding positions in <cite>ref</cite>
- Modifies the <cite>cut</cite> structure in-place to match the aligned orientation
- Returns both structures and the RMSD value of the alignment</p>
<p>It assumes:
- One-to-one correspondence between residues after gap removal
- Structures are predicted by AlphaFold2 / ESMFold (no missing atoms)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>Bio.PDB.Structure.Structure</em>) – Reference structure path or loaded Structure.</p></li>
<li><p><strong>cut</strong> (<em>str</em><em> or </em><em>Bio.PDB.Structure.Structure</em>) – Truncated structure path or loaded Structure.</p></li>
<li><p><strong>cal_rmsd</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to calculate RMSD. Default is True.</p></li>
<li><p><strong>cal_tmscore</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to calculate TM-score using USalign. Default is False.</p></li>
<li><p><strong>label1</strong> (<em>str</em><em>, </em><em>optional</em>) – Name for the reference structure. Default is “ref”.</p></li>
<li><p><strong>label2</strong> (<em>str</em><em>, </em><em>optional</em>) – Name for the cut structure. Default is “cut”.</p></li>
<li><p><strong>usalign_bin</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the USalign binary for TM-score calculation. Default is “usalign”.</p></li>
<li><p><strong>log_level</strong> (<em>str</em><em>, </em><em>optional</em>) – Logging level. Default is “WARNING”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl>
<dt>{</dt><dd><blockquote>
<div><p>“{label1}”: aln label1 structure,  # if cal_rmsd is True, unaltered label1 structure
“{label2}}”: fixed label2 structure,  # if cal_rmsd is True, fix label2 coords in-place
“RMSD”: 0.123  # if cal_rmsd is True, the RMSD value between label1 and label2
f”{label1}_seq”: ref_seq,  # if cal_rmsd is True, the sequence of label1 structure
f”{label2}_seq”: cut_seq,  # if cal_rmsd is True, the sequence of label2 structure
“alignment_dict”: alignment_dict,  # if cal_rmsd is True, the alignment dict of label1 and label2
“gap_indices”: gap_indices,  # if cal_rmsd is True, the indices of gaps in label1 structure
“TM-score:mean”: 0.623,  # if cal_tmscore is True, the mean TM-score value
“TM-score:TM1”: 0.456,  # if cal_tmscore is True, use label1 as ref &lt;L_N&gt; in calculation
“TM-score:TM2”: 0.789,  # if cal_tmscore is True, use label2 as ref &lt;L_N&gt; in calculation
…</p>
</div></blockquote>
<p>}</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libpdb.load_structure">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libpdb.</span></span><span class="sig-name descname"><span class="pre">load_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'WARNING'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libpdb.html#load_structure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libpdb.load_structure" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libpdb.show_ref_cut">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libpdb.</span></span><span class="sig-name descname"><span class="pre">show_ref_cut</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_pdb</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Seq</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_pdb</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Structure</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Structure</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_color</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_map_colors</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_map_values</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lightgray'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'purple'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cartoon'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cartoon'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cartoon'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_map_value_random</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_fig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_interval</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'WARNING'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libpdb.html#show_ref_cut"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libpdb.show_ref_cut" title="Link to this definition"></a></dt>
<dd><p>Visualizes the alignment of sequences and highlights changes in PDB structures using py3Dmol.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_seq</strong> (<em>str</em><em> or </em><em>Seq</em>) – Amino acid sequence content for the ref protein.</p></li>
<li><p><strong>ref_pdb</strong> (<em>str</em><em> or </em><em>Bio.PDB.Structure.Structure</em>) – Path to the PDB file of the reference structure.</p></li>
<li><p><strong>cut_seq</strong> (<em>str</em><em>, </em><em>Seq</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Amino acid sequence content for the cut protein.</p></li>
<li><p><strong>cut_pdb</strong> (<em>str</em><em>, </em><em>Bio.PDB.Structure.Structure</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Path to the PDB file of the cut structure.</p></li>
<li><p><strong>cut_labels</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Label for the cut proteins. If None, the label will be set to “cut”.</p></li>
<li><p><strong>ref_color</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for reference residues.</p></li>
<li><p><strong>ref_map_colors</strong> (<em>tuple</em><em>[</em><em>str</em><em>, </em><em>str</em><em>] or </em><em>None</em><em>, </em><em>optional</em>) – ref_map_colors will be used as color bar from ref_map_colors[0] to ref_map_colors[1]. If None, do not apply color mapping. Defaults to None.</p></li>
<li><p><strong>ref_map_values</strong> (<em>dict</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – A dictionary of values for the ref color map, it will be normalized to the range of [0 - 1]. If None, all residues will be colored with the same color. e.g. ref_map_values = {‘V_0’: 0.4177215189873418, ‘S_1’: 0.8185654008438819, ‘K_2’: 0.9915611814345991, ‘G_3’: 0.42616033755274263, …}</p></li>
<li><p><strong>cut_color</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for cut residues.</p></li>
<li><p><strong>gap_color</strong> (<em>str</em><em>, </em><em>optional</em>) – Color for gaps or removed residues.</p></li>
<li><p><strong>ref_style</strong> (<em>str</em><em>, </em><em>optional</em>) – “stick”, “sphere”, “cartoon”, or “line”</p></li>
<li><p><strong>cut_style</strong> (<em>str</em><em>, </em><em>optional</em>) – “stick”, “sphere”, “cartoon”, or “line”</p></li>
<li><p><strong>gap_style</strong> (<em>str</em><em>, </em><em>optional</em>) – “stick”, “sphere”, “cartoon”, or “line”</p></li>
<li><p><strong>ref_map_value_random</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, ref_map_values will be randomly generated. Defaults to False.</p></li>
<li><p><strong>output_fig</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Output figure file path. If None, the figure will not be saved in html format. Defaults to None.</p></li>
<li><p><strong>col</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of columns for the visualization. Defaults to 3.</p></li>
<li><p><strong>scale</strong> (<em>float</em><em>, </em><em>optional</em>) – Scale factor for the visualization. Defaults to 1.0.</p></li>
<li><p><strong>annotate</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to annotate the visualization with labels. Defaults to True.</p></li>
<li><p><strong>text_interval</strong> (<em>int</em><em>, </em><em>optional</em>) – The interval between text annotations. Defaults to 5.</p></li>
<li><p><strong>log_level</strong> (<em>str</em><em>, </em><em>optional</em>) – Log level. Defaults to “WARNING”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libpdb.structure2string">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libpdb.</span></span><span class="sig-name descname"><span class="pre">structure2string</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Structure</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libpdb.html#structure2string"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libpdb.structure2string" title="Link to this definition"></a></dt>
<dd><p>Convert Bio.PDB.Structure.Structure to string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>structure</strong> (<em>Bio.PDB.Structure.Structure</em>) – Structure object to convert.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>PDB context。</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-bioat.lib.libphylo">
<span id="bioat-lib-libphylo-module"></span><h2>bioat.lib.libphylo module<a class="headerlink" href="#module-bioat.lib.libphylo" title="Link to this heading"></a></h2>
</section>
<section id="module-bioat.lib.libplot">
<span id="bioat-lib-libplot-module"></span><h2>bioat.lib.libplot module<a class="headerlink" href="#module-bioat.lib.libplot" title="Link to this heading"></a></h2>
<p>bioat.lib.libplot</p>
<p>author: Herman Huanan Zhao
email: <a class="reference external" href="mailto:hermanzhaozzzz&#37;&#52;&#48;gmail&#46;com">hermanzhaozzzz<span>&#64;</span>gmail<span>&#46;</span>com</a>
homepage: <a class="reference external" href="https://github.com/hermanzhaozzzz">https://github.com/hermanzhaozzzz</a></p>
<p>This module provides functions for plotting.</p>
<dl>
<dt>example 1:</dt><dd><dl>
<dt>init_matplotlib</dt><dd><dl>
<dt>&lt;in python consolo&gt;:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">bioat.lib.libplot</span><span class="w"> </span><span class="kn">import</span> <span class="n">init_matplotlib</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">init_matplotlib</span><span class="p">(</span><span class="n">log_level</span><span class="o">=</span><span class="s1">&#39;info&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</dd>
</dl>
</dd>
</dl>
</dd>
<dt>example 2:</dt><dd><dl>
<dt>plot_colortable</dt><dd><dl>
<dt>&lt;in python consolo&gt;:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">bioat.lib.libplot</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_colortable</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;#64C1E8&#39;</span><span class="p">,</span> <span class="s1">&#39;#80CED7&#39;</span><span class="p">,</span> <span class="s1">&#39;#63C7B2&#39;</span><span class="p">,</span> <span class="s1">&#39;#8E6C88&#39;</span><span class="p">,</span> <span class="s1">&#39;#CA61C3&#39;</span><span class="p">,</span> <span class="s1">&#39;#FF958C&#39;</span><span class="p">,</span> <span class="s1">&#39;#883677&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plot_colortable</span><span class="p">(</span><span class="n">colors</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libplot.init_matplotlib">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libplot.</span></span><span class="sig-name descname"><span class="pre">init_matplotlib</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">style</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'seaborn-v0_8-paper'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Helvetica'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">refresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'INFO'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libplot.html#init_matplotlib"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libplot.init_matplotlib" title="Link to this definition"></a></dt>
<dd><p>easily set matplotlib style</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>style</strong> (<em>str</em><em>, </em><em>optional</em>) – matplotlib style, defaults to ‘seaborn-v0_8-paper’</p></li>
<li><p><strong>font</strong> (<em>str</em><em>, </em><em>optional</em>) – use what font in matplotlib, defaults to ‘Helvetica’</p></li>
<li><p><strong>refresh</strong> (<em>bool</em><em>, </em><em>optional</em>) – wether to remove matplotlib font cache and reset it, defaults to False</p></li>
<li><p><strong>log_level</strong> (<em>str</em><em>, </em><em>optional</em>) – log level, defaults to ‘INFO’</p></li>
<li><p><strong>set_backend_pdf</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether to use core fonts for the PDF backend, defaults to True</p></li>
<li><p><strong>set_backend_ps</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether to use core fonts for the PS backend, defaults to True</p></li>
<li><p><strong>set_backend_svg</strong> (<em>str</em><em>, </em><em>optional</em>) – whether to use ‘none’ to replace ‘path’ (use font but not plot path for characters)for the SVG backend, defaults to ‘none’</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference internal" href="bioat.html#bioat.exceptions.BioatError" title="bioat.exceptions.BioatError"><strong>BioatError</strong></a> – if failed to copy fonts</p></li>
<li><p><a class="reference internal" href="bioat.html#bioat.exceptions.BioatRuntimeError" title="bioat.exceptions.BioatRuntimeError"><strong>BioatRuntimeError</strong></a> – if site-packages not found in sys.path</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libplot.plot_colortable">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libplot.</span></span><span class="sig-name descname"><span class="pre">plot_colortable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">colors</span></span></em>, <em class="sig-param"><span class="keyword-only-separator o"><abbr title="Keyword-only parameters separator (PEP 3102)"><span class="pre">*</span></abbr></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libplot.html#plot_colortable"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libplot.plot_colortable" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-bioat.lib.libsnakemake">
<span id="bioat-lib-libsnakemake-module"></span><h2>bioat.lib.libsnakemake module<a class="headerlink" href="#module-bioat.lib.libsnakemake" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libsnakemake.check_cmd">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libsnakemake.</span></span><span class="sig-name descname"><span class="pre">check_cmd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libsnakemake.html#check_cmd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libsnakemake.check_cmd" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libsnakemake.check_read">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libsnakemake.</span></span><span class="sig-name descname"><span class="pre">check_read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libsnakemake.html#check_read"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libsnakemake.check_read" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libsnakemake.print_head">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libsnakemake.</span></span><span class="sig-name descname"><span class="pre">print_head</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">SAMPLES</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">MODE</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libsnakemake.html#print_head"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libsnakemake.print_head" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-bioat.lib.libspider">
<span id="bioat-lib-libspider-module"></span><h2>bioat.lib.libspider module<a class="headerlink" href="#module-bioat.lib.libspider" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="bioat.lib.libspider.ProxyPool">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bioat.lib.libspider.</span></span><span class="sig-name descname"><span class="pre">ProxyPool</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">url</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libspider.html#ProxyPool"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libspider.ProxyPool" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libspider.ProxyPool.delete_proxy">
<span class="sig-name descname"><span class="pre">delete_proxy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">proxy</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libspider.html#ProxyPool.delete_proxy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libspider.ProxyPool.delete_proxy" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bioat.lib.libspider.ProxyPool.get_proxy">
<span class="sig-name descname"><span class="pre">get_proxy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libspider.html#ProxyPool.get_proxy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libspider.ProxyPool.get_proxy" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="bioat.lib.libspider.get_random_user_agents">
<span class="sig-prename descclassname"><span class="pre">bioat.lib.libspider.</span></span><span class="sig-name descname"><span class="pre">get_random_user_agents</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bioat/lib/libspider.html#get_random_user_agents"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bioat.lib.libspider.get_random_user_agents" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-bioat.lib">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-bioat.lib" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="bioat.html" class="btn btn-neutral float-left" title="bioat package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>