

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bioat.lib.libcrispr &mdash; bioat Documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=27e171c2"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            bioat
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">Modules</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">bioat</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bioat.lib.libcrispr</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bioat.lib.libcrispr</h1><div class="highlight"><pre>
<span></span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">PairwiseAligner</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>

<span class="kn">from</span> <span class="nn">bioat.lib.libalignment</span> <span class="kn">import</span> <span class="n">get_alignment_info</span>
<span class="kn">from</span> <span class="nn">bioat.logger</span> <span class="kn">import</span> <span class="n">LoggerManager</span>

<span class="n">lm</span> <span class="o">=</span> <span class="n">LoggerManager</span><span class="p">(</span><span class="n">mod_name</span><span class="o">=</span><span class="s2">&quot;bioat.lib.libcrispr&quot;</span><span class="p">)</span>

<span class="c1"># TARGET_REGIONS_LIB</span>
<span class="c1">#     for target_seq alignment</span>
<span class="c1">#</span>
<span class="c1"># bioat target_seq region_heatmap test_sorted.mpileup.info.tsv test.pdf --get_built_in_target_region</span>
<span class="c1">#</span>
<span class="c1"># INFO:root:You can use &lt;key&gt; in built-in &lt;target_regions&gt; to represent your target_region:</span>
<span class="c1">#         &lt;key&gt;   &lt;target_region&gt;</span>
<span class="c1">#         EMX1    GAGTCCGAGCAGAAGAAGAA^NGG^</span>
<span class="c1">#         HEK3    GGCCCAGACTGAGCACGTGA^NGG^</span>
<span class="c1">#         HEK4    GGCACTGCGGCTGGAGGTGG^NGG^</span>
<span class="c1">#         ...</span>

<span class="n">TARGET_SEQ_LIB</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;EMX1&quot;</span><span class="p">:</span> <span class="s2">&quot;GAGTCCGAGCAGAAGAAGAA^NGG^&quot;</span><span class="p">,</span>
    <span class="s2">&quot;HEK3&quot;</span><span class="p">:</span> <span class="s2">&quot;GGCCCAGACTGAGCACGTGA^NGG^&quot;</span><span class="p">,</span>
    <span class="s2">&quot;HEK4&quot;</span><span class="p">:</span> <span class="s2">&quot;GGCACTGCGGCTGGAGGTGG^NGG^&quot;</span><span class="p">,</span>
    <span class="s2">&quot;RNF2&quot;</span><span class="p">:</span> <span class="s2">&quot;GTCATCTTAGTCATTACCTG^NGG^&quot;</span><span class="p">,</span>
    <span class="s2">&quot;VEGFA&quot;</span><span class="p">:</span> <span class="s2">&quot;GACCCCCTCCACCCCGCCTC^NGG^&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CDKN2A&quot;</span><span class="p">:</span> <span class="s2">&quot;^TTTA^GCCCCAATAATCCCCACATGTCA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;DYRK1A&quot;</span><span class="p">:</span> <span class="s2">&quot;^TTTA^GAAGCACATCAAGGACATTCTAA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ND4&quot;</span><span class="p">:</span> <span class="s2">&quot;TGCTAGTAACCACGTTCTCCTGATCAAATATCACTCTCCTACTTACAGGA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ND5.1&quot;</span><span class="p">:</span> <span class="s2">&quot;TAGCATTAGCAGGAATACCTTTCCTCACAGGTTTCTACTCCAAAGA&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ND6&quot;</span><span class="p">:</span> <span class="s2">&quot;TGACCCCCATGCCTCAGGATACTCCTCAATAGCCATCG&quot;</span><span class="p">,</span>
<span class="p">}</span>


<div class="viewcode-block" id="cmp_align_list">
<a class="viewcode-back" href="../../../bioat.lib.html#bioat.lib.libcrispr.cmp_align_list">[docs]</a>
<span class="k">def</span> <span class="nf">cmp_align_list</span><span class="p">(</span><span class="n">aln_a</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span> <span class="n">aln_b</span><span class="p">:</span> <span class="nb">dict</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compare function for alignment lists.</span>

<span class="sd">    This function processes alignment data and performs comparisons based on the input structure.</span>

<span class="sd">    Args:</span>
<span class="sd">        input_data (list | dict): </span>
<span class="sd">            The alignment data to be processed. Can be provided as:</span>
<span class="sd">            - A list, e.g.: [17, 3, 0, 73.0, 1, 22, &#39;AAGAAGAAGACGAGTCTGCA&#39;, &#39;||||||||||||||X|||XX&#39;, &#39;AAGAAGAAGACGAGCCTGAG&#39;]</span>
<span class="sd">            - A dictionary, e.g.: {&#39;match_count&#39;: 26, &#39;mismatch_count&#39;: 3, &#39;gap_count&#39;: 4, &#39;aln_score&#39;: 96.0, &#39;ref_aln_start&#39;: 0, &#39;ref_aln_end&#39;: 32, &#39;alignment&#39;: {&#39;reference_seq&#39;: &#39;GGCACTGCGGCTGGAAAAAAAAAAAAAAA--GT&#39;,&#39;aln_info&#39;: &#39;--..|.|||||||||||||||||||||||--||&#39;, &#39;target_seq&#39;: &#39;--GGCAGCGGCTGGAAAAAAAAAAAAAAAAGGT&#39;}}</span>

<span class="sd">    Returns:</span>
<span class="sd">        None: The function does not return a value but performs comparisons.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; compare_alignments([</span>
<span class="sd">        ...     17, 3, 0, 73.0, 1, 22, </span>
<span class="sd">        ...     &#39;AAGAAGAAGACGAGTCTGCA&#39;, </span>
<span class="sd">        ...     &#39;||||||||||||||X|||XX&#39;, </span>
<span class="sd">        ...     &#39;AAGAAGAAGACGAGCCTGAG&#39;</span>
<span class="sd">        ... ])</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># sort reason score -&gt; gap -&gt; mismatch -&gt; match</span>
    <span class="k">def</span> <span class="nf">sign_value</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        HELP</span>
<span class="sd">            sign function</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">x</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="o">-</span><span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="mi">0</span>

    <span class="n">align_alphabet_score</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;-&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;|&quot;</span><span class="p">:</span> <span class="mi">2</span><span class="p">}</span>
    <span class="n">sort_keys</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;alignment_score&#39;</span><span class="p">,</span> <span class="s1">&#39;gap_count&#39;</span><span class="p">,</span> <span class="s1">&#39;mismatch_count&#39;</span><span class="p">,</span> <span class="s1">&#39;match_count&#39;</span><span class="p">]</span>
    <span class="n">sort_rev_state_list</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sort_keys</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">aln_a</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">-</span> <span class="n">aln_b</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">sort_rev_state_list</span><span class="p">[</span><span class="n">idx</span><span class="p">]:</span>
                <span class="k">return</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">sign_value</span><span class="p">(</span><span class="n">aln_a</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">-</span> <span class="n">aln_b</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">sign_value</span><span class="p">(</span><span class="n">aln_a</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">-</span> <span class="n">aln_b</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>

    <span class="n">aln_info_a</span> <span class="o">=</span> <span class="n">aln_a</span><span class="p">[</span><span class="s1">&#39;alignment&#39;</span><span class="p">][</span><span class="s1">&#39;aln_info&#39;</span><span class="p">]</span>
    <span class="n">aln_info_b</span> <span class="o">=</span> <span class="n">aln_b</span><span class="p">[</span><span class="s1">&#39;alignment&#39;</span><span class="p">][</span><span class="s1">&#39;aln_info&#39;</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">char_a</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">aln_info_a</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="o">&lt;=</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">aln_info_b</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
            <span class="n">value_a</span> <span class="o">=</span> <span class="n">align_alphabet_score</span><span class="p">[</span><span class="n">char_a</span><span class="p">]</span>
            <span class="n">value_b</span> <span class="o">=</span> <span class="n">align_alphabet_score</span><span class="p">[</span><span class="n">aln_info_a</span><span class="p">[</span><span class="n">idx</span><span class="p">]]</span>
            <span class="k">return</span> <span class="n">sign_value</span><span class="p">(</span><span class="n">value_a</span> <span class="o">-</span> <span class="n">value_b</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span></div>



<div class="viewcode-block" id="run_target_seq_align">
<a class="viewcode-back" href="../../../bioat.lib.html#bioat.lib.libcrispr.run_target_seq_align">[docs]</a>
<span class="k">def</span> <span class="nf">run_target_seq_align</span><span class="p">(</span>
        <span class="n">ref_seq</span><span class="p">:</span> <span class="n">Seq</span><span class="p">,</span>
        <span class="n">target_seq</span><span class="p">:</span> <span class="n">Seq</span><span class="p">,</span>
        <span class="n">aligner</span><span class="p">:</span> <span class="n">PairwiseAligner</span><span class="p">,</span>
        <span class="n">PAM</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">log_level</span><span class="o">=</span><span class="s1">&#39;WARNING&#39;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Perform global alignment for the target sequence.</span>

<span class="sd">    This function performs global alignment between a reference sequence and a target sequence (usually the sgRNA sequence without PAM) using the specified aligner.</span>

<span class="sd">    Args:</span>
<span class="sd">        ref_seq (Seq): </span>
<span class="sd">            A Seq object from BioPython representing the reference sequence </span>
<span class="sd">            for the targeted deep sequencing.</span>
<span class="sd">        target_seq (Seq): </span>
<span class="sd">            A Seq object from BioPython representing the target region sequence </span>
<span class="sd">            for the editing window (usually the sgRNA sequence without PAM).</span>
<span class="sd">        aligner (object): </span>
<span class="sd">            A PairwiseAligner object from BioPython used for performing the alignment.</span>
<span class="sd">        PAM (dict, optional): </span>
<span class="sd">            A dictionary containing PAM information with the following keys:</span>
<span class="sd">            - &#39;PAM&#39; (str): The PAM sequence (e.g., &quot;AGG&quot;).</span>
<span class="sd">            - &#39;position&#39; (int): The insertion site of `target_seq`.</span>
<span class="sd">            - &#39;weight&#39; (float): The priority weight for PAM alignment. The alignment score will be multiplied by this weight.</span>
<span class="sd">            - Example: {&#39;PAM&#39;: &#39;AGG&#39;, &#39;position&#39;: 20, &#39;weight&#39;: 1.0}.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list[dict]: </span>
<span class="sd">            A list of dictionaries, each containing alignment results. Each dictionary has the following keys:</span>
<span class="sd">            - &#39;match_count&#39; (int): The number of matching bases in the alignment.</span>
<span class="sd">            - &#39;mismatch_count&#39; (int): The number of mismatched bases.</span>
<span class="sd">            - &#39;gap_count&#39; (int): The number of gaps.</span>
<span class="sd">            - &#39;aln_score&#39; (float): The alignment score.</span>
<span class="sd">            - &#39;ref_aln_start&#39; (int): The starting position of the alignment on the reference sequence.</span>
<span class="sd">            - &#39;ref_aln_end&#39; (int): The ending position of the alignment on the reference sequence.</span>
<span class="sd">            - &#39;alignment&#39; (dict): The alignment details containing:&#39;reference_seq&#39; (str): The aligned reference sequence.&#39;aln_info&#39; (str): The alignment information (e.g., matches, mismatches, and gaps).&#39;target_seq&#39; (str): The aligned target sequence.</span>

<span class="sd">    Example:</span>
<span class="sd">        An example alignment result might look like this:[{&#39;match_count&#39;: 26,&#39;mismatch_count&#39;: 3,&#39;gap_count&#39;: 4,&#39;aln_score&#39;: 96.0,&#39;ref_aln_start&#39;: 0,&#39;ref_aln_end&#39;: 32,&#39;alignment&#39;: {&#39;reference_seq&#39;: &#39;GGCACTGCGGCTGGAAAAAAAAAAAAAAA--GT&#39;,&#39;aln_info&#39;: &#39;--..|.|||||||||||||||||||||||--||&#39;,&#39;target_seq&#39;: &#39;--GGCAGCGGCTGGAAAAAAAAAAAAAAAAGGT&#39;}}]</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">lm</span><span class="o">.</span><span class="n">set_names</span><span class="p">(</span><span class="n">func_name</span><span class="o">=</span><span class="s2">&quot;run_target_seq_align&quot;</span><span class="p">)</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">set_level</span><span class="p">(</span><span class="n">log_level</span><span class="p">)</span>


    <span class="k">if</span> <span class="n">PAM</span><span class="p">:</span>
        <span class="c1"># considering PAM</span>
        <span class="c1"># DNA info https://www.bioinformatics.org/sms/iupac.html</span>
        <span class="n">alignments_pam</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">ref_seq</span><span class="p">,</span> <span class="n">Seq</span><span class="p">(</span><span class="n">PAM</span><span class="p">[</span><span class="s1">&#39;PAM&#39;</span><span class="p">]))</span>
        <span class="c1"># use pam to find aim sequence on ref and then use target_seq to align them and give it a higher priority</span>
        <span class="n">dt_PAMs</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">alignments_pam</span><span class="p">):</span>
            <span class="n">alignment_analysis_result</span> <span class="o">=</span> <span class="n">get_alignment_info</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;alignment_analysis_result for PAM:</span><span class="se">\n</span><span class="si">{</span><span class="n">alignment_analysis_result</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">ref_aln_start</span> <span class="o">=</span> <span class="n">alignment_analysis_result</span><span class="p">[</span><span class="s1">&#39;ref_aln_start&#39;</span><span class="p">]</span>
            <span class="n">ref_aln_end</span> <span class="o">=</span> <span class="n">alignment_analysis_result</span><span class="p">[</span><span class="s1">&#39;ref_aln_end&#39;</span><span class="p">]</span>
            <span class="n">aln_score</span> <span class="o">=</span> <span class="n">alignment_analysis_result</span><span class="p">[</span><span class="s1">&#39;aln_score&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="n">PAM</span><span class="p">[</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span>
            <span class="n">dt_PAMs</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">ref_aln_start</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">ref_aln_end</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">aln_score</span><span class="o">=</span><span class="n">aln_score</span> <span class="o">*</span> <span class="n">PAM</span><span class="p">[</span><span class="s1">&#39;weight&#39;</span><span class="p">])</span>

        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;find PAMs on reference: </span><span class="si">{</span><span class="n">dt_PAMs</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="c1"># add PAM left or right or any position</span>
        <span class="n">target_seq</span> <span class="o">=</span> <span class="n">target_seq</span><span class="p">[:</span><span class="n">PAM</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]]</span> <span class="o">+</span> <span class="n">Seq</span><span class="p">(</span><span class="n">PAM</span><span class="p">[</span><span class="s1">&#39;PAM&#39;</span><span class="p">])</span> <span class="o">+</span> <span class="n">target_seq</span><span class="p">[</span><span class="n">PAM</span><span class="p">[</span><span class="s1">&#39;position&#39;</span><span class="p">]:]</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;find PAM on target_seq, target_seq is updated to: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">target_seq</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># alignment part</span>
    <span class="n">alignments</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">ref_seq</span><span class="p">,</span> <span class="n">target_seq</span><span class="p">)</span>
    <span class="c1"># parse alignment</span>
    <span class="n">alignment_results</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">alignment</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
        <span class="n">alignment_analysis_result</span> <span class="o">=</span> <span class="n">get_alignment_info</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;alignment_analysis_result:</span><span class="se">\n</span><span class="si">{</span><span class="n">alignment_analysis_result</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">alignment_results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment_analysis_result</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment_results</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;find 1 alignment result:</span><span class="se">\n</span><span class="si">{</span><span class="n">alignment_results</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">alignment_results</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;find </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">alignment_results</span><span class="p">)</span><span class="si">}</span><span class="s1"> alignment results:</span><span class="se">\n</span><span class="si">{</span><span class="n">alignment_results</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">alignment_results</span><span class="p">)</span>
        <span class="c1"># if contain multiple alignment result with the same score, keep the best one;</span>
        <span class="c1"># sort reason score -&gt; gap -&gt; mismatch -&gt; match</span>
        <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span>
            <span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;aln_score&#39;</span><span class="p">,</span> <span class="s1">&#39;gap_count&#39;</span><span class="p">,</span> <span class="s1">&#39;mismatch_count&#39;</span><span class="p">,</span> <span class="s1">&#39;match_count&#39;</span><span class="p">],</span>
            <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span>
            <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>