

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bioat.lib.libpdb &mdash; Unknown Project Documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=01f34227"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            Unknown Project
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">Modules</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Unknown Project</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bioat.lib.libpdb</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bioat.lib.libpdb</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;TODO.&quot;&quot;&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">math</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">shutil</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">subprocess</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">tempfile</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">io</span><span class="w"> </span><span class="kn">import</span> <span class="n">StringIO</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">cast</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">py3Dmol</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.MMCIFParser</span><span class="w"> </span><span class="kn">import</span> <span class="n">MMCIFParser</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.PDBIO</span><span class="w"> </span><span class="kn">import</span> <span class="n">PDBIO</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.PDBParser</span><span class="w"> </span><span class="kn">import</span> <span class="n">PDBParser</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.Structure</span><span class="w"> </span><span class="kn">import</span> <span class="n">Structure</span> <span class="k">as</span> <span class="n">BiopythonStructure</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB.Superimposer</span><span class="w"> </span><span class="kn">import</span> <span class="n">Superimposer</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqRecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.exceptions</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">BioatFileFormatError</span><span class="p">,</span>
    <span class="n">BioatFileNotFoundError</span><span class="p">,</span>
    <span class="n">BioatInvalidParameterError</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.lib.libalignment</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">get_aligned_seq</span><span class="p">,</span>
    <span class="n">get_best_alignment</span><span class="p">,</span>
    <span class="n">instantiate_pairwise_aligner</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.lib.libcolor</span><span class="w"> </span><span class="kn">import</span> <span class="n">map_colors_between_two</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.lib.libfastx</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">AMINO_ACIDS_1CODE</span><span class="p">,</span>
    <span class="n">AMINO_ACIDS_3CODE</span><span class="p">,</span>
    <span class="n">AMINO_ACIDS_3CODE_EXTEND</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.logger</span><span class="w"> </span><span class="kn">import</span> <span class="n">LoggerManager</span>

<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;get_cut2ref_aln_info&quot;</span><span class="p">,</span>
    <span class="s2">&quot;load_structure&quot;</span><span class="p">,</span>
    <span class="s2">&quot;show_ref_cut&quot;</span><span class="p">,</span>
    <span class="s2">&quot;structure2string&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="n">lm</span> <span class="o">=</span> <span class="n">LoggerManager</span><span class="p">(</span><span class="n">mod_name</span><span class="o">=</span><span class="s2">&quot;bioat.lib.libpdb&quot;</span><span class="p">)</span>
<span class="n">lm</span><span class="o">.</span><span class="n">set_level</span><span class="p">(</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">)</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_load_seq</span><span class="p">(</span><span class="n">seq</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">Seq</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

    <span class="k">if</span> <span class="p">(</span>
        <span class="nb">isinstance</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">Path</span><span class="p">)</span>
        <span class="ow">and</span> <span class="n">seq</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span>
        <span class="ow">and</span> <span class="n">seq</span><span class="o">.</span><span class="n">suffix</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;.fa&quot;</span><span class="p">]</span>
    <span class="p">):</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>
        <span class="n">label</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">id</span> <span class="k">if</span> <span class="n">label</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">label</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">seq</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">Seq</span><span class="p">):</span>
        <span class="n">label</span> <span class="o">=</span> <span class="s2">&quot;no_name&quot;</span> <span class="k">if</span> <span class="n">label</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">label</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Invalid sequence format. seq: </span><span class="si">{</span><span class="n">seq</span><span class="si">}</span><span class="s2">, type=</span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span><span class="si">}</span><span class="s2">, seq.is_file()=</span><span class="si">{</span><span class="n">seq</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span><span class="si">}</span><span class="s2">, seq.suffix.lower()=</span><span class="si">{</span><span class="n">seq</span><span class="o">.</span><span class="n">suffix</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">raise</span> <span class="n">BioatInvalidParameterError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">seq</span><span class="p">,</span> <span class="n">label</span>


<div class="viewcode-block" id="load_structure">
<a class="viewcode-back" href="../../../bioat.lib.html#bioat.lib.libpdb.load_structure">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">load_structure</span><span class="p">(</span>
    <span class="n">structure</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">BiopythonStructure</span><span class="p">,</span> <span class="n">label</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="n">BiopythonStructure</span><span class="p">,</span> <span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Load a PDB structure file or a BiopythonStructure object.</span>

<span class="sd">    Args:</span>
<span class="sd">        structure (str | Path | BiopythonStructure): Structure file path or BiopythonStructure object.</span>
<span class="sd">        label (str | None, optional): Structure label. If None, the label will be set to the file name. Defaults to None.</span>

<span class="sd">    Raises:</span>
<span class="sd">        BioatInvalidParameterError: If the input structure is not a valid file path or a BiopythonStructure object.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple[BiopythonStructure, str]: A tuple of BiopythonStructure object and label.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">structure</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">structure</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="n">Path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">structure</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
        <span class="n">parser</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">(</span><span class="n">QUIET</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">structure</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="n">BiopythonStructure</span><span class="p">,</span> <span class="n">parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">label</span><span class="p">,</span> <span class="n">structure</span><span class="p">))</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="n">BiopythonStructure</span><span class="p">):</span>
        <span class="n">label</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="n">structure</span><span class="o">.</span><span class="n">id</span> <span class="k">if</span> <span class="n">label</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">label</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">structure</span><span class="p">,</span> <span class="n">label</span>
    <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Invalid structure format. structure: </span><span class="si">{</span><span class="n">structure</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="k">raise</span> <span class="n">BioatInvalidParameterError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div>



<div class="viewcode-block" id="structure2string">
<a class="viewcode-back" href="../../../bioat.lib.html#bioat.lib.libpdb.structure2string">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">structure2string</span><span class="p">(</span><span class="n">structure</span><span class="p">:</span> <span class="n">BiopythonStructure</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert BiopythonStructure object to string format.</span>

<span class="sd">    Args:</span>
<span class="sd">        structure (BiopythonStructure): Structure object to convert.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: PDB context.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pdb_io</span> <span class="o">=</span> <span class="n">PDBIO</span><span class="p">()</span>
    <span class="n">pdb_io</span><span class="o">.</span><span class="n">set_structure</span><span class="p">(</span><span class="n">structure</span><span class="p">)</span>
    <span class="c1"># Use StringIO to save to memory string</span>
    <span class="n">pdb_string</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
    <span class="n">pdb_io</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">pdb_string</span><span class="p">)</span>
    <span class="c1"># Return string</span>
    <span class="k">return</span> <span class="n">pdb_string</span><span class="o">.</span><span class="n">getvalue</span><span class="p">()</span></div>



<span class="k">def</span><span class="w"> </span><span class="nf">pdb2fasta</span><span class="p">(</span>
    <span class="n">input_pdb</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">BiopythonStructure</span><span class="p">,</span>
    <span class="n">output_fasta</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">func_return</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">log_level</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">Seq</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Converts a PDB file to a FASTA file.</span>

<span class="sd">    Details:</span>
<span class="sd">        1. **Proteins**:</span>
<span class="sd">            The protein sequence for each chain will be extracted as &quot;Chain X Protein&quot;.</span>
<span class="sd">        2. **DNA and RNA**:</span>
<span class="sd">            Bases for DNA (A, T, G, C) will be saved as &quot;Chain X DNA&quot;, and bases for RNA (A, U, G, C) will be saved as &quot;Chain X RNA&quot;.</span>
<span class="sd">        3. **Other molecules**:</span>
<span class="sd">            Any unrecognized molecules (e.g., ions, modified molecules) will be labeled as [residue] and stored as &quot;Chain X Other molecules&quot;.</span>
<span class="sd">        4. **Multi-chain complexes**:</span>
<span class="sd">            The program supports multi-chain structures in complexes, and the content of each chain will be recorded separately.</span>

<span class="sd">    Args:</span>
<span class="sd">        input_pdb (str or Path or BiopythonStructure):</span>
<span class="sd">            Path to the input PDB/CIF file or Biopython Structure.</span>
<span class="sd">        output_fasta (str or Path, optional):</span>
<span class="sd">            Output file path. If None, the output file will be named as the</span>
<span class="sd">            basename of the input file with a &quot;.fa&quot; extension. Defaults to None.</span>
<span class="sd">        func_return: (bool, optional)</span>
<span class="sd">            Whether to return a list of SeqRecord objects, useful when used as a function but not for command line. Defaults to False.</span>
<span class="sd">        log_level (str, optional):</span>
<span class="sd">            Logging level. Defaults to &quot;WARNING&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of SeqRecord if func_return is True, otherwise None.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">set_names</span><span class="p">(</span><span class="n">func_name</span><span class="o">=</span><span class="s2">&quot;pdb2fasta&quot;</span><span class="p">)</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">set_level</span><span class="p">(</span><span class="n">log_level</span><span class="p">)</span>

    <span class="n">structure</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">structure_id</span> <span class="o">=</span> <span class="s2">&quot;structure&quot;</span>

    <span class="c1"># === Handle input file or structure object ===</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_pdb</span><span class="p">,</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span><span class="p">):</span>
        <span class="n">input_pdb</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">input_pdb</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">input_pdb</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Input file not found: </span><span class="si">{</span><span class="n">input_pdb</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">raise</span> <span class="n">BioatFileNotFoundError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="n">suffix</span> <span class="o">=</span> <span class="n">input_pdb</span><span class="o">.</span><span class="n">suffix</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="n">parser</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">MMCIFParser</span><span class="p">(</span><span class="n">QUIET</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">suffix</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.cif&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">suffix</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.cif.gz&quot;</span><span class="p">)</span>
            <span class="k">else</span> <span class="n">PDBParser</span><span class="p">(</span><span class="n">QUIET</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">suffix</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.pdb&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">suffix</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.pdb.gz&quot;</span><span class="p">)</span>
            <span class="k">else</span> <span class="kc">None</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">parser</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">BioatFileFormatError</span><span class="p">(</span><span class="s2">&quot;Only .pdb and .cif files are allowed.&quot;</span><span class="p">)</span>

        <span class="n">structure_id</span> <span class="o">=</span> <span class="n">input_pdb</span><span class="o">.</span><span class="n">stem</span>
        <span class="k">with</span> <span class="n">input_pdb</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;rt&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">structure</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">structure_id</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">output_fasta</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">output_fasta</span> <span class="o">=</span> <span class="n">input_pdb</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span><span class="s2">&quot;.fa&quot;</span><span class="p">)</span>
            <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Output is not defined, set output to </span><span class="si">{</span><span class="n">output_fasta</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">output_fasta</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_fasta</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">output_fasta</span><span class="o">.</span><span class="n">suffix</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;.fasta&quot;</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Output file must have a &#39;.fa&#39; / &#39;.fasta&#39; suffix&quot;</span><span class="p">)</span>

    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">input_pdb</span><span class="p">,</span> <span class="n">BiopythonStructure</span><span class="p">):</span>
        <span class="n">structure</span> <span class="o">=</span> <span class="n">input_pdb</span>
        <span class="n">structure_id</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="s2">&quot;id&quot;</span><span class="p">,</span> <span class="s2">&quot;structure&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Invalid input type. Expect Path or BiopythonStructure, got </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">input_pdb</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">raise</span> <span class="n">BioatInvalidParameterError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing structure: </span><span class="si">{</span><span class="n">structure_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">records</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="n">BiopythonStructure</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
                <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing chain </span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">residues</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">chain</span><span class="p">)</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">residues</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="n">res_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">residues</span><span class="p">]</span>
                <span class="n">offset</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">res_ids</span><span class="p">)</span>
                <span class="n">length</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">res_ids</span><span class="p">)</span> <span class="o">-</span> <span class="n">offset</span> <span class="o">+</span> <span class="mi">1</span>

                <span class="n">protein</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">length</span>
                <span class="n">dna</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">length</span>
                <span class="n">rna</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">length</span>
                <span class="n">other</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="n">length</span>

                <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">residues</span><span class="p">:</span>
                    <span class="n">idx</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">offset</span>
                    <span class="n">name</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">resname</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

                    <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">AMINO_ACIDS_3CODE</span><span class="p">:</span>
                        <span class="n">protein</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">AMINO_ACIDS_3CODE</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
                    <span class="k">elif</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">AMINO_ACIDS_3CODE_EXTEND</span><span class="p">:</span>
                        <span class="n">protein</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">AMINO_ACIDS_3CODE_EXTEND</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
                    <span class="k">elif</span> <span class="n">name</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;DA&quot;</span><span class="p">,</span> <span class="s2">&quot;DT&quot;</span><span class="p">,</span> <span class="s2">&quot;DG&quot;</span><span class="p">,</span> <span class="s2">&quot;DC&quot;</span><span class="p">):</span>
                        <span class="n">dna</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">name</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="k">elif</span> <span class="n">name</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;U&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">):</span>
                        <span class="n">rna</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">name</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">other</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">]&quot;</span>

                <span class="k">def</span><span class="w"> </span><span class="nf">_append</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">chain</span><span class="p">):</span>
                    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">c</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">):</span>
                        <span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                            <span class="n">SeqRecord</span><span class="p">(</span>
                                <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seq</span><span class="p">)),</span>
                                <span class="nb">id</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">structure_id</span><span class="si">}</span><span class="s2">|Chain_</span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">|</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                                <span class="n">description</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Chain </span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                            <span class="p">)</span>
                        <span class="p">)</span>

                <span class="n">_append</span><span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="s2">&quot;Protein&quot;</span><span class="p">,</span> <span class="n">chain</span><span class="p">)</span>
                <span class="n">_append</span><span class="p">(</span><span class="n">dna</span><span class="p">,</span> <span class="s2">&quot;DNA&quot;</span><span class="p">,</span> <span class="n">chain</span><span class="p">)</span>
                <span class="n">_append</span><span class="p">(</span><span class="n">rna</span><span class="p">,</span> <span class="s2">&quot;RNA&quot;</span><span class="p">,</span> <span class="n">chain</span><span class="p">)</span>
                <span class="n">_append</span><span class="p">(</span><span class="n">other</span><span class="p">,</span> <span class="s2">&quot;UNKNOWN&quot;</span><span class="p">,</span> <span class="n">chain</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">BioatInvalidParameterError</span><span class="p">(</span><span class="s2">&quot;Invalid structure format.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">output_fasta</span><span class="p">:</span>
        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">output_fasta</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;FASTA saved to: </span><span class="si">{</span><span class="n">output_fasta</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">[</span><span class="n">s</span><span class="o">.</span><span class="n">seq</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">records</span><span class="p">]</span> <span class="k">if</span> <span class="n">func_return</span> <span class="k">else</span> <span class="kc">None</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_map_ref_colors</span><span class="p">(</span>
    <span class="n">ref_map_colors</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">,</span>
    <span class="n">ref_map_values</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span>
    <span class="n">ref_map_value_random</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Map reference residue colors based on values.</span>

<span class="sd">    This function maps residue-level scalar values (e.g., importance scores) to colors</span>
<span class="sd">    along a linear gradient defined by `ref_map_colors`. Each residue is identified by</span>
<span class="sd">    a string key in the format &quot;X_123&quot;, where X is the 1-letter amino acid code and 123 is the residue index.</span>

<span class="sd">    Args:</span>
<span class="sd">        ref_map_colors (tuple): A tuple of two color values (e.g., hex or RGB), representing the color gradient endpoints.</span>
<span class="sd">                                For example: (&quot;blue&quot;, &quot;red&quot;) or (&quot;#0000FF&quot;, &quot;#FF0000&quot;).</span>
<span class="sd">        ref_map_values (dict): A dictionary mapping residue keys to float values.</span>
<span class="sd">                               Each key must be in the format &quot;&lt;residue&gt;_&lt;index&gt;&quot;, such as &quot;A_25&quot;.</span>
<span class="sd">                               The value represents a scalar quantity (e.g., attention weight) to be color-mapped.</span>
<span class="sd">        ref_map_value_random (bool): If True, the actual mapped values will be replaced with random numbers in [0, 1],</span>
<span class="sd">                                     ignoring the values in `ref_map_values`. Useful for debugging or visualization testing.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple:</span>
<span class="sd">            - colors (List[str or tuple]): List of color values corresponding to each residue, generated by interpolation.</span>
<span class="sd">            - res_names (List[str]): List of 1-letter residue codes.</span>
<span class="sd">            - res_idxes (List[int]): List of residue indices (integers parsed from the keys).</span>

<span class="sd">    Raises:</span>
<span class="sd">        BioatInvalidParameterError:</span>
<span class="sd">            - If `ref_map_colors` is not a tuple of length 2.</span>
<span class="sd">            - If a residue name in `ref_map_values` is not a valid 1-letter amino acid code.</span>

<span class="sd">    Example:</span>
<span class="sd">        &gt;&gt;&gt; ref_map_colors = (&quot;blue&quot;, &quot;red&quot;)</span>
<span class="sd">        &gt;&gt;&gt; ref_map_values = {&quot;A_25&quot;: 0.1, &quot;V_26&quot;: 0.9}</span>
<span class="sd">        &gt;&gt;&gt; colors, names, idxes = _map_ref_colors(ref_map_colors, ref_map_values, False)</span>
<span class="sd">        &gt;&gt;&gt; print(colors)  # List of interpolated colors between blue and red</span>
<span class="sd">        &gt;&gt;&gt; print(names)  # [&#39;A&#39;, &#39;V&#39;]</span>
<span class="sd">        &gt;&gt;&gt; print(idxes)  # [25, 26]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_map_colors</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">BioatInvalidParameterError</span><span class="p">(</span><span class="s2">&quot;ref_map_colors must be a tuple of length 2&quot;</span><span class="p">)</span>

    <span class="n">color1</span><span class="p">,</span> <span class="n">color2</span> <span class="o">=</span> <span class="n">ref_map_colors</span>

    <span class="n">values</span><span class="p">,</span> <span class="n">res_names</span><span class="p">,</span> <span class="n">res_idxes</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[],</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">ref_map_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">res_name</span><span class="p">,</span> <span class="n">res_idx</span> <span class="o">=</span> <span class="n">k</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">res_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">AMINO_ACIDS_1CODE</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Invalid residue name: </span><span class="si">{</span><span class="n">res_name</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">raise</span> <span class="n">BioatInvalidParameterError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="n">res_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res_name</span><span class="p">)</span>
        <span class="n">res_idxes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">res_idx</span><span class="p">))</span>
        <span class="n">values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>

    <span class="c1"># 映射颜色</span>
    <span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">()</span>
    <span class="n">mapped_values</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">))</span> <span class="k">if</span> <span class="n">ref_map_value_random</span> <span class="k">else</span> <span class="n">values</span>

    <span class="n">colors</span> <span class="o">=</span> <span class="n">map_colors_between_two</span><span class="p">(</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span> <span class="n">mapped_values</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">colors</span><span class="p">,</span> <span class="n">res_names</span><span class="p">,</span> <span class="n">res_idxes</span>


<div class="viewcode-block" id="show_ref_cut">
<a class="viewcode-back" href="../../../bioat.lib.html#bioat.lib.libpdb.show_ref_cut">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">show_ref_cut</span><span class="p">(</span>
    <span class="n">ref_seq</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">Seq</span><span class="p">,</span>
    <span class="n">ref_pdb</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">BiopythonStructure</span><span class="p">,</span>
    <span class="n">cut_seq</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">Seq</span><span class="p">]</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">Seq</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">cut_pdb</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">BiopythonStructure</span><span class="p">]</span>
    <span class="o">|</span> <span class="nb">str</span>
    <span class="o">|</span> <span class="n">Path</span>
    <span class="o">|</span> <span class="n">BiopythonStructure</span>
    <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">cut_labels</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ref_color</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;red&quot;</span><span class="p">,</span>
    <span class="n">ref_map_colors</span><span class="p">:</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ref_map_values</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">cut_color</span><span class="o">=</span><span class="s2">&quot;lightgray&quot;</span><span class="p">,</span>
    <span class="n">gap_color</span><span class="o">=</span><span class="s2">&quot;purple&quot;</span><span class="p">,</span>
    <span class="n">ref_style</span><span class="o">=</span><span class="s2">&quot;cartoon&quot;</span><span class="p">,</span>
    <span class="n">cut_style</span><span class="o">=</span><span class="s2">&quot;cartoon&quot;</span><span class="p">,</span>
    <span class="n">gap_style</span><span class="o">=</span><span class="s2">&quot;cartoon&quot;</span><span class="p">,</span>
    <span class="n">ref_map_value_random</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">output_fig</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">col</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span>
    <span class="n">scale</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
    <span class="n">annotate</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">text_interval</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Visualizes the alignment of sequences and highlights changes in PDB structures using py3Dmol.</span>

<span class="sd">    Args:</span>
<span class="sd">        ref_seq (str or Seq): Amino acid sequence content for the ref protein.</span>
<span class="sd">        ref_pdb (str or BiopythonStructure): Path to the PDB file of the reference structure.</span>
<span class="sd">        cut_seq (str, Seq or None, optional): Amino acid sequence content for the cut protein.</span>
<span class="sd">        cut_pdb (str, BiopythonStructure or None, optional): Path to the PDB file of the cut structure.</span>
<span class="sd">        cut_labels (list[str] or str or None, optional): Label for the cut proteins. If None, the label will be set to &quot;cut&quot;.</span>
<span class="sd">        ref_color (str, optional): Color for reference residues.</span>
<span class="sd">        ref_map_colors (tuple[str, str] or None, optional): ref_map_colors will be used as color bar from ref_map_colors[0] to ref_map_colors[1]. If None, do not apply color mapping. Defaults to None.</span>
<span class="sd">        ref_map_values (dict or None, optional): A dictionary of values for the ref color map, it will be normalized to the range of [0 - 1]. If None, all residues will be colored with the same color. e.g. ref_map_values = {&#39;V_0&#39;: 0.4177215189873418, &#39;S_1&#39;: 0.8185654008438819, &#39;K_2&#39;: 0.9915611814345991, &#39;G_3&#39;: 0.42616033755274263, ...}</span>
<span class="sd">        cut_color (str, optional): Color for cut residues.</span>
<span class="sd">        gap_color (str, optional): Color for gaps or removed residues.</span>
<span class="sd">        ref_style (str, optional): &quot;stick&quot;, &quot;sphere&quot;, &quot;cartoon&quot;, or &quot;line&quot;</span>
<span class="sd">        cut_style (str, optional): &quot;stick&quot;, &quot;sphere&quot;, &quot;cartoon&quot;, or &quot;line&quot;</span>
<span class="sd">        gap_style (str, optional): &quot;stick&quot;, &quot;sphere&quot;, &quot;cartoon&quot;, or &quot;line&quot;</span>
<span class="sd">        ref_map_value_random (bool, optional): If True, ref_map_values will be randomly generated. Defaults to False.</span>
<span class="sd">        output_fig (str or Path or None, optional): Output figure file path. If None, the figure will not be saved in html format. Defaults to None.</span>
<span class="sd">        col (int, optional): Number of columns for the visualization. Defaults to 3.</span>
<span class="sd">        scale (float, optional): Scale factor for the visualization. Defaults to 1.0.</span>
<span class="sd">        annotate (bool, optional): Whether to annotate the visualization with labels. Defaults to True.</span>
<span class="sd">        text_interval (int, optional): The interval between text annotations. Defaults to 5.</span>
<span class="sd">        log_level (str, optional): Log level. Defaults to &quot;WARNING&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">set_names</span><span class="p">(</span><span class="n">func_name</span><span class="o">=</span><span class="s2">&quot;show_ref_cut&quot;</span><span class="p">)</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">set_level</span><span class="p">(</span><span class="n">log_level</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cut_seq</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="s2">&quot;,&quot;</span> <span class="ow">in</span> <span class="n">cut_seq</span><span class="p">:</span>
        <span class="n">cut_seq</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">Seq</span><span class="p">],</span> <span class="n">cut_seq</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">))</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cut_pdb</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span> <span class="s2">&quot;,&quot;</span> <span class="ow">in</span> <span class="n">cut_pdb</span><span class="p">:</span>
        <span class="n">cut_pdb</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="nb">list</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">BiopythonStructure</span><span class="p">],</span> <span class="n">cut_pdb</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">))</span>
    <span class="n">ref_seq</span><span class="p">,</span> <span class="n">ref_label</span> <span class="o">=</span> <span class="n">_load_seq</span><span class="p">(</span><span class="n">ref_seq</span><span class="p">,</span> <span class="s2">&quot;ref&quot;</span><span class="p">)</span>
    <span class="n">ref_pdb</span><span class="p">,</span> <span class="n">ref_label</span> <span class="o">=</span> <span class="n">load_structure</span><span class="p">(</span><span class="n">ref_pdb</span><span class="p">,</span> <span class="n">ref_label</span><span class="p">)</span>

    <span class="n">aligner</span> <span class="o">=</span> <span class="n">instantiate_pairwise_aligner</span><span class="p">(</span>
        <span class="n">scoring_match</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
        <span class="n">penalty_mismatch</span><span class="o">=-</span><span class="mi">100</span><span class="p">,</span>
        <span class="n">penalty_gap_open</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">penalty_gap_extension</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">penalty_query_left_gap_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">penalty_query_right_gap_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">log_level</span><span class="o">=</span><span class="n">log_level</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">cut_seq</span><span class="p">:</span>  <span class="c1"># not None</span>
        <span class="n">cut_seqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">cut_seq</span><span class="p">]</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cut_seq</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="k">else</span> <span class="n">cut_seq</span>
        <span class="n">cut_pdbs</span> <span class="o">=</span> <span class="p">[</span><span class="n">cut_pdb</span><span class="p">]</span> <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cut_pdb</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="k">else</span> <span class="n">cut_pdb</span>

        <span class="n">ref_seqs_fix</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">ref_pdbs_fix</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">ref_labels</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">cut_seqs_fix</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">cut_pdbs_fix</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">cut_labels_fix</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">gap_indices</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">rmsds</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">alignment_dicts</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">_cut_seq</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cut_seqs</span><span class="p">):</span>
            <span class="n">ref_seqs_fix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ref_seq</span><span class="p">)</span>
            <span class="n">ref_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ref_label</span><span class="p">)</span>
            <span class="c1"># 执行Sequence Alignment并得到gaps的位置</span>
            <span class="n">_cut_seq</span><span class="p">,</span> <span class="n">cut_label</span> <span class="o">=</span> <span class="n">_load_seq</span><span class="p">(</span><span class="n">_cut_seq</span><span class="p">,</span> <span class="s2">&quot;cut&quot;</span><span class="p">)</span>
            <span class="n">cut_seqs_fix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_cut_seq</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">cut_labels</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cut_labels</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
                    <span class="n">cut_labels_fix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cut_labels</span><span class="p">[</span><span class="n">idx</span><span class="p">])</span>
                <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cut_labels</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                    <span class="n">cut_labels_fix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cut_labels</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cut_labels_fix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cut_label</span><span class="p">)</span>

            <span class="c1"># 获得gaps的位置</span>
            <span class="c1"># Perform alignment</span>
            <span class="n">alignment</span> <span class="o">=</span> <span class="n">get_best_alignment</span><span class="p">(</span>
                <span class="n">seq_a</span><span class="o">=</span><span class="n">ref_seq</span><span class="p">,</span>
                <span class="n">seq_b</span><span class="o">=</span><span class="n">_cut_seq</span><span class="p">,</span>
                <span class="n">aligner</span><span class="o">=</span><span class="n">aligner</span><span class="p">,</span>
                <span class="n">consider_strand</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">alignment_dict</span> <span class="o">=</span> <span class="n">get_aligned_seq</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">letn_match</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">alignment_dicts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment_dict</span><span class="p">)</span>
            <span class="n">gap_indices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>  <span class="c1"># 不指定 cut pdb 时必须在这里计算而不是get_cut2ref_aln_info中</span>
                <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">alignment_dict</span><span class="p">[</span><span class="s2">&quot;target_seq&quot;</span><span class="p">])</span> <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">],</span>
            <span class="p">)</span>

            <span class="n">cut_pdb</span> <span class="o">=</span> <span class="n">cut_pdbs</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">cut_pdb</span><span class="p">:</span>
                <span class="c1"># 执行PDB Alignment并得对齐后的PDB文件和RMSD</span>
                <span class="n">res</span> <span class="o">=</span> <span class="n">get_cut2ref_aln_info</span><span class="p">(</span>
                    <span class="n">ref_pdb</span><span class="p">,</span>
                    <span class="n">cut_pdb</span><span class="p">,</span>
                    <span class="n">cal_rmsd</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">cal_tmscore</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                    <span class="n">label1</span><span class="o">=</span><span class="n">ref_label</span><span class="p">,</span>
                    <span class="n">label2</span><span class="o">=</span><span class="n">cut_label</span><span class="p">,</span>
                    <span class="n">log_level</span><span class="o">=</span><span class="n">log_level</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">ref_pdbs_fix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="n">ref_label</span><span class="p">])</span>
                <span class="n">cut_pdbs_fix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="n">cut_label</span><span class="p">])</span>
                <span class="n">rmsds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s2">&quot;RMSD&quot;</span><span class="p">])</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="n">cut_label</span><span class="p">],</span> <span class="n">BiopythonStructure</span><span class="p">):</span>
                    <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;Invalid input format. cut_pdb must be (one or list of) BiopythonStructure&quot;</span>
                    <span class="k">raise</span> <span class="n">BioatInvalidParameterError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">ref_pdbs_fix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ref_pdb</span><span class="p">)</span>
                <span class="n">cut_pdbs_fix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">rmsds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Cut sequence is not provided, skip alignment and annotation for gaps, and just show ref visualization.&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">ref_seqs_fix</span> <span class="o">=</span> <span class="p">[</span><span class="n">ref_seq</span><span class="p">]</span>
        <span class="n">ref_pdbs_fix</span> <span class="o">=</span> <span class="p">[</span><span class="n">ref_pdb</span><span class="p">]</span>
        <span class="n">ref_labels</span> <span class="o">=</span> <span class="p">[</span><span class="n">ref_label</span><span class="p">]</span>
        <span class="n">cut_seqs_fix</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span>
        <span class="n">cut_pdbs_fix</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span>
        <span class="n">cut_labels_fix</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span>
        <span class="n">gap_indices</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span>
        <span class="n">rmsds</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span>
        <span class="n">alignment_dicts</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span>

    <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_seqs_fix</span><span class="p">)</span>
    <span class="n">col</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">col</span><span class="p">)</span>
    <span class="n">row</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">n</span> <span class="o">/</span> <span class="n">col</span><span class="p">)</span>

    <span class="n">view</span> <span class="o">=</span> <span class="n">py3Dmol</span><span class="o">.</span><span class="n">view</span><span class="p">(</span>
        <span class="n">width</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mi">500</span> <span class="o">*</span> <span class="n">scale</span> <span class="o">*</span> <span class="n">col</span><span class="p">),</span>
        <span class="n">height</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="mi">500</span> <span class="o">*</span> <span class="n">scale</span> <span class="o">*</span> <span class="n">row</span><span class="p">),</span>
        <span class="n">viewergrid</span><span class="o">=</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">col</span><span class="p">)</span> <span class="k">if</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">linked</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="c1"># 是否同步网格间的分子运动</span>
    <span class="p">)</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;viewergrid: (</span><span class="si">{</span><span class="n">row</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="n">viewers</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span> <span class="o">//</span> <span class="n">col</span><span class="p">,</span> <span class="n">i</span> <span class="o">%</span> <span class="n">col</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)]</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;viewers = </span><span class="si">{</span><span class="n">viewers</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">viewer</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">viewers</span><span class="p">):</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;viewer = </span><span class="si">{</span><span class="n">viewer</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">_add_one_submodel</span><span class="p">(</span>
            <span class="n">view</span><span class="o">=</span><span class="n">view</span><span class="p">,</span>
            <span class="n">ref_pdb</span><span class="o">=</span><span class="n">ref_pdbs_fix</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
            <span class="n">ref_label</span><span class="o">=</span><span class="n">ref_labels</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
            <span class="n">cut_pdb</span><span class="o">=</span><span class="n">cut_pdbs_fix</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
            <span class="n">cut_label</span><span class="o">=</span><span class="n">cut_labels_fix</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
            <span class="n">ref_color</span><span class="o">=</span><span class="n">ref_color</span><span class="p">,</span>
            <span class="n">ref_map_colors</span><span class="o">=</span><span class="n">ref_map_colors</span><span class="p">,</span>
            <span class="n">ref_map_values</span><span class="o">=</span><span class="n">ref_map_values</span><span class="p">,</span>
            <span class="n">ref_map_value_random</span><span class="o">=</span><span class="n">ref_map_value_random</span><span class="p">,</span>
            <span class="n">cut_color</span><span class="o">=</span><span class="n">cut_color</span><span class="p">,</span>
            <span class="n">gap_color</span><span class="o">=</span><span class="n">gap_color</span><span class="p">,</span>
            <span class="n">gap_indices</span><span class="o">=</span><span class="n">gap_indices</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
            <span class="n">ref_style</span><span class="o">=</span><span class="n">ref_style</span><span class="p">,</span>
            <span class="n">cut_style</span><span class="o">=</span><span class="n">cut_style</span><span class="p">,</span>
            <span class="n">gap_style</span><span class="o">=</span><span class="n">gap_style</span><span class="p">,</span>
            <span class="n">alignment_dict</span><span class="o">=</span><span class="n">alignment_dicts</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
            <span class="n">rmsd</span><span class="o">=</span><span class="n">rmsds</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
            <span class="n">annotate</span><span class="o">=</span><span class="n">annotate</span><span class="p">,</span>
            <span class="n">text_interval</span><span class="o">=</span><span class="n">text_interval</span><span class="p">,</span>
            <span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">output_fig</span><span class="p">:</span>
        <span class="n">output_fig</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">output_fig</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">output_fig</span><span class="o">.</span><span class="n">suffix</span> <span class="o">!=</span> <span class="s2">&quot;.html&quot;</span><span class="p">:</span>
            <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Output file must have a &#39;.html&#39; extension, but got </span><span class="si">{</span><span class="n">output_fig</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">output_fig</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">view</span><span class="o">.</span><span class="n">_make_html</span><span class="p">())</span>  <span class="c1"># noqa: SLF001</span>
            <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Visualization saved to: </span><span class="si">{</span><span class="n">output_fig</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">view</span><span class="o">.</span><span class="n">show</span><span class="p">()</span></div>



<span class="k">def</span><span class="w"> </span><span class="nf">_add_one_submodel</span><span class="p">(</span>
    <span class="n">view</span><span class="p">,</span>
    <span class="n">ref_pdb</span><span class="p">,</span>
    <span class="n">ref_label</span><span class="p">,</span>
    <span class="n">cut_pdb</span><span class="p">,</span>
    <span class="n">cut_label</span><span class="p">,</span>
    <span class="n">ref_color</span><span class="p">,</span>
    <span class="n">ref_map_colors</span><span class="p">,</span>
    <span class="n">ref_map_values</span><span class="p">,</span>
    <span class="n">ref_map_value_random</span><span class="p">,</span>
    <span class="n">cut_color</span><span class="p">,</span>
    <span class="n">gap_color</span><span class="p">,</span>
    <span class="n">gap_indices</span><span class="p">,</span>
    <span class="n">ref_style</span><span class="p">,</span>
    <span class="n">cut_style</span><span class="p">,</span>
    <span class="n">gap_style</span><span class="p">,</span>
    <span class="n">alignment_dict</span><span class="p">,</span>
    <span class="n">rmsd</span><span class="p">,</span>
    <span class="n">annotate</span><span class="p">,</span>
    <span class="n">text_interval</span><span class="p">,</span>
    <span class="n">viewer</span><span class="p">,</span>
<span class="p">):</span>
    <span class="c1"># for annotation</span>
    <span class="n">label_xpos</span> <span class="o">=</span> <span class="o">-</span><span class="mi">70</span>
    <span class="n">label_ypos</span> <span class="o">=</span> <span class="mi">75</span>
    <span class="n">label_zpos</span> <span class="o">=</span> <span class="o">-</span><span class="mi">30</span>

    <span class="k">if</span> <span class="n">ref_map_colors</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">colors</span><span class="p">,</span> <span class="n">res_names</span><span class="p">,</span> <span class="n">res_idxes</span> <span class="o">=</span> <span class="n">_map_ref_colors</span><span class="p">(</span>
            <span class="n">ref_map_colors</span><span class="p">,</span>
            <span class="n">ref_map_values</span><span class="p">,</span>
            <span class="n">ref_map_value_random</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">colors</span><span class="p">,</span> <span class="n">res_names</span><span class="p">,</span> <span class="n">res_idxes</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>
    <span class="n">color1</span><span class="p">,</span> <span class="n">color2</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">ref_map_colors</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">color1</span><span class="p">,</span> <span class="n">color2</span> <span class="o">=</span> <span class="n">ref_map_colors</span>
    <span class="k">if</span> <span class="n">ref_map_colors</span> <span class="ow">and</span> <span class="n">ref_map_values</span><span class="p">:</span>
        <span class="n">values</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">res_names</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">res_idxes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">ref_map_values</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">res_name</span><span class="p">,</span> <span class="n">res_idx</span> <span class="o">=</span> <span class="n">k</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">res_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">AMINO_ACIDS_1CODE</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Invalid residue name: </span><span class="si">{</span><span class="n">res_name</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">raise</span> <span class="n">BioatInvalidParameterError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="n">res_idx</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">res_idx</span><span class="p">)</span>
            <span class="n">values</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
            <span class="n">res_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res_name</span><span class="p">)</span>
            <span class="n">res_idxes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res_idx</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">ref_map_value_random</span><span class="p">:</span>
            <span class="n">colors</span> <span class="o">=</span> <span class="n">map_colors_between_two</span><span class="p">(</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span> <span class="n">values</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">rng</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">default_rng</span><span class="p">()</span>
            <span class="n">mapped_values</span> <span class="o">=</span> <span class="n">rng</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">))</span>
            <span class="n">colors</span> <span class="o">=</span> <span class="n">map_colors_between_two</span><span class="p">(</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span> <span class="n">mapped_values</span><span class="p">)</span>

    <span class="n">view</span><span class="o">.</span><span class="n">setViewStyle</span><span class="p">({</span><span class="s2">&quot;style&quot;</span><span class="p">:</span> <span class="s2">&quot;outline&quot;</span><span class="p">,</span> <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="s2">&quot;width&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">})</span>

    <span class="n">view</span><span class="o">.</span><span class="n">addModel</span><span class="p">(</span>
        <span class="n">structure2string</span><span class="p">(</span><span class="n">ref_pdb</span><span class="p">),</span>
        <span class="s2">&quot;pdb&quot;</span><span class="p">,</span>
        <span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">view</span><span class="o">.</span><span class="n">setStyle</span><span class="p">(</span>
        <span class="p">{</span><span class="s2">&quot;model&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">},</span>
        <span class="p">{</span>
            <span class="n">ref_style</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;arrows&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                <span class="s2">&quot;tubes&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                <span class="s2">&quot;style&quot;</span><span class="p">:</span> <span class="s2">&quot;oval&quot;</span><span class="p">,</span>
                <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="n">ref_color</span><span class="p">,</span>
            <span class="p">},</span>
        <span class="p">},</span>
        <span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">label_ypos</span> <span class="o">-=</span> <span class="n">text_interval</span>
    <span class="n">view</span><span class="o">.</span><span class="n">addLabel</span><span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ref_label</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">ref_color</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="p">{</span>
            <span class="s2">&quot;fontColor&quot;</span><span class="p">:</span> <span class="n">ref_color</span><span class="p">,</span>
            <span class="s2">&quot;fontSize&quot;</span><span class="p">:</span> <span class="mi">14</span><span class="p">,</span>
            <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="n">label_xpos</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">:</span> <span class="n">label_ypos</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">:</span> <span class="n">label_zpos</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># 标记ref 值的颜色</span>
    <span class="k">if</span> <span class="p">(</span>
        <span class="n">ref_map_colors</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="ow">and</span> <span class="n">colors</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="ow">and</span> <span class="n">res_names</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="ow">and</span> <span class="n">res_idxes</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
    <span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">colors</span><span class="p">)):</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="n">res_idxes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">res_names</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">view</span><span class="o">.</span><span class="n">addStyle</span><span class="p">(</span>
                <span class="p">{</span><span class="s2">&quot;model&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;resi&quot;</span><span class="p">:</span> <span class="n">idx</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;resn&quot;</span><span class="p">:</span> <span class="n">AMINO_ACIDS_1CODE</span><span class="p">[</span><span class="n">res</span><span class="p">]},</span>
                <span class="p">{</span><span class="n">ref_style</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="n">color</span><span class="p">,</span> <span class="s2">&quot;radius&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">}},</span>
                <span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">annotate</span><span class="p">:</span>
            <span class="n">label_ypos</span> <span class="o">-=</span> <span class="n">text_interval</span>
            <span class="n">view</span><span class="o">.</span><span class="n">addLabel</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ref_label</span><span class="si">}</span><span class="s2"> values: </span><span class="si">{</span><span class="n">ref_color</span><span class="si">}</span><span class="s2"> (0) to </span><span class="si">{</span><span class="n">ref_color</span><span class="si">}</span><span class="s2"> (1)&quot;</span><span class="p">,</span>
                <span class="p">{</span>
                    <span class="s2">&quot;fontColor&quot;</span><span class="p">:</span> <span class="n">ref_color</span><span class="p">,</span>
                    <span class="s2">&quot;fontSize&quot;</span><span class="p">:</span> <span class="mi">14</span><span class="p">,</span>
                    <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="n">label_xpos</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">:</span> <span class="n">label_ypos</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">:</span> <span class="n">label_zpos</span><span class="p">},</span>
                <span class="p">},</span>
                <span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">,</span>
            <span class="p">)</span>
    <span class="c1"># 标记gaps的颜色</span>
    <span class="k">if</span> <span class="n">gap_indices</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">res</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">alignment_dict</span><span class="p">[</span><span class="s2">&quot;reference_seq&quot;</span><span class="p">]):</span>
            <span class="c1"># 标记 gap位置的颜色</span>
            <span class="k">if</span> <span class="n">idx</span> <span class="ow">in</span> <span class="n">gap_indices</span><span class="p">:</span>
                <span class="n">view</span><span class="o">.</span><span class="n">addStyle</span><span class="p">(</span>
                    <span class="p">{</span><span class="s2">&quot;model&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;resi&quot;</span><span class="p">:</span> <span class="n">idx</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;resn&quot;</span><span class="p">:</span> <span class="n">AMINO_ACIDS_1CODE</span><span class="p">[</span><span class="n">res</span><span class="p">]},</span>
                    <span class="p">{</span><span class="n">gap_style</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="n">gap_color</span><span class="p">,</span> <span class="s2">&quot;radius&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">}},</span>
                    <span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">,</span>
                <span class="p">)</span>
        <span class="k">if</span> <span class="n">annotate</span><span class="p">:</span>
            <span class="n">label_ypos</span> <span class="o">-=</span> <span class="n">text_interval</span>
            <span class="n">view</span><span class="o">.</span><span class="n">addLabel</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;gaps: </span><span class="si">{</span><span class="n">gap_color</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="p">{</span>
                    <span class="s2">&quot;fontColor&quot;</span><span class="p">:</span> <span class="n">gap_color</span><span class="p">,</span>
                    <span class="s2">&quot;fontSize&quot;</span><span class="p">:</span> <span class="mi">14</span><span class="p">,</span>
                    <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="n">label_xpos</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">:</span> <span class="n">label_ypos</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">:</span> <span class="n">label_zpos</span><span class="p">},</span>
                <span class="p">},</span>
                <span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">,</span>
            <span class="p">)</span>
    <span class="k">if</span> <span class="n">cut_pdb</span> <span class="ow">and</span> <span class="n">rmsd</span><span class="p">:</span>
        <span class="n">view</span><span class="o">.</span><span class="n">addModel</span><span class="p">(</span><span class="n">structure2string</span><span class="p">(</span><span class="n">cut_pdb</span><span class="p">),</span> <span class="s2">&quot;pdb&quot;</span><span class="p">,</span> <span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">)</span>
        <span class="n">view</span><span class="o">.</span><span class="n">setStyle</span><span class="p">(</span>
            <span class="p">{</span><span class="s2">&quot;model&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
            <span class="p">{</span>
                <span class="n">cut_style</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;arrows&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                    <span class="s2">&quot;tubes&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
                    <span class="s2">&quot;style&quot;</span><span class="p">:</span> <span class="s2">&quot;oval&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="n">cut_color</span><span class="p">,</span>
                <span class="p">},</span>
            <span class="p">},</span>
            <span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">annotate</span><span class="p">:</span>
            <span class="n">label_ypos</span> <span class="o">-=</span> <span class="n">text_interval</span>
            <span class="n">view</span><span class="o">.</span><span class="n">addLabel</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cut_label</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">cut_color</span><span class="si">}</span><span class="se">\t</span><span class="s2">RMSD: </span><span class="si">{</span><span class="n">rmsd</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="p">{</span>
                    <span class="s2">&quot;fontColor&quot;</span><span class="p">:</span> <span class="n">cut_color</span><span class="p">,</span>
                    <span class="s2">&quot;fontSize&quot;</span><span class="p">:</span> <span class="mi">14</span><span class="p">,</span>
                    <span class="s2">&quot;position&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;x&quot;</span><span class="p">:</span> <span class="n">label_xpos</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">:</span> <span class="n">label_ypos</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">:</span> <span class="n">label_zpos</span><span class="p">},</span>
                <span class="p">},</span>
                <span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="c1"># 聚焦显示</span>
    <span class="c1"># Finalize the view</span>
    <span class="n">view</span><span class="o">.</span><span class="n">zoomTo</span><span class="p">(</span><span class="n">viewer</span><span class="o">=</span><span class="n">viewer</span><span class="p">)</span>
    <span class="c1"># view.render(viewer=viewer)</span>


<span class="c1"># 对齐函数</span>
<div class="viewcode-block" id="get_cut2ref_aln_info">
<a class="viewcode-back" href="../../../bioat.lib.html#bioat.lib.libpdb.get_cut2ref_aln_info">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_cut2ref_aln_info</span><span class="p">(</span>
    <span class="n">ref</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">BiopythonStructure</span><span class="p">,</span>
    <span class="n">cut</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">|</span> <span class="n">BiopythonStructure</span><span class="p">,</span>
    <span class="n">cal_rmsd</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">cal_tmscore</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">label1</span><span class="o">=</span><span class="s2">&quot;ref&quot;</span><span class="p">,</span>
    <span class="n">label2</span><span class="o">=</span><span class="s2">&quot;cut&quot;</span><span class="p">,</span>
    <span class="n">usalign_bin</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span> <span class="o">=</span> <span class="s2">&quot;usalign&quot;</span><span class="p">,</span>
    <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Align cutted pdb to ref pdb using the CA atoms.</span>

<span class="sd">    Aligns a truncated protein structure (cut) to its full-length reference structure (ref)</span>
<span class="sd">    using Ca atoms and Biopython&#39;s Superimposer.</span>

<span class="sd">    This function:</span>
<span class="sd">    - Extracts all Ca atoms from `ref` and `cut`</span>
<span class="sd">    - Removes atoms from `ref` at the indices listed in `gap_indices`</span>
<span class="sd">    - Aligns the remaining atoms from `cut` to the corresponding positions in `ref`</span>
<span class="sd">    - Modifies the `cut` structure in-place to match the aligned orientation</span>
<span class="sd">    - Returns both structures and the RMSD value of the alignment</span>

<span class="sd">    It assumes:</span>
<span class="sd">    - One-to-one correspondence between residues after gap removal</span>
<span class="sd">    - Structures are predicted by AlphaFold2 / ESMFold (no missing atoms)</span>

<span class="sd">    Args:</span>
<span class="sd">        ref (str or Bio.PDB.Structure.Structure): Reference structure path or loaded Structure.</span>
<span class="sd">        cut (str or Bio.PDB.Structure.Structure): Truncated structure path or loaded Structure.</span>
<span class="sd">        cal_rmsd (bool, optional): Whether to calculate RMSD. Default is True.</span>
<span class="sd">        cal_tmscore (bool, optional): Whether to calculate TM-score using USalign. Default is False.</span>
<span class="sd">        label1 (str, optional): Name for the reference structure. Default is &quot;ref&quot;.</span>
<span class="sd">        label2 (str, optional): Name for the cut structure. Default is &quot;cut&quot;.</span>
<span class="sd">        usalign_bin (str or Path, optional): Path to the USalign binary for TM-score calculation. Default is &quot;usalign&quot;.</span>
<span class="sd">        log_level (str, optional): Logging level. Default is &quot;WARNING&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: {</span>
<span class="sd">                &quot;{label1}&quot;: aln label1 structure,  # if cal_rmsd is True, unaltered label1 structure</span>
<span class="sd">                &quot;{label2}}&quot;: fixed label2 structure,  # if cal_rmsd is True, fix label2 coords in-place</span>
<span class="sd">                &quot;RMSD&quot;: 0.123  # if cal_rmsd is True, the RMSD value between label1 and label2</span>
<span class="sd">                f&quot;{label1}_seq&quot;: ref_seq,  # if cal_rmsd is True, the sequence of label1 structure</span>
<span class="sd">                f&quot;{label2}_seq&quot;: cut_seq,  # if cal_rmsd is True, the sequence of label2 structure</span>
<span class="sd">                &quot;alignment_dict&quot;: alignment_dict,  # if cal_rmsd is True, the alignment dict of label1 and label2</span>
<span class="sd">                &quot;gap_indices&quot;: gap_indices,  # if cal_rmsd is True, the indices of gaps in label1 structure</span>
<span class="sd">                &quot;TM-score:mean&quot;: 0.623,  # if cal_tmscore is True, the mean TM-score value</span>
<span class="sd">                &quot;TM-score:TM1&quot;: 0.456,  # if cal_tmscore is True, use label1 as ref &lt;L_N&gt; in calculation</span>
<span class="sd">                &quot;TM-score:TM2&quot;: 0.789,  # if cal_tmscore is True, use label2 as ref &lt;L_N&gt; in calculation</span>
<span class="sd">                ...</span>
<span class="sd">            }</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># 读取 PDB 文件</span>
    <span class="n">ref</span><span class="p">,</span> <span class="n">cut</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">load_structure</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">label1</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
        <span class="n">load_structure</span><span class="p">(</span><span class="n">cut</span><span class="p">,</span> <span class="n">label2</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">aln_info</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">if</span> <span class="n">cal_rmsd</span><span class="p">:</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Calculating RMSD between </span><span class="si">{</span><span class="n">label1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">label2</span><span class="si">}</span><span class="s2"> using CA atoms.&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">ref_seq</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="n">Seq</span><span class="p">,</span> <span class="n">pdb2fasta</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">func_return</span><span class="o">=</span><span class="kc">True</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>  <span class="c1"># type: ignore</span>
        <span class="n">cut_seq</span> <span class="o">=</span> <span class="n">cast</span><span class="p">(</span><span class="n">Seq</span><span class="p">,</span> <span class="n">pdb2fasta</span><span class="p">(</span><span class="n">cut</span><span class="p">,</span> <span class="n">func_return</span><span class="o">=</span><span class="kc">True</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>  <span class="c1"># type: ignore</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fetch seq info from structure: </span><span class="si">{</span><span class="n">ref</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="n">ref_seq</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;fetch seq info from structure:  </span><span class="si">{</span><span class="n">cut</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> -&gt; </span><span class="si">{</span><span class="n">cut_seq</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">aligner</span> <span class="o">=</span> <span class="n">instantiate_pairwise_aligner</span><span class="p">(</span>
            <span class="n">scoring_match</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
            <span class="n">penalty_mismatch</span><span class="o">=-</span><span class="mi">100</span><span class="p">,</span>
            <span class="n">penalty_gap_open</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">penalty_gap_extension</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">penalty_query_left_gap_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">penalty_query_right_gap_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">log_level</span><span class="o">=</span><span class="n">log_level</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="c1"># 获得gaps的位置</span>
        <span class="c1"># Perform alignment</span>
        <span class="n">alignment</span> <span class="o">=</span> <span class="n">get_best_alignment</span><span class="p">(</span>
            <span class="n">seq_a</span><span class="o">=</span><span class="n">ref_seq</span><span class="p">,</span>
            <span class="n">seq_b</span><span class="o">=</span><span class="n">cut_seq</span><span class="p">,</span>
            <span class="n">aligner</span><span class="o">=</span><span class="n">aligner</span><span class="p">,</span>
            <span class="n">consider_strand</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">alignment_dict</span> <span class="o">=</span> <span class="n">get_aligned_seq</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">letn_match</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">gap_indices</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">alignment_dict</span><span class="p">[</span><span class="s2">&quot;target_seq&quot;</span><span class="p">])</span> <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span>
        <span class="p">]</span>
        <span class="c1"># 选择对齐的原子 (e.g., CA)</span>
        <span class="n">atoms1</span> <span class="o">=</span> <span class="p">[</span><span class="n">atom</span> <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">ref</span><span class="o">.</span><span class="n">get_atoms</span><span class="p">()</span> <span class="k">if</span> <span class="n">atom</span><span class="o">.</span><span class="n">get_name</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span><span class="p">]</span>
        <span class="n">atoms1</span> <span class="o">=</span> <span class="p">[</span><span class="n">atoms1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">atoms1</span><span class="p">))</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gap_indices</span><span class="p">]</span>
        <span class="n">atoms2</span> <span class="o">=</span> <span class="p">[</span><span class="n">atom</span> <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">cut</span><span class="o">.</span><span class="n">get_atoms</span><span class="p">()</span> <span class="k">if</span> <span class="n">atom</span><span class="o">.</span><span class="n">get_name</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;CA&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">atoms1</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">atoms2</span><span class="p">):</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;The number of CA atoms in </span><span class="si">{</span><span class="n">label1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">label2</span><span class="si">}</span><span class="s2"> are not equal. &quot;</span>
                <span class="s2">&quot;Use ref and cut pdbs predicted by AlphaFold2/ESMFold or other tools &quot;</span>
                <span class="s2">&quot;but not pdb downloaded from &lt;PDB database&gt; to ensure the CA atoms are aligned.&quot;</span>
            <span class="p">)</span>
            <span class="k">raise</span> <span class="n">BioatInvalidParameterError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="c1"># 使用 Biopython 的 Superimposer 进行对齐</span>
        <span class="n">super_imposer</span> <span class="o">=</span> <span class="n">Superimposer</span><span class="p">()</span>
        <span class="n">super_imposer</span><span class="o">.</span><span class="n">set_atoms</span><span class="p">(</span><span class="n">atoms1</span><span class="p">,</span> <span class="n">atoms2</span><span class="p">)</span>
        <span class="n">super_imposer</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">cut</span><span class="o">.</span><span class="n">get_atoms</span><span class="p">())</span>  <span class="c1"># 修改 pdb 的坐标</span>
        <span class="n">rmsd</span> <span class="o">=</span> <span class="n">super_imposer</span><span class="o">.</span><span class="n">rms</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;RMSD (strict, remove gaps in </span><span class="si">{</span><span class="n">label1</span><span class="si">}</span><span class="s2">) between </span><span class="si">{</span><span class="n">label1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">label2</span><span class="si">}</span><span class="s2">: </span><span class="si">{</span><span class="n">rmsd</span><span class="si">:</span><span class="s2">.3f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">aln_info</span><span class="p">[</span><span class="n">label1</span><span class="p">]</span> <span class="o">=</span> <span class="n">ref</span>
        <span class="n">aln_info</span><span class="p">[</span><span class="n">label2</span><span class="p">]</span> <span class="o">=</span> <span class="n">cut</span>  <span class="c1"># 坐标已对齐</span>
        <span class="n">aln_info</span><span class="p">[</span><span class="s2">&quot;RMSD&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">rmsd</span>  <span class="c1"># it should be close to the result of pymol cmd.align(&quot;cut and name CA&quot;, &quot;ref and name CA&quot;, cutoff=10.0, cycles=0)[0]</span>
        <span class="p">)</span>
        <span class="n">aln_info</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label1</span><span class="si">}</span><span class="s2">_seq&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">ref_seq</span><span class="p">)</span>
        <span class="n">aln_info</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">label2</span><span class="si">}</span><span class="s2">_seq&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">cut_seq</span><span class="p">)</span>
        <span class="n">aln_info</span><span class="p">[</span><span class="s2">&quot;alignment_dict&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">alignment_dict</span>
        <span class="n">aln_info</span><span class="p">[</span><span class="s2">&quot;gap_indices&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gap_indices</span>

    <span class="k">if</span> <span class="n">cal_tmscore</span><span class="p">:</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Calculating TM-score between </span><span class="si">{</span><span class="n">label1</span><span class="si">}</span><span class="s2"> and </span><span class="si">{</span><span class="n">label2</span><span class="si">}</span><span class="s2"> using CA atoms.&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># 判断 usalign_bin 是否可执行</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="nb">isinstance</span><span class="p">(</span><span class="n">usalign_bin</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span>
            <span class="ow">and</span> <span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">usalign_bin</span>
            <span class="ow">and</span> <span class="n">shutil</span><span class="o">.</span><span class="n">which</span><span class="p">(</span><span class="n">usalign_bin</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">None</span>
        <span class="p">)</span> <span class="ow">or</span> <span class="p">(</span>
            <span class="nb">isinstance</span><span class="p">(</span><span class="n">usalign_bin</span><span class="p">,</span> <span class="nb">str</span> <span class="o">|</span> <span class="n">Path</span><span class="p">)</span>
            <span class="ow">and</span> <span class="s2">&quot;/&quot;</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">usalign_bin</span><span class="p">)</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span>
                <span class="n">Path</span><span class="p">(</span><span class="n">usalign_bin</span><span class="p">)</span><span class="o">.</span><span class="n">expanduser</span><span class="p">()</span><span class="o">.</span><span class="n">resolve</span><span class="p">()</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span>
                <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">access</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="n">usalign_bin</span><span class="p">)</span><span class="o">.</span><span class="n">expanduser</span><span class="p">()</span><span class="o">.</span><span class="n">resolve</span><span class="p">(),</span> <span class="n">os</span><span class="o">.</span><span class="n">X_OK</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">):</span>
            <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;usalign binary not found in var &#39;PATH&#39; or not executable: </span><span class="si">{</span><span class="n">usalign_bin</span><span class="si">}</span><span class="s2">, please set the --usalign_bin &lt;path/to/usalign&gt; param or use conda install -c conda-forge usalign to install it!&quot;</span>
            <span class="p">)</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;usalign binary not found or not executable: </span><span class="si">{</span><span class="n">usalign_bin</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">raise</span> <span class="n">BioatFileNotFoundError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">save_structure_to_tempfile</span><span class="p">(</span><span class="n">structure</span><span class="p">):</span>
            <span class="n">io</span> <span class="o">=</span> <span class="n">PDBIO</span><span class="p">()</span>
            <span class="n">io</span><span class="o">.</span><span class="n">set_structure</span><span class="p">(</span><span class="n">structure</span><span class="p">)</span>
            <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">delete</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s2">&quot;.pdb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">tmp</span><span class="p">:</span>
                <span class="n">io</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">tmp</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">tmp</span><span class="o">.</span><span class="n">name</span>

        <span class="n">ref_path</span> <span class="o">=</span> <span class="n">save_structure_to_tempfile</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span>
        <span class="n">cut_path</span> <span class="o">=</span> <span class="n">save_structure_to_tempfile</span><span class="p">(</span><span class="n">cut</span><span class="p">)</span>
        <span class="c1"># 使用 usalign 计算 TM-score</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">usalign_bin</span><span class="p">),</span> <span class="n">ref_path</span><span class="p">,</span> <span class="n">cut_path</span><span class="p">,</span> <span class="s2">&quot;-outfmt&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;-mol&quot;</span><span class="p">,</span> <span class="s2">&quot;prot&quot;</span><span class="p">]</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">capture_output</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">text</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>  <span class="c1"># noqa: S603</span>

        <span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">returncode</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;USalign 执行失败: </span><span class="se">\n</span><span class="si">{</span><span class="n">result</span><span class="o">.</span><span class="n">stderr</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="n">lines</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">lines</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;USalign 输出异常, 结果缺失。&quot;</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="n">fields</span> <span class="o">=</span> <span class="n">lines</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="n">tm1</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
        <span class="n">tm2</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
        <span class="n">aln_info</span><span class="p">[</span><span class="s2">&quot;TM-score:mean&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">tm1</span> <span class="o">+</span> <span class="n">tm2</span><span class="p">)</span> <span class="o">/</span> <span class="mf">2.0</span>
        <span class="n">aln_info</span><span class="p">[</span><span class="s2">&quot;TM-score:TM1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tm1</span>
        <span class="n">aln_info</span><span class="p">[</span><span class="s2">&quot;TM-score:TM2&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tm2</span>
    <span class="k">return</span> <span class="n">aln_info</span></div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">ref_seq</span> <span class="o">=</span> <span class="s2">&quot;data/pdb/EGFP_4EUL_ref.fa&quot;</span>
    <span class="n">cut_seq</span> <span class="o">=</span> <span class="s2">&quot;data/pdb/EGFP_4EUL_cut1.fa&quot;</span>
    <span class="n">ref_pdb</span> <span class="o">=</span> <span class="s2">&quot;data/pdb/EGFP_4EUL_ref.pdb&quot;</span>
    <span class="n">cut_pdb</span> <span class="o">=</span> <span class="s2">&quot;data/pdb/EGFP_4EUL_cut1.pdb&quot;</span>
    <span class="n">ref_map_values</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;M_0&quot;</span><span class="p">:</span> <span class="mf">0.0</span><span class="p">,</span>
        <span class="s2">&quot;V_1&quot;</span><span class="p">:</span> <span class="mf">0.890295358649789</span><span class="p">,</span>
        <span class="s2">&quot;S_2&quot;</span><span class="p">:</span> <span class="mf">0.270042194092827</span><span class="p">,</span>
        <span class="s2">&quot;K_3&quot;</span><span class="p">:</span> <span class="mf">0.24050632911392406</span><span class="p">,</span>
        <span class="s2">&quot;G_4&quot;</span><span class="p">:</span> <span class="mf">0.8523206751054853</span><span class="p">,</span>
        <span class="s2">&quot;E_5&quot;</span><span class="p">:</span> <span class="mf">0.5991561181434599</span><span class="p">,</span>
        <span class="s2">&quot;K_238&quot;</span><span class="p">:</span> <span class="mf">0.5316455696202531</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">show_ref_cut</span><span class="p">(</span>
        <span class="n">ref_seq</span><span class="o">=</span><span class="n">ref_seq</span><span class="p">,</span>
        <span class="n">ref_pdb</span><span class="o">=</span><span class="n">ref_pdb</span><span class="p">,</span>
        <span class="n">cut_seq</span><span class="o">=</span><span class="n">cut_seq</span><span class="p">,</span>
        <span class="n">cut_pdb</span><span class="o">=</span><span class="n">cut_pdb</span><span class="p">,</span>
        <span class="n">ref_color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span>
        <span class="n">ref_map_colors</span><span class="o">=</span><span class="p">(</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="s2">&quot;blue&quot;</span><span class="p">),</span>
        <span class="n">ref_map_values</span><span class="o">=</span><span class="n">ref_map_values</span><span class="p">,</span>
        <span class="n">cut_color</span><span class="o">=</span><span class="s2">&quot;green&quot;</span><span class="p">,</span>
        <span class="n">gap_color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span>
        <span class="n">ref_style</span><span class="o">=</span><span class="s2">&quot;cartoon&quot;</span><span class="p">,</span>
        <span class="n">cut_style</span><span class="o">=</span><span class="s2">&quot;cartoon&quot;</span><span class="p">,</span>
        <span class="n">gap_style</span><span class="o">=</span><span class="s2">&quot;cartoon&quot;</span><span class="p">,</span>
        <span class="n">ref_map_value_random</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">output_fig</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">aln_info</span> <span class="o">=</span> <span class="n">get_cut2ref_aln_info</span><span class="p">(</span>
        <span class="n">ref</span><span class="o">=</span><span class="n">ref_pdb</span><span class="p">,</span>
        <span class="n">cut</span><span class="o">=</span><span class="n">cut_pdb</span><span class="p">,</span>
        <span class="n">cal_rmsd</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">cal_tmscore</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">label1</span><span class="o">=</span><span class="s2">&quot;ref&quot;</span><span class="p">,</span>
        <span class="n">label2</span><span class="o">=</span><span class="s2">&quot;cut&quot;</span><span class="p">,</span>
        <span class="n">usalign_bin</span><span class="o">=</span><span class="s2">&quot;/Users/zhaohuanan/micromamba/envs/ProEvaluator/bin/usalign&quot;</span><span class="p">,</span>
        <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">aln_info</span><span class="p">)</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>