

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bioat.lib.libpdb &mdash; bioat Documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=48106516"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            bioat
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">Modules</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">bioat</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">bioat.lib.libpdb</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for bioat.lib.libpdb</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">gzip</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">py3Dmol</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.PDB</span><span class="w"> </span><span class="kn">import</span> <span class="n">MMCIFParser</span><span class="p">,</span> <span class="n">PDBParser</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.Seq</span><span class="w"> </span><span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">Bio.SeqRecord</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqRecord</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.exceptions</span><span class="w"> </span><span class="kn">import</span> <span class="n">BioatFileFormatError</span><span class="p">,</span> <span class="n">BioatFileNotFoundError</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.lib.libalignment</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">get_aligned_seq</span><span class="p">,</span>
    <span class="n">get_best_alignment</span><span class="p">,</span>
    <span class="n">instantiate_pairwise_aligner</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.lib.libfastx</span><span class="w"> </span><span class="kn">import</span> <span class="n">AMINO_ACIDS</span><span class="p">,</span> <span class="n">AMINO_ACIDS_EXTEND</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioat.logger</span><span class="w"> </span><span class="kn">import</span> <span class="n">LoggerManager</span>

<span class="n">lm</span> <span class="o">=</span> <span class="n">LoggerManager</span><span class="p">(</span><span class="n">mod_name</span><span class="o">=</span><span class="s2">&quot;bioat.lib.libpdb&quot;</span><span class="p">)</span>


<div class="viewcode-block" id="show_ref_cut">
<a class="viewcode-back" href="../../../bioat.lib.html#bioat.lib.libpdb.show_ref_cut">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">show_ref_cut</span><span class="p">(</span>
    <span class="n">ref_seq</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cut_seq</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">ref_pdb</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cut_pdb</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ref_color</span><span class="o">=</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span>
    <span class="n">cut_color</span><span class="o">=</span><span class="s2">&quot;green&quot;</span><span class="p">,</span>
    <span class="n">gap_color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span>
    <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualizes the alignment of sequences and highlights changes in PDB structures using py3Dmol.</span>

<span class="sd">    Args:</span>
<span class="sd">        ref_seq (str): Amino acid sequence content for the ref protein.</span>
<span class="sd">        cut_seq (str): Amino acid sequence content for the cut protein</span>
<span class="sd">        ref_pdb (str): Path to the PDB file of the reference structure.</span>
<span class="sd">        cut_pdb (str): Path to the PDB file of the cut structure.</span>
<span class="sd">        gap_color (str): Color for gaps or removed residues.</span>
<span class="sd">        ref_color (str): Color for reference residues.</span>
<span class="sd">        cut_color (str): Color for cut residues.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">set_names</span><span class="p">(</span><span class="n">func_name</span><span class="o">=</span><span class="s2">&quot;show_ref_cut&quot;</span><span class="p">)</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">set_level</span><span class="p">(</span><span class="n">log_level</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">highlight_gaps_py3dmol</span><span class="p">(</span><span class="n">viewer</span><span class="p">,</span> <span class="n">pdb_file</span><span class="p">,</span> <span class="n">gaps</span><span class="p">,</span> <span class="n">color</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Highlights specific residue ranges in a PDB file with a given color.</span>

<span class="sd">        Args:</span>
<span class="sd">            viewer: py3Dmol.View instance.</span>
<span class="sd">            pdb_file (str): Path to the PDB file.</span>
<span class="sd">            gaps (list of tuples): List of (start, end) residue ranges to highlight.</span>
<span class="sd">            color (str): Color for highlighting.</span>
<span class="sd">            label (str): Label for the highlighted residues.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">pdb_file</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">pdb</span><span class="p">:</span>
            <span class="n">pdb_data</span> <span class="o">=</span> <span class="n">pdb</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
        <span class="n">viewer</span><span class="o">.</span><span class="n">addModel</span><span class="p">(</span><span class="n">pdb_data</span><span class="p">,</span> <span class="s2">&quot;pdb&quot;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span> <span class="ow">in</span> <span class="n">gaps</span><span class="p">:</span>
            <span class="n">viewer</span><span class="o">.</span><span class="n">setStyle</span><span class="p">(</span>
                <span class="p">{</span><span class="s2">&quot;resi&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">))},</span>
                <span class="p">{</span><span class="s2">&quot;stick&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="n">color</span><span class="p">}},</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">label</span><span class="p">:</span>
                <span class="n">viewer</span><span class="o">.</span><span class="n">addLabel</span><span class="p">(</span>
                    <span class="n">label</span><span class="p">,</span>
                    <span class="p">{</span><span class="s2">&quot;fontColor&quot;</span><span class="p">:</span> <span class="n">color</span><span class="p">,</span> <span class="s2">&quot;backgroundColor&quot;</span><span class="p">:</span> <span class="s2">&quot;white&quot;</span><span class="p">},</span>
                    <span class="p">{</span><span class="s2">&quot;resi&quot;</span><span class="p">:</span> <span class="n">start</span><span class="p">},</span>
                <span class="p">)</span>

    <span class="c1"># Perform alignment</span>
    <span class="n">aligner</span> <span class="o">=</span> <span class="n">instantiate_pairwise_aligner</span><span class="p">(</span>
        <span class="n">scoring_match</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
        <span class="n">penalty_mismatch</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">penalty_gap_open</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span>
        <span class="n">penalty_gap_extension</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">penalty_query_left_gap_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">penalty_query_right_gap_score</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">alignment</span> <span class="o">=</span> <span class="n">get_best_alignment</span><span class="p">(</span>
        <span class="n">seq_a</span><span class="o">=</span><span class="n">ref_seq</span><span class="p">,</span> <span class="n">seq_b</span><span class="o">=</span><span class="n">cut_seq</span><span class="p">,</span> <span class="n">aligner</span><span class="o">=</span><span class="n">aligner</span><span class="p">,</span> <span class="n">consider_strand</span><span class="o">=</span><span class="kc">False</span>
    <span class="p">)</span>
    <span class="n">alignment_dict</span> <span class="o">=</span> <span class="n">get_aligned_seq</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">letn_match</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Create mapping of aligned indices and gaps</span>
    <span class="n">ref_indices</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">aligned</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">cut_indices</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">aligned</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">gap_ranges</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="n">ref_indices</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="n">cut_indices</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ref_indices</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">cut_indices</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">ref_indices</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
    <span class="p">]</span>

    <span class="c1"># Visualize PDB structures using py3Dmol</span>
    <span class="n">viewer</span> <span class="o">=</span> <span class="n">py3Dmol</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="mi">400</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
    <span class="n">viewer</span><span class="o">.</span><span class="n">setStyle</span><span class="p">({</span><span class="s2">&quot;cartoon&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;color&quot;</span><span class="p">:</span> <span class="s2">&quot;spectrum&quot;</span><span class="p">}})</span>

    <span class="c1"># Add reference and cut structures</span>
    <span class="n">highlight_gaps_py3dmol</span><span class="p">(</span><span class="n">viewer</span><span class="p">,</span> <span class="n">ref_pdb</span><span class="p">,</span> <span class="p">[],</span> <span class="n">ref_color</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Reference&quot;</span><span class="p">)</span>
    <span class="n">highlight_gaps_py3dmol</span><span class="p">(</span><span class="n">viewer</span><span class="p">,</span> <span class="n">cut_pdb</span><span class="p">,</span> <span class="p">[],</span> <span class="n">cut_color</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Cut&quot;</span><span class="p">)</span>
    <span class="c1"># Highlight gaps</span>
    <span class="n">highlight_gaps_py3dmol</span><span class="p">(</span>
        <span class="n">viewer</span><span class="p">,</span> <span class="n">ref_pdb</span><span class="p">,</span> <span class="n">gap_ranges</span><span class="p">,</span> <span class="n">gap_color</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Deleted Residues&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Finalize the view</span>
    <span class="n">viewer</span><span class="o">.</span><span class="n">zoomTo</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">viewer</span><span class="o">.</span><span class="n">show</span><span class="p">()</span></div>



<div class="viewcode-block" id="pdb2fasta">
<a class="viewcode-back" href="../../../bioat.lib.html#bioat.lib.libpdb.pdb2fasta">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">pdb2fasta</span><span class="p">(</span><span class="n">pdb_file</span><span class="p">,</span> <span class="n">output_fasta</span><span class="p">,</span> <span class="n">log_level</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Converts a PDB / CIF file to a FASTA file.</span>

<span class="sd">        This function processes the provided PDB file and extracts protein, DNA,</span>
<span class="sd">        RNA sequences, and other molecules appropriately to create a FASTA file.</span>
<span class="sd">    e   1. Proteins:The protein sequence for each chain will be extracted as Chain X Protein.</span>
<span class="sd">        2. DNA and RNA: Bases for DNA (A, T, G, C) will be saved as Chain X DNA and bases for RNA (A, U, G, C) will be saved as Chain X RNA.</span>
<span class="sd">        3. Other molecules: Any unrecognized molecules (e.g., ions, modified molecules) will be labeled as [residue] and stored as Chain X Other molecules.</span>
<span class="sd">        4. Multi-chain complexes: The program supports multi-chain structures in complexes, and the content of each chain will be recorded separately.</span>

<span class="sd">        Args:</span>
<span class="sd">            pdb_file (str): input file path.</span>
<span class="sd">            output_fasta (str | None, optional): output file path. If None, the output file will be named as the basename of the input file with a &quot;.fa&quot; extension. Defaults to None.</span>
<span class="sd">            log_level (str, optional): log level. Defaults to &quot;WARNING&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">set_names</span><span class="p">(</span><span class="n">func_name</span><span class="o">=</span><span class="s2">&quot;pdb2fasta&quot;</span><span class="p">)</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">set_level</span><span class="p">(</span><span class="n">log_level</span><span class="p">)</span>

    <span class="c1"># 设置输出文件名</span>
    <span class="k">if</span> <span class="n">output_fasta</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">pdb_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.gz&quot;</span><span class="p">):</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">pdb_file</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">2</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">pdb_file</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">output_fasta</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">idx</span><span class="si">}</span><span class="s2">.fa&quot;</span> <span class="k">if</span> <span class="n">idx</span> <span class="k">else</span> <span class="s2">&quot;output.fa&quot;</span>
        <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Output is not defined, set output to </span><span class="si">{</span><span class="n">output_fasta</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">output_fasta</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.fa&quot;</span><span class="p">),</span> <span class="s2">&quot;Output file must have a &#39;.fa&#39; extension&quot;</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="s2">&quot;.&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">output_fasta</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.fa&quot;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

    <span class="c1"># 检查输入文件是否存在</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">pdb_file</span><span class="p">):</span>
        <span class="k">raise</span> <span class="n">BioatFileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Input file not found: </span><span class="si">{</span><span class="n">pdb_file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># 检测文件格式并选择解析器</span>
    <span class="k">if</span> <span class="n">pdb_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.cif&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">pdb_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.cif.gz&quot;</span><span class="p">):</span>
        <span class="n">parser</span> <span class="o">=</span> <span class="n">MMCIFParser</span><span class="p">(</span><span class="n">QUIET</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">pdb_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.pdb&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">pdb_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.pdb.gz&quot;</span><span class="p">):</span>
        <span class="n">parser</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">(</span><span class="n">QUIET</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">BioatFileFormatError</span><span class="p">(</span><span class="s2">&quot;Only .pdb and .cif files are allowed.&quot;</span><span class="p">)</span>

    <span class="c1"># 打开文件，支持 gzip 格式</span>
    <span class="n">input_handler</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">pdb_file</span><span class="p">,</span> <span class="s2">&quot;rt&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="n">pdb_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.gz&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="nb">open</span><span class="p">(</span><span class="n">pdb_file</span><span class="p">,</span> <span class="s2">&quot;rt&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">structure</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">input_handler</span><span class="p">)</span>

    <span class="n">records</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># 遍历结构中的所有模型和链，提取序列</span>
    <span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
            <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;processing chain </span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1"># 找出当前链中所有氨基酸的编号范围</span>
            <span class="n">residue_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">]</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">residue_ids</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="n">min_res_id</span><span class="p">,</span> <span class="n">max_res_id</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">residue_ids</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="n">residue_ids</span><span class="p">)</span>
            <span class="c1"># 分别初始化序列列表，使用 &quot;-&quot; 占位符填充缺失位置</span>
            <span class="n">protein_seq</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">max_res_id</span> <span class="o">-</span> <span class="n">min_res_id</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">dna_seq</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">max_res_id</span> <span class="o">-</span> <span class="n">min_res_id</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">rna_seq</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">max_res_id</span> <span class="o">-</span> <span class="n">min_res_id</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">other_seq</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">max_res_id</span> <span class="o">-</span> <span class="n">min_res_id</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">:</span>
                <span class="n">residue_name</span> <span class="o">=</span> <span class="n">residue</span><span class="o">.</span><span class="n">resname</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>  <span class="c1"># 标准化残基名称</span>
                <span class="n">residue_id</span> <span class="o">=</span> <span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">-</span> <span class="n">min_res_id</span>  <span class="c1"># 计算相对于最小编号的偏移</span>

                <span class="k">if</span> <span class="n">residue_name</span> <span class="ow">in</span> <span class="n">AMINO_ACIDS</span><span class="p">:</span>
                    <span class="c1"># 蛋白质序列，填充氨基酸到相应位置</span>
                    <span class="n">protein_seq</span><span class="p">[</span><span class="n">residue_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">AMINO_ACIDS</span><span class="p">[</span><span class="n">residue_name</span><span class="p">]</span>
                <span class="k">elif</span> <span class="n">residue_name</span> <span class="ow">in</span> <span class="n">AMINO_ACIDS_EXTEND</span><span class="p">:</span>
                    <span class="c1"># 蛋白质序列，填充氨基酸到相应位置</span>
                    <span class="n">protein_seq</span><span class="p">[</span><span class="n">residue_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">AMINO_ACIDS_EXTEND</span><span class="p">[</span><span class="n">residue_name</span><span class="p">]</span>
                <span class="k">elif</span> <span class="n">residue_name</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;DA&quot;</span><span class="p">,</span> <span class="s2">&quot;DT&quot;</span><span class="p">,</span> <span class="s2">&quot;DG&quot;</span><span class="p">,</span> <span class="s2">&quot;DC&quot;</span><span class="p">):</span>
                    <span class="c1"># DNA 序列，填充 A, T, G, C</span>
                    <span class="n">dna_seq</span><span class="p">[</span><span class="n">residue_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">residue_name</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="k">elif</span> <span class="n">residue_name</span> <span class="ow">in</span> <span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;U&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">):</span>
                    <span class="c1"># RNA 碱基</span>
                    <span class="n">rna_seq</span><span class="p">[</span><span class="n">residue_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">residue_name</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># 其他分子类型，用 [] 标记</span>
                    <span class="n">other_seq</span><span class="p">[</span><span class="n">residue_id</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">residue_name</span><span class="si">}</span><span class="s2">]&quot;</span>

            <span class="c1"># 根据链类型创建 SeqRecord，并添加到 records 列表</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">res</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">protein_seq</span><span class="p">):</span>
                <span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">SeqRecord</span><span class="p">(</span>
                        <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">protein_seq</span><span class="p">)),</span>
                        <span class="nb">id</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">structure</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">|Chain_</span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">|Protein&quot;</span><span class="p">,</span>
                        <span class="n">description</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Chain </span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> Protein&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">res</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">dna_seq</span><span class="p">):</span>
                <span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">SeqRecord</span><span class="p">(</span>
                        <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">)),</span>
                        <span class="nb">id</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">structure</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">|Chain_</span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">|DNA&quot;</span><span class="p">,</span>
                        <span class="n">description</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Chain </span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> DNA&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">res</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">rna_seq</span><span class="p">):</span>
                <span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">SeqRecord</span><span class="p">(</span>
                        <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">rna_seq</span><span class="p">)),</span>
                        <span class="nb">id</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">structure</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">|Chain_</span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">|RNA&quot;</span><span class="p">,</span>
                        <span class="n">description</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Chain </span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> RNA&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">res</span> <span class="o">!=</span> <span class="s2">&quot;-&quot;</span> <span class="k">for</span> <span class="n">res</span> <span class="ow">in</span> <span class="n">other_seq</span><span class="p">):</span>
                <span class="n">records</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">SeqRecord</span><span class="p">(</span>
                        <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">other_seq</span><span class="p">)),</span>
                        <span class="nb">id</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">structure</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">|Chain_</span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">|UNKNOWN&quot;</span><span class="p">,</span>
                        <span class="n">description</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;Chain </span><span class="si">{</span><span class="n">chain</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2"> Other molecules&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>

    <span class="c1"># 将所有链的序列保存为 FASTA 文件</span>
    <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">output_fasta</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
    <span class="n">input_handler</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">lm</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;FASTA saved to: </span><span class="si">{</span><span class="n">output_fasta</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>



<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="c1"># ref_seq</span>
    <span class="k">pass</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>